// @btc-vision/transaction/browser/btc-vision-bitcoin.js
import { d as I, B as h, a as U, t as c, i as Ut, b as Kn, e as Br, c as kt, v as Se, p as qe, u as N } from "./vendors.js";
const M = {
  messagePrefix: `Bitcoin Signed Message:
`,
  bech32: "bc",
  bech32Opnet: "op",
  bip32: {
    public: 76067358,
    private: 76066276
  },
  pubKeyHash: 0,
  scriptHash: 5,
  wif: 128
}, so = {
  messagePrefix: `Bitcoin Signed Message:
`,
  bech32: "bcrt",
  bech32Opnet: "opr",
  bip32: {
    public: 70617039,
    private: 70615956
  },
  pubKeyHash: 111,
  scriptHash: 196,
  wif: 239
}, io = {
  messagePrefix: `Bitcoin Signed Message:
`,
  bech32: "tb",
  bech32Opnet: "opt",
  bip32: {
    public: 70617039,
    private: 70615956
  },
  pubKeyHash: 111,
  scriptHash: 196,
  wif: 239
};
function Be(t) {
  let e, r;
  try {
    e = I.bech32.decode(t);
  } catch {
  }
  if (e) {
    if (r = e.words[0], r !== 0)
      throw new TypeError(t + " uses wrong encoding");
  } else if (e = I.bech32m.decode(t), r = e.words[0], r === 0)
    throw new TypeError(t + " uses wrong encoding");
  const n = I.bech32.fromWords(e.words.slice(1));
  return {
    version: r,
    prefix: e.prefix,
    data: h.from(n)
  };
}
function Mn(t) {
  if (t.length < 8 || t.length > 72 || t[0] !== 48 || t[1] !== t.length - 2 || t[2] !== 2)
    return !1;
  const e = t[3];
  if (e === 0 || 5 + e >= t.length || t[4 + e] !== 2)
    return !1;
  const r = t[5 + e];
  return !(r === 0 || 6 + e + r !== t.length || t[4] & 128 || e > 1 && t[4] === 0 && !(t[5] & 128) || t[e + 6] & 128 || r > 1 && t[e + 6] === 0 && !(t[e + 7] & 128));
}
function Dn(t) {
  if (t.length < 8)
    throw new Error("DER sequence length is too short");
  if (t.length > 72)
    throw new Error("DER sequence length is too long");
  if (t[0] !== 48)
    throw new Error("Expected DER sequence");
  if (t[1] !== t.length - 2)
    throw new Error("DER sequence length is invalid");
  if (t[2] !== 2)
    throw new Error("Expected DER integer");
  const e = t[3];
  if (e === 0)
    throw new Error("R length is zero");
  if (5 + e >= t.length)
    throw new Error("R length is too long");
  if (t[4 + e] !== 2)
    throw new Error("Expected DER integer (2)");
  const r = t[5 + e];
  if (r === 0)
    throw new Error("S length is zero");
  if (6 + e + r !== t.length)
    throw new Error("S length is invalid");
  if (t[4] & 128)
    throw new Error("R value is negative");
  if (e > 1 && t[4] === 0 && !(t[5] & 128))
    throw new Error("R value excessively padded");
  if (t[e + 6] & 128)
    throw new Error("S value is negative");
  if (r > 1 && t[e + 6] === 0 && !(t[e + 7] & 128))
    throw new Error("S value excessively padded");
  return {
    r: t.subarray(4, 4 + e),
    s: t.subarray(6 + e)
  };
}
function $n(t, e) {
  const r = t.length, n = e.length;
  if (r === 0)
    throw new Error("R length is zero");
  if (n === 0)
    throw new Error("S length is zero");
  if (r > 33)
    throw new Error("R length is too long");
  if (n > 33)
    throw new Error("S length is too long");
  if (t[0] & 128)
    throw new Error("R value is negative");
  if (e[0] & 128)
    throw new Error("S value is negative");
  if (r > 1 && t[0] === 0 && !(t[1] & 128))
    throw new Error("R value excessively padded");
  if (n > 1 && e[0] === 0 && !(e[1] & 128))
    throw new Error("S value excessively padded");
  const s = h.allocUnsafe(6 + r + n);
  return s[0] = 48, s[1] = s.length - 2, s[2] = 2, s[3] = t.length, t.copy(s, 4), s[4 + r] = 2, s[5 + r] = e.length, e.copy(s, 6 + r), s;
}
const E = {
  OP_FALSE: 0,
  OP_0: 0,
  OP_PUSHDATA1: 76,
  OP_PUSHDATA2: 77,
  OP_PUSHDATA4: 78,
  OP_1NEGATE: 79,
  OP_RESERVED: 80,
  OP_TRUE: 81,
  OP_1: 81,
  OP_2: 82,
  OP_3: 83,
  OP_4: 84,
  OP_5: 85,
  OP_6: 86,
  OP_7: 87,
  OP_8: 88,
  OP_9: 89,
  OP_10: 90,
  OP_11: 91,
  OP_12: 92,
  OP_13: 93,
  OP_14: 94,
  OP_15: 95,
  OP_16: 96,
  OP_NOP: 97,
  OP_VER: 98,
  OP_IF: 99,
  OP_NOTIF: 100,
  OP_VERIF: 101,
  OP_VERNOTIF: 102,
  OP_ELSE: 103,
  OP_ENDIF: 104,
  OP_VERIFY: 105,
  OP_RETURN: 106,
  OP_TOALTSTACK: 107,
  OP_FROMALTSTACK: 108,
  OP_2DROP: 109,
  OP_2DUP: 110,
  OP_3DUP: 111,
  OP_2OVER: 112,
  OP_2ROT: 113,
  OP_2SWAP: 114,
  OP_IFDUP: 115,
  OP_DEPTH: 116,
  OP_DROP: 117,
  OP_DUP: 118,
  OP_NIP: 119,
  OP_OVER: 120,
  OP_PICK: 121,
  OP_ROLL: 122,
  OP_ROT: 123,
  OP_SWAP: 124,
  OP_TUCK: 125,
  OP_CAT: 126,
  OP_SUBSTR: 127,
  OP_LEFT: 128,
  OP_RIGHT: 129,
  OP_SIZE: 130,
  OP_INVERT: 131,
  OP_AND: 132,
  OP_OR: 133,
  OP_XOR: 134,
  OP_EQUAL: 135,
  OP_EQUALVERIFY: 136,
  OP_RESERVED1: 137,
  OP_RESERVED2: 138,
  OP_1ADD: 139,
  OP_1SUB: 140,
  OP_2MUL: 141,
  OP_2DIV: 142,
  OP_NEGATE: 143,
  OP_ABS: 144,
  OP_NOT: 145,
  OP_0NOTEQUAL: 146,
  OP_ADD: 147,
  OP_SUB: 148,
  OP_MUL: 149,
  OP_DIV: 150,
  OP_MOD: 151,
  OP_LSHIFT: 152,
  OP_RSHIFT: 153,
  OP_BOOLAND: 154,
  OP_BOOLOR: 155,
  OP_NUMEQUAL: 156,
  OP_NUMEQUALVERIFY: 157,
  OP_NUMNOTEQUAL: 158,
  OP_LESSTHAN: 159,
  OP_GREATERTHAN: 160,
  OP_LESSTHANOREQUAL: 161,
  OP_GREATERTHANOREQUAL: 162,
  OP_MIN: 163,
  OP_MAX: 164,
  OP_WITHIN: 165,
  OP_RIPEMD160: 166,
  OP_SHA1: 167,
  OP_SHA256: 168,
  OP_HASH160: 169,
  OP_HASH256: 170,
  OP_CODESEPARATOR: 171,
  OP_CHECKSIG: 172,
  OP_CHECKSIGVERIFY: 173,
  OP_CHECKMULTISIG: 174,
  OP_CHECKMULTISIGVERIFY: 175,
  OP_NOP1: 176,
  OP_NOP2: 177,
  OP_CHECKLOCKTIMEVERIFY: 177,
  OP_NOP3: 178,
  OP_CHECKSEQUENCEVERIFY: 178,
  OP_NOP4: 179,
  OP_NOP5: 180,
  OP_NOP6: 181,
  OP_NOP7: 182,
  OP_NOP8: 183,
  OP_NOP9: 184,
  OP_NOP10: 185,
  OP_CHECKSIGADD: 186,
  OP_PUBKEYHASH: 253,
  OP_PUBKEY: 254,
  OP_INVALIDOPCODE: 255
}, Rr = {};
for (const t of Object.keys(E)) {
  const e = E[t];
  Rr[e] = t;
}
function Fr(t) {
  return t < E.OP_PUSHDATA1 ? 1 : t <= 255 ? 2 : t <= 65535 ? 3 : 5;
}
function qn(t, e, r) {
  const n = Fr(e);
  return n === 1 ? t.writeUInt8(e, r) : n === 2 ? (t.writeUInt8(E.OP_PUSHDATA1, r), t.writeUInt8(e, r + 1)) : n === 3 ? (t.writeUInt8(E.OP_PUSHDATA2, r), t.writeUInt16LE(e, r + 1)) : (t.writeUInt8(E.OP_PUSHDATA4, r), t.writeUInt32LE(e, r + 1)), n;
}
function Xn(t, e) {
  const r = t.readUInt8(e);
  let n, s;
  if (r < E.OP_PUSHDATA1)
    n = r, s = 1;
  else if (r === E.OP_PUSHDATA1) {
    if (e + 2 > t.length)
      return null;
    n = t.readUInt8(e + 1), s = 2;
  } else if (r === E.OP_PUSHDATA2) {
    if (e + 3 > t.length)
      return null;
    n = t.readUInt16LE(e + 1), s = 3;
  } else {
    if (e + 5 > t.length)
      return null;
    if (r !== E.OP_PUSHDATA4)
      throw new Error("Unexpected opcode");
    n = t.readUInt32LE(e + 1), s = 5;
  }
  return {
    opcode: r,
    number: n,
    size: s
  };
}
function jn(t, e, r) {
  e = e || 4, r = r === void 0 ? !0 : r;
  const n = t.length;
  if (n === 0)
    return 0;
  if (n > e)
    throw new TypeError("Script number overflow");
  if (r && (t[n - 1] & 127) === 0 && (n <= 1 || (t[n - 2] & 128) === 0))
    throw new Error("Non-minimally encoded script number");
  if (n === 5) {
    const i = t.readUInt32LE(0), o = t.readUInt8(4);
    return o & 128 ? -((o & -129) * 4294967296 + i) : o * 4294967296 + i;
  }
  let s = 0;
  for (let i = 0; i < n; ++i)
    s |= t[i] << 8 * i;
  return t[n - 1] & 128 ? -(s & ~(128 << 8 * (n - 1))) : s;
}
function zn(t) {
  return t > 2147483647 ? 5 : t > 8388607 ? 4 : t > 32767 ? 3 : t > 127 ? 2 : t > 0 ? 1 : 0;
}
function Lr(t) {
  let e = Math.abs(t);
  const r = zn(e), n = h.allocUnsafe(r), s = t < 0;
  for (let i = 0; i < r; ++i)
    n.writeUInt8(e & 255, i), e >>= 8;
  return n[r - 1] & 128 ? n.writeUInt8(s ? 128 : 0, r - 1) : s && (n[r - 1] |= 128), n;
}
const Yn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  decode: jn,
  encode: Lr
}, Symbol.toStringTag, { value: "Module" })), Xe = U.alloc(32, 0), je = U.from("fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f", "hex");
function at(t, e) {
  return t.length !== e.length ? !1 : t.every((r, n) => r.equals(e[n]));
}
function W(t) {
  if (!U.isBuffer(t) || t.length < 33)
    return !1;
  const e = t[0], r = t.subarray(1, 33);
  if (r.compare(Xe) === 0 || r.compare(je) >= 0)
    return !1;
  if ((e === 2 || e === 3) && t.length === 33)
    return !0;
  if (t.length !== 65)
    return !1;
  const n = t.subarray(33);
  return n.compare(Xe) === 0 || n.compare(je) >= 0 ? !1 : e === 4 || e === 6 || e === 7;
}
const Zn = 21 * 1e14;
function qt(t) {
  return c.UInt53(t) && t <= Zn;
}
const Pt = 254;
function ne(t) {
  if (!t || typeof t != "object" || !("output" in t))
    return !1;
  const e = t;
  return U.isBuffer(e.output) ? e.version !== void 0 ? (e.version & Pt) === e.version : !0 : !1;
}
function Re(t) {
  return globalThis.Array.isArray(t) ? t.length !== 2 ? !1 : t.every((e) => Re(e)) : ne(t);
}
const Qn = c.BufferN(32), Jn = c.BufferN(20), _e = c.BufferN(32), It = c.Number, se = c.Array, ts = c.Boolean, Vr = c.String, V = c.Buffer, Wr = c.Hex, Te = c.maybe, z = c.tuple, Gr = c.UInt8, Y = c.UInt32, es = c.Function, Kr = c.BufferN, Mr = c.Null, rs = c.oneOf, ie = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  Array: se,
  Boolean: ts,
  Buffer: V,
  Buffer256bit: Qn,
  BufferN: Kr,
  Function: es,
  Hash160bit: Jn,
  Hash256bit: _e,
  Hex: Wr,
  Null: Mr,
  Number: It,
  Satoshi: qt,
  String: Vr,
  TAPLEAF_VERSION_MASK: Pt,
  UInt32: Y,
  UInt8: Gr,
  isPoint: W,
  isTapleaf: ne,
  isTaptree: Re,
  maybe: Te,
  oneOf: rs,
  stacksEqual: at,
  tuple: z,
  typeforce: c
}, Symbol.toStringTag, { value: "Module" })), { typeforce: ns } = ie, ze = h.alloc(1, 0);
function oe(t) {
  const e = t & -129;
  return e > 0 && e < 4;
}
function Ye(t) {
  let e = 0;
  for (; t[e] === 0; )
    ++e;
  return e === t.length ? ze : (t = t.subarray(e), t[0] & 128 ? h.concat([ze, t], 1 + t.length) : t);
}
function Ze(t) {
  t[0] === 0 && (t = t.subarray(1));
  const e = h.alloc(32, 0), r = Math.max(0, 32 - t.length);
  return t.copy(e, r), e;
}
function ss(t) {
  const e = t.readUInt8(t.length - 1);
  if (!oe(e))
    throw new Error(`Invalid hashType ${e}`);
  const r = Dn(t.subarray(0, -1)), n = Ze(r.r), s = Ze(r.s);
  return { signature: h.concat([n, s], 64), hashType: e };
}
function is(t, e) {
  if (ns({
    signature: Kr(64),
    hashType: Gr
  }, { signature: t, hashType: e }), !oe(e))
    throw new Error(`Invalid hashType ${e}`);
  const r = h.allocUnsafe(1);
  r.writeUInt8(e, 0);
  const n = Ye(t.subarray(0, 32)), s = Ye(t.subarray(32, 64));
  return h.concat([$n(n, s), r]);
}
const os = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  decode: ss,
  encode: is,
  isDefinedHashType: oe
}, Symbol.toStringTag, { value: "Module" })), { typeforce: Nt } = ie, Dr = E.OP_RESERVED;
function us(t) {
  return It(t) && (t === E.OP_0 || t >= E.OP_1 && t <= E.OP_16 || t === E.OP_1NEGATE);
}
function $r(t) {
  return V(t) || us(t);
}
function ue(t) {
  return se(t) && t.every($r);
}
function Fe(t) {
  return t.length - t.filter($r).length;
}
function zt(t) {
  if (t.length === 0)
    return E.OP_0;
  if (t.length === 1) {
    if (t[0] >= 1 && t[0] <= 16)
      return Dr + t[0];
    if (t[0] === 129)
      return E.OP_1NEGATE;
  }
}
function qr(t) {
  return h.isBuffer(t);
}
function fs(t) {
  return se(t);
}
function Yt(t) {
  return h.isBuffer(t);
}
function v(t) {
  if (qr(t))
    return t;
  Nt(se, t);
  const e = t.reduce((s, i) => Yt(i) ? i.length === 1 && zt(i) !== void 0 ? s + 1 : s + Fr(i.length) + i.length : s + 1, 0), r = h.allocUnsafe(e);
  let n = 0;
  if (t.forEach((s) => {
    if (Yt(s)) {
      const i = zt(s);
      if (i !== void 0) {
        r.writeUInt8(i, n), n += 1;
        return;
      }
      n += qn(r, s.length, n), s.copy(r, n), n += s.length;
    } else
      r.writeUInt8(s, n), n += 1;
  }), n !== r.length)
    throw new Error("Could not decode chunks");
  return r;
}
function A(t) {
  if (fs(t))
    return t;
  Nt(V, t);
  const e = [];
  let r = 0;
  for (; r < t.length; ) {
    const n = t[r];
    if (n > E.OP_0 && n <= E.OP_PUSHDATA4) {
      const s = Xn(t, r);
      if (s === null || (r += s.size, r + s.number > t.length))
        return null;
      const i = t.subarray(r, r + s.number);
      r += s.number;
      const o = zt(i);
      o !== void 0 ? e.push(o) : e.push(i);
    } else
      e.push(n), r += 1;
  }
  return e;
}
function Xr(t) {
  if (qr(t) && (t = A(t)), !t)
    throw new Error("Could not convert invalid chunks to ASM");
  return t.map((e) => {
    if (Yt(e)) {
      const r = zt(e);
      if (r === void 0)
        return e.toString("hex");
      e = r;
    }
    return Rr[e];
  }).join(" ");
}
function cs(t) {
  return Nt(Vr, t), v(t.split(" ").map((e) => E[e] !== void 0 ? E[e] : (Nt(Wr, e), h.from(e, "hex"))));
}
function jr(t) {
  return t = A(t), Nt(ue, t), t.map((e) => Yt(e) ? e : e === E.OP_0 ? h.allocUnsafe(0) : Lr(e - Dr));
}
function zr(t) {
  return W(t);
}
function Q(t) {
  return !h.isBuffer(t) || !oe(t[t.length - 1]) ? !1 : Mn(t.subarray(0, -1));
}
const as = Yn, dt = os, hs = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  compile: v,
  countNonPushOnlyOPs: Fe,
  decompile: A,
  fromASM: cs,
  isCanonicalPubKey: zr,
  isCanonicalScriptSignature: Q,
  isPushOnly: ue,
  number: as,
  opcodes: E,
  signature: dt,
  toASM: Xr,
  toStack: jr
}, Symbol.toStringTag, { value: "Module" }));
var K;
(function(t) {
  t.P2PK = "p2pk", t.P2PKH = "p2pkh", t.P2SH = "p2sh", t.P2MS = "p2ms", t.P2WPKH = "p2wpkh", t.P2WSH = "p2wsh", t.P2TR = "p2tr", t.P2OP = "p2op", t.Embed = "embed", t.ScriptRedeem = "scriptRedeem";
})(K || (K = {}));
function y(t, e, r) {
  Object.defineProperty(t, e, {
    configurable: !0,
    enumerable: !0,
    get() {
      const n = r.call(this);
      return this[e] = n, n;
    },
    set(n) {
      Object.defineProperty(this, e, {
        configurable: !0,
        enumerable: !0,
        value: n,
        writable: !0
      });
    }
  });
}
function B(t) {
  let e;
  return () => (e !== void 0 || (e = t()), e);
}
const Qe = E, Je = 16, tr = 2, er = 40;
function Yr(t, e) {
  if (!t.address && !t.output && !t.program && (typeof t.deploymentVersion > "u" || !t.hash160))
    throw new TypeError("At least one of address, output or program must be provided");
  e = Object.assign({ validate: !0 }, e || {}), c({
    address: c.maybe(c.String),
    output: c.maybe(c.Buffer),
    program: c.maybe(c.Buffer),
    network: c.maybe(c.Object),
    deploymentVersion: c.maybe(c.Number),
    hash160: c.maybe(c.BufferN(20))
  }, t);
  const r = () => {
    if (typeof t.deploymentVersion < "u" && typeof t.hash160 < "u") {
      if (t.hash160.length !== 20)
        throw new TypeError("hash160 must be exactly 20 bytes");
      if (t.deploymentVersion < 0 || t.deploymentVersion > 255)
        throw new TypeError("deploymentVersion must fit in one byte");
      return h.concat([h.of(t.deploymentVersion), t.hash160]);
    }
  }, n = B(() => Be(t.address)), s = t.network || M, i = {
    name: K.P2OP,
    network: s,
    deploymentVersion: 0
  };
  if (y(i, "program", () => {
    if (t.program)
      return t.program;
    const o = r();
    if (o)
      return o;
    if (t.output) {
      if (t.output[0] !== Qe.OP_16)
        throw new TypeError("Invalid P2OP script");
      let u = 1, f;
      if (t.output[1] < 76)
        f = t.output[1], u = 2;
      else if (t.output[1] === 76)
        f = t.output[2], u = 3;
      else
        throw new TypeError("Unsupported push opcode in P2OP script");
      return t.output.subarray(u, u + f);
    }
    if (t.address)
      return n().data;
  }), y(i, "deploymentVersion", () => {
    if (i.program)
      return i.program[0];
  }), y(i, "hash160", () => {
    if (i.program)
      return i.program.subarray(1);
  }), y(i, "output", () => {
    if (i.program)
      return v([Qe.OP_16, i.program]);
  }), y(i, "address", () => {
    if (!i.program)
      return;
    if (!s.bech32Opnet)
      throw new TypeError("Network does not support opnet");
    const o = I.bech32m.toWords(i.program);
    return o.unshift(Je), I.bech32m.encode(s.bech32Opnet, o);
  }), e.validate) {
    let o = U.alloc(0);
    if (t.address) {
      const u = n();
      if (s.bech32Opnet !== u.prefix)
        throw new TypeError("Invalid prefix or network mismatch");
      if (u.version !== Je)
        throw new TypeError("Invalid witness version for p2op");
      if (u.data.length < tr || u.data.length > er)
        throw new TypeError("Invalid witness program length");
      o = u.data;
    }
    if (t.program) {
      if (o.length && !o.equals(t.program))
        throw new TypeError("Program mismatch");
      o = t.program;
    }
    if (!o.length && t.deploymentVersion !== void 0 && t.hash160 && (o = r()), t.output) {
      const u = i.program;
      if (o.length && !o.equals(u))
        throw new TypeError("Program mismatch (output vs other source)");
      o = u;
    }
    if (o.length < tr || o.length > er)
      throw new TypeError(`Witness program must be 2–40 bytes. Was ${o.length} bytes`);
    if (t.deploymentVersion !== void 0 && t.deploymentVersion !== o[0])
      throw new TypeError("deploymentVersion mismatch");
    if (t.hash160 && !t.hash160.equals(o.subarray(1)))
      throw new TypeError("hash160 mismatch");
  }
  return Object.assign(i, t);
}
function ps(t) {
  return t instanceof Uint8Array || ArrayBuffer.isView(t) && t.constructor.name === "Uint8Array";
}
function Zr(t, e, r = "") {
  const n = ps(t), s = t?.length;
  if (!n || e !== void 0) {
    const o = r && `"${r}" `, u = "", f = n ? `length=${s}` : `type=${typeof t}`;
    throw new Error(o + "expected Uint8Array" + u + ", got " + f);
  }
  return t;
}
function rr(t, e = !0) {
  if (t.destroyed)
    throw new Error("Hash instance has been destroyed");
  if (e && t.finished)
    throw new Error("Hash#digest() has already been called");
}
function ls(t, e) {
  Zr(t, void 0, "digestInto() output");
  const r = e.outputLen;
  if (t.length < r)
    throw new Error('"digestInto() output" expected to be of length >=' + r);
}
function ht(...t) {
  for (let e = 0; e < t.length; e++)
    t[e].fill(0);
}
function we(t) {
  return new DataView(t.buffer, t.byteOffset, t.byteLength);
}
function q(t, e) {
  return t << 32 - e | t >>> e;
}
function ft(t, e) {
  return t << e | t >>> 32 - e >>> 0;
}
function Le(t, e = {}) {
  const r = (s, i) => t(i).update(s).digest(), n = t(void 0);
  return r.outputLen = n.outputLen, r.blockLen = n.blockLen, r.create = (s) => t(s), Object.assign(r, e), Object.freeze(r);
}
const ds = (t) => ({
  oid: Uint8Array.from([6, 9, 96, 134, 72, 1, 101, 3, 4, 2, t])
});
function Qr(t, e, r) {
  return t & e ^ ~t & r;
}
function Jr(t, e, r) {
  return t & e ^ t & r ^ e & r;
}
class Ve {
  blockLen;
  outputLen;
  padOffset;
  isLE;
  // For partial updates less than block size
  buffer;
  view;
  finished = !1;
  length = 0;
  pos = 0;
  destroyed = !1;
  constructor(e, r, n, s) {
    this.blockLen = e, this.outputLen = r, this.padOffset = n, this.isLE = s, this.buffer = new Uint8Array(e), this.view = we(this.buffer);
  }
  update(e) {
    rr(this), Zr(e);
    const { view: r, buffer: n, blockLen: s } = this, i = e.length;
    for (let o = 0; o < i; ) {
      const u = Math.min(s - this.pos, i - o);
      if (u === s) {
        const f = we(e);
        for (; s <= i - o; o += s)
          this.process(f, o);
        continue;
      }
      n.set(e.subarray(o, o + u), this.pos), this.pos += u, o += u, this.pos === s && (this.process(r, 0), this.pos = 0);
    }
    return this.length += e.length, this.roundClean(), this;
  }
  digestInto(e) {
    rr(this), ls(e, this), this.finished = !0;
    const { buffer: r, view: n, blockLen: s, isLE: i } = this;
    let { pos: o } = this;
    r[o++] = 128, ht(this.buffer.subarray(o)), this.padOffset > s - o && (this.process(n, 0), o = 0);
    for (let l = o; l < s; l++)
      r[l] = 0;
    n.setBigUint64(s - 8, BigInt(this.length * 8), i), this.process(n, 0);
    const u = we(e), f = this.outputLen;
    if (f % 4)
      throw new Error("_sha2: outputLen must be aligned to 32bit");
    const a = f / 4, p = this.get();
    if (a > p.length)
      throw new Error("_sha2: outputLen bigger than state");
    for (let l = 0; l < a; l++)
      u.setUint32(4 * l, p[l], i);
  }
  digest() {
    const { buffer: e, outputLen: r } = this;
    this.digestInto(e);
    const n = e.slice(0, r);
    return this.destroy(), n;
  }
  _cloneInto(e) {
    e ||= new this.constructor(), e.set(...this.get());
    const { blockLen: r, buffer: n, length: s, finished: i, destroyed: o, pos: u } = this;
    return e.destroyed = o, e.finished = i, e.length = s, e.pos = u, s % r && e.buffer.set(n), e;
  }
  clone() {
    return this._cloneInto();
  }
}
const et = /* @__PURE__ */ Uint32Array.from([
  1779033703,
  3144134277,
  1013904242,
  2773480762,
  1359893119,
  2600822924,
  528734635,
  1541459225
]), St = /* @__PURE__ */ Uint32Array.from([
  1732584193,
  4023233417,
  2562383102,
  271733878,
  3285377520
]), rt = /* @__PURE__ */ new Uint32Array(80);
class ws extends Ve {
  A = St[0] | 0;
  B = St[1] | 0;
  C = St[2] | 0;
  D = St[3] | 0;
  E = St[4] | 0;
  constructor() {
    super(64, 20, 8, !1);
  }
  get() {
    const { A: e, B: r, C: n, D: s, E: i } = this;
    return [e, r, n, s, i];
  }
  set(e, r, n, s, i) {
    this.A = e | 0, this.B = r | 0, this.C = n | 0, this.D = s | 0, this.E = i | 0;
  }
  process(e, r) {
    for (let f = 0; f < 16; f++, r += 4)
      rt[f] = e.getUint32(r, !1);
    for (let f = 16; f < 80; f++)
      rt[f] = ft(rt[f - 3] ^ rt[f - 8] ^ rt[f - 14] ^ rt[f - 16], 1);
    let { A: n, B: s, C: i, D: o, E: u } = this;
    for (let f = 0; f < 80; f++) {
      let a, p;
      f < 20 ? (a = Qr(s, i, o), p = 1518500249) : f < 40 ? (a = s ^ i ^ o, p = 1859775393) : f < 60 ? (a = Jr(s, i, o), p = 2400959708) : (a = s ^ i ^ o, p = 3395469782);
      const l = ft(n, 5) + a + u + p + rt[f] | 0;
      u = o, o = i, i = ft(s, 30), s = n, n = l;
    }
    n = n + this.A | 0, s = s + this.B | 0, i = i + this.C | 0, o = o + this.D | 0, u = u + this.E | 0, this.set(n, s, i, o, u);
  }
  roundClean() {
    ht(rt);
  }
  destroy() {
    this.set(0, 0, 0, 0, 0), ht(this.buffer);
  }
}
const gs = /* @__PURE__ */ Le(() => new ws()), ys = /* @__PURE__ */ Uint8Array.from([
  7,
  4,
  13,
  1,
  10,
  6,
  15,
  3,
  12,
  0,
  9,
  5,
  2,
  14,
  11,
  8
]), tn = Uint8Array.from(new Array(16).fill(0).map((t, e) => e)), ms = tn.map((t) => (9 * t + 5) % 16), en = /* @__PURE__ */ (() => {
  const r = [[tn], [ms]];
  for (let n = 0; n < 4; n++)
    for (let s of r)
      s.push(s[n].map((i) => ys[i]));
  return r;
})(), rn = en[0], nn = en[1], sn = /* @__PURE__ */ [
  [11, 14, 15, 12, 5, 8, 7, 9, 11, 13, 14, 15, 6, 7, 9, 8],
  [12, 13, 11, 15, 6, 9, 9, 7, 12, 15, 11, 13, 7, 8, 7, 7],
  [13, 15, 14, 11, 7, 7, 6, 8, 13, 14, 13, 12, 5, 5, 6, 9],
  [14, 11, 12, 14, 8, 6, 5, 5, 15, 12, 15, 14, 9, 9, 8, 6],
  [15, 12, 13, 13, 9, 5, 8, 6, 14, 11, 12, 11, 8, 6, 5, 5]
].map((t) => Uint8Array.from(t)), bs = /* @__PURE__ */ rn.map((t, e) => t.map((r) => sn[e][r])), Es = /* @__PURE__ */ nn.map((t, e) => t.map((r) => sn[e][r])), Ss = /* @__PURE__ */ Uint32Array.from([
  0,
  1518500249,
  1859775393,
  2400959708,
  2840853838
]), _s = /* @__PURE__ */ Uint32Array.from([
  1352829926,
  1548603684,
  1836072691,
  2053994217,
  0
]);
function nr(t, e, r, n) {
  return t === 0 ? e ^ r ^ n : t === 1 ? e & r | ~e & n : t === 2 ? (e | ~r) ^ n : t === 3 ? e & n | r & ~n : e ^ (r | ~n);
}
const Wt = /* @__PURE__ */ new Uint32Array(16);
class Ts extends Ve {
  h0 = 1732584193;
  h1 = -271733879;
  h2 = -1732584194;
  h3 = 271733878;
  h4 = -1009589776;
  constructor() {
    super(64, 20, 8, !0);
  }
  get() {
    const { h0: e, h1: r, h2: n, h3: s, h4: i } = this;
    return [e, r, n, s, i];
  }
  set(e, r, n, s, i) {
    this.h0 = e | 0, this.h1 = r | 0, this.h2 = n | 0, this.h3 = s | 0, this.h4 = i | 0;
  }
  process(e, r) {
    for (let b = 0; b < 16; b++, r += 4)
      Wt[b] = e.getUint32(r, !0);
    let n = this.h0 | 0, s = n, i = this.h1 | 0, o = i, u = this.h2 | 0, f = u, a = this.h3 | 0, p = a, l = this.h4 | 0, d = l;
    for (let b = 0; b < 5; b++) {
      const w = 4 - b, S = Ss[b], P = _s[b], k = rn[b], H = nn[b], _ = bs[b], O = Es[b];
      for (let T = 0; T < 16; T++) {
        const F = ft(n + nr(b, i, u, a) + Wt[k[T]] + S, _[T]) + l | 0;
        n = l, l = a, a = ft(u, 10) | 0, u = i, i = F;
      }
      for (let T = 0; T < 16; T++) {
        const F = ft(s + nr(w, o, f, p) + Wt[H[T]] + P, O[T]) + d | 0;
        s = d, d = p, p = ft(f, 10) | 0, f = o, o = F;
      }
    }
    this.set(this.h1 + u + p | 0, this.h2 + a + d | 0, this.h3 + l + s | 0, this.h4 + n + o | 0, this.h0 + i + f | 0);
  }
  roundClean() {
    ht(Wt);
  }
  destroy() {
    this.destroyed = !0, ht(this.buffer), this.set(0, 0, 0, 0, 0);
  }
}
const on = /* @__PURE__ */ Le(() => new Ts()), Os = /* @__PURE__ */ Uint32Array.from([
  1116352408,
  1899447441,
  3049323471,
  3921009573,
  961987163,
  1508970993,
  2453635748,
  2870763221,
  3624381080,
  310598401,
  607225278,
  1426881987,
  1925078388,
  2162078206,
  2614888103,
  3248222580,
  3835390401,
  4022224774,
  264347078,
  604807628,
  770255983,
  1249150122,
  1555081692,
  1996064986,
  2554220882,
  2821834349,
  2952996808,
  3210313671,
  3336571891,
  3584528711,
  113926993,
  338241895,
  666307205,
  773529912,
  1294757372,
  1396182291,
  1695183700,
  1986661051,
  2177026350,
  2456956037,
  2730485921,
  2820302411,
  3259730800,
  3345764771,
  3516065817,
  3600352804,
  4094571909,
  275423344,
  430227734,
  506948616,
  659060556,
  883997877,
  958139571,
  1322822218,
  1537002063,
  1747873779,
  1955562222,
  2024104815,
  2227730452,
  2361852424,
  2428436474,
  2756734187,
  3204031479,
  3329325298
]), nt = /* @__PURE__ */ new Uint32Array(64);
class As extends Ve {
  constructor(e) {
    super(64, e, 8, !1);
  }
  get() {
    const { A: e, B: r, C: n, D: s, E: i, F: o, G: u, H: f } = this;
    return [e, r, n, s, i, o, u, f];
  }
  // prettier-ignore
  set(e, r, n, s, i, o, u, f) {
    this.A = e | 0, this.B = r | 0, this.C = n | 0, this.D = s | 0, this.E = i | 0, this.F = o | 0, this.G = u | 0, this.H = f | 0;
  }
  process(e, r) {
    for (let l = 0; l < 16; l++, r += 4)
      nt[l] = e.getUint32(r, !1);
    for (let l = 16; l < 64; l++) {
      const d = nt[l - 15], b = nt[l - 2], w = q(d, 7) ^ q(d, 18) ^ d >>> 3, S = q(b, 17) ^ q(b, 19) ^ b >>> 10;
      nt[l] = S + nt[l - 7] + w + nt[l - 16] | 0;
    }
    let { A: n, B: s, C: i, D: o, E: u, F: f, G: a, H: p } = this;
    for (let l = 0; l < 64; l++) {
      const d = q(u, 6) ^ q(u, 11) ^ q(u, 25), b = p + d + Qr(u, f, a) + Os[l] + nt[l] | 0, S = (q(n, 2) ^ q(n, 13) ^ q(n, 22)) + Jr(n, s, i) | 0;
      p = a, a = f, f = u, u = o + b | 0, o = i, i = s, s = n, n = b + S | 0;
    }
    n = n + this.A | 0, s = s + this.B | 0, i = i + this.C | 0, o = o + this.D | 0, u = u + this.E | 0, f = f + this.F | 0, a = a + this.G | 0, p = p + this.H | 0, this.set(n, s, i, o, u, f, a, p);
  }
  roundClean() {
    ht(nt);
  }
  destroy() {
    this.set(0, 0, 0, 0, 0, 0, 0, 0), ht(this.buffer);
  }
}
class Ps extends As {
  // We cannot use array here since array allows indexing by variable
  // which means optimizer/compiler cannot use registers.
  A = et[0] | 0;
  B = et[1] | 0;
  C = et[2] | 0;
  D = et[3] | 0;
  E = et[4] | 0;
  F = et[5] | 0;
  G = et[6] | 0;
  H = et[7] | 0;
  constructor() {
    super(32);
  }
}
const Zt = /* @__PURE__ */ Le(
  () => new Ps(),
  /* @__PURE__ */ ds(1)
);
function Is(t) {
  return h.from(on(Uint8Array.from(t)));
}
function Hs(t) {
  return h.from(gs(Uint8Array.from(t)));
}
function D(t) {
  return h.from(Zt(Uint8Array.from(t)));
}
function C(t) {
  return h.from(on(Zt(Uint8Array.from(t))));
}
function it(t) {
  return h.from(Zt(Zt(Uint8Array.from(t))));
}
const vs = [
  "BIP0340/challenge",
  "BIP0340/aux",
  "BIP0340/nonce",
  "TapLeaf",
  "TapBranch",
  "TapSighash",
  "TapTweak",
  "KeyAgg list",
  "KeyAgg coefficient"
], un = {
  "BIP0340/challenge": h.from([
    123,
    181,
    45,
    122,
    159,
    239,
    88,
    50,
    62,
    177,
    191,
    122,
    64,
    125,
    179,
    130,
    210,
    243,
    242,
    216,
    27,
    177,
    34,
    79,
    73,
    254,
    81,
    143,
    109,
    72,
    211,
    124,
    123,
    181,
    45,
    122,
    159,
    239,
    88,
    50,
    62,
    177,
    191,
    122,
    64,
    125,
    179,
    130,
    210,
    243,
    242,
    216,
    27,
    177,
    34,
    79,
    73,
    254,
    81,
    143,
    109,
    72,
    211,
    124
  ]),
  "BIP0340/aux": h.from([
    241,
    239,
    78,
    94,
    192,
    99,
    202,
    218,
    109,
    148,
    202,
    250,
    157,
    152,
    126,
    160,
    105,
    38,
    88,
    57,
    236,
    193,
    31,
    151,
    45,
    119,
    165,
    46,
    216,
    193,
    204,
    144,
    241,
    239,
    78,
    94,
    192,
    99,
    202,
    218,
    109,
    148,
    202,
    250,
    157,
    152,
    126,
    160,
    105,
    38,
    88,
    57,
    236,
    193,
    31,
    151,
    45,
    119,
    165,
    46,
    216,
    193,
    204,
    144
  ]),
  "BIP0340/nonce": h.from([
    7,
    73,
    119,
    52,
    167,
    155,
    203,
    53,
    91,
    155,
    140,
    125,
    3,
    79,
    18,
    28,
    244,
    52,
    215,
    62,
    247,
    45,
    218,
    25,
    135,
    0,
    97,
    251,
    82,
    191,
    235,
    47,
    7,
    73,
    119,
    52,
    167,
    155,
    203,
    53,
    91,
    155,
    140,
    125,
    3,
    79,
    18,
    28,
    244,
    52,
    215,
    62,
    247,
    45,
    218,
    25,
    135,
    0,
    97,
    251,
    82,
    191,
    235,
    47
  ]),
  TapLeaf: h.from([
    174,
    234,
    143,
    220,
    66,
    8,
    152,
    49,
    5,
    115,
    75,
    88,
    8,
    29,
    30,
    38,
    56,
    211,
    95,
    28,
    181,
    64,
    8,
    212,
    211,
    87,
    202,
    3,
    190,
    120,
    233,
    238,
    174,
    234,
    143,
    220,
    66,
    8,
    152,
    49,
    5,
    115,
    75,
    88,
    8,
    29,
    30,
    38,
    56,
    211,
    95,
    28,
    181,
    64,
    8,
    212,
    211,
    87,
    202,
    3,
    190,
    120,
    233,
    238
  ]),
  TapBranch: h.from([
    25,
    65,
    161,
    242,
    229,
    110,
    185,
    95,
    162,
    169,
    241,
    148,
    190,
    92,
    1,
    247,
    33,
    111,
    51,
    237,
    130,
    176,
    145,
    70,
    52,
    144,
    208,
    91,
    245,
    22,
    160,
    21,
    25,
    65,
    161,
    242,
    229,
    110,
    185,
    95,
    162,
    169,
    241,
    148,
    190,
    92,
    1,
    247,
    33,
    111,
    51,
    237,
    130,
    176,
    145,
    70,
    52,
    144,
    208,
    91,
    245,
    22,
    160,
    21
  ]),
  TapSighash: h.from([
    244,
    10,
    72,
    223,
    75,
    42,
    112,
    200,
    180,
    146,
    75,
    242,
    101,
    70,
    97,
    237,
    61,
    149,
    253,
    102,
    163,
    19,
    235,
    135,
    35,
    117,
    151,
    198,
    40,
    228,
    160,
    49,
    244,
    10,
    72,
    223,
    75,
    42,
    112,
    200,
    180,
    146,
    75,
    242,
    101,
    70,
    97,
    237,
    61,
    149,
    253,
    102,
    163,
    19,
    235,
    135,
    35,
    117,
    151,
    198,
    40,
    228,
    160,
    49
  ]),
  TapTweak: h.from([
    232,
    15,
    225,
    99,
    156,
    156,
    160,
    80,
    227,
    175,
    27,
    57,
    193,
    67,
    198,
    62,
    66,
    156,
    188,
    235,
    21,
    217,
    64,
    251,
    181,
    197,
    161,
    244,
    175,
    87,
    197,
    233,
    232,
    15,
    225,
    99,
    156,
    156,
    160,
    80,
    227,
    175,
    27,
    57,
    193,
    67,
    198,
    62,
    66,
    156,
    188,
    235,
    21,
    217,
    64,
    251,
    181,
    197,
    161,
    244,
    175,
    87,
    197,
    233
  ]),
  "KeyAgg list": h.from([
    72,
    28,
    151,
    28,
    60,
    11,
    70,
    215,
    240,
    178,
    117,
    174,
    89,
    141,
    78,
    44,
    126,
    215,
    49,
    156,
    89,
    74,
    92,
    110,
    199,
    158,
    160,
    212,
    153,
    2,
    148,
    240,
    72,
    28,
    151,
    28,
    60,
    11,
    70,
    215,
    240,
    178,
    117,
    174,
    89,
    141,
    78,
    44,
    126,
    215,
    49,
    156,
    89,
    74,
    92,
    110,
    199,
    158,
    160,
    212,
    153,
    2,
    148,
    240
  ]),
  "KeyAgg coefficient": h.from([
    191,
    201,
    4,
    3,
    77,
    28,
    136,
    232,
    200,
    14,
    34,
    229,
    61,
    36,
    86,
    109,
    100,
    130,
    78,
    214,
    66,
    114,
    129,
    192,
    145,
    0,
    249,
    77,
    205,
    82,
    201,
    129,
    191,
    201,
    4,
    3,
    77,
    28,
    136,
    232,
    200,
    14,
    34,
    229,
    61,
    36,
    86,
    109,
    100,
    130,
    78,
    214,
    66,
    114,
    129,
    192,
    145,
    0,
    249,
    77,
    205,
    82,
    201,
    129
  ])
};
function Rt(t, e) {
  return D(h.concat([un[t], e]));
}
const xs = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  TAGGED_HASH_PREFIXES: un,
  TAGS: vs,
  hash160: C,
  hash256: it,
  ripemd160: Is,
  sha1: Hs,
  sha256: D,
  taggedHash: Rt
}, Symbol.toStringTag, { value: "Module" }));
const Us = {
  p: 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2fn,
  n: 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141n,
  b: 7n,
  Gx: 0x79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798n,
  Gy: 0x483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8n
}, { p: ct, n: fn, Gx: ks, Gy: Ns, b: cn } = Us, Xt = 32, Oe = 64, G = (t = "") => {
  throw new Error(t);
}, an = (t) => typeof t == "bigint", hn = (t) => typeof t == "string", Cs = (t) => t instanceof Uint8Array || ArrayBuffer.isView(t) && t.constructor.name === "Uint8Array", Ct = (t, e) => !Cs(t) || typeof e == "number" && e > 0 && t.length !== e ? G("Uint8Array expected") : t, pn = (t) => new Uint8Array(t), Bs = (t) => Uint8Array.from(t), ln = (t, e) => t.toString(16).padStart(e, "0"), dn = (t) => Array.from(Ct(t)).map((e) => ln(e, 2)).join(""), J = { _0: 48, _9: 57, A: 65, F: 70, a: 97, f: 102 }, sr = (t) => {
  if (t >= J._0 && t <= J._9)
    return t - J._0;
  if (t >= J.A && t <= J.F)
    return t - (J.A - 10);
  if (t >= J.a && t <= J.f)
    return t - (J.a - 10);
}, wn = (t) => {
  const e = "hex invalid";
  if (!hn(t))
    return G(e);
  const r = t.length, n = r / 2;
  if (r % 2)
    return G(e);
  const s = pn(n);
  for (let i = 0, o = 0; i < n; i++, o += 2) {
    const u = sr(t.charCodeAt(o)), f = sr(t.charCodeAt(o + 1));
    if (u === void 0 || f === void 0)
      return G(e);
    s[i] = u * 16 + f;
  }
  return s;
}, gn = (t, e) => Ct(hn(t) ? wn(t) : Bs(Ct(t)), e), ir = (...t) => {
  const e = pn(t.reduce((n, s) => n + Ct(s).length, 0));
  let r = 0;
  return t.forEach((n) => {
    e.set(n, r), r += n.length;
  }), e;
}, Qt = BigInt, Ft = (t, e, r, n = "bad number: out of range") => an(t) && e <= t && t < r ? t : G(n), m = (t, e = ct) => {
  const r = t % e;
  return r >= 0n ? r : e + r;
}, Rs = (t, e) => {
  (t === 0n || e <= 0n) && G("no inverse n=" + t + " mod=" + e);
  let r = m(t, e), n = e, s = 0n, i = 1n;
  for (; r !== 0n; ) {
    const o = n / r, u = n % r, f = s - i * o;
    n = r, r = u, s = i, i = f;
  }
  return n === 1n ? m(s, e) : G("no inverse");
}, or = (t) => t instanceof R ? t : G("Point expected"), yn = (t) => m(m(t * t) * t + cn), ur = (t) => Ft(t, 0n, ct), jt = (t) => Ft(t, 1n, ct), Fs = (t) => Ft(t, 1n, fn), Ae = (t) => (t & 1n) === 0n, mn = (t) => Uint8Array.of(t), Ls = (t) => mn(Ae(t) ? 2 : 3), Vs = (t) => {
  const e = yn(jt(t));
  let r = 1n;
  for (let n = e, s = (ct + 1n) / 4n; s > 0n; s >>= 1n)
    s & 1n && (r = r * n % ct), n = n * n % ct;
  return m(r * r) === e ? r : G("sqrt invalid");
};
class R {
  static BASE;
  static ZERO;
  px;
  py;
  pz;
  constructor(e, r, n) {
    this.px = ur(e), this.py = jt(r), this.pz = ur(n), Object.freeze(this);
  }
  /** Convert Uint8Array or hex string to Point. */
  static fromBytes(e) {
    Ct(e);
    let r;
    const n = e[0], s = e.subarray(1), i = fr(s, 0, Xt), o = e.length;
    if (o === Xt + 1 && [2, 3].includes(n)) {
      let u = Vs(i);
      const f = Ae(u);
      Ae(Qt(n)) !== f && (u = m(-u)), r = new R(i, u, 1n);
    }
    return o === Oe + 1 && n === 4 && (r = new R(i, fr(s, Xt, Oe), 1n)), r ? r.assertValidity() : G("bad point: not on curve");
  }
  /** Equality check: compare points P&Q. */
  equals(e) {
    const { px: r, py: n, pz: s } = this, { px: i, py: o, pz: u } = or(e), f = m(r * u), a = m(i * s), p = m(n * u), l = m(o * s);
    return f === a && p === l;
  }
  is0() {
    return this.equals(ut);
  }
  /** Flip point over y coordinate. */
  negate() {
    return new R(this.px, m(-this.py), this.pz);
  }
  /** Point doubling: P+P, complete formula. */
  double() {
    return this.add(this);
  }
  /**
   * Point addition: P+Q, complete, exception-free formula
   * (Renes-Costello-Batina, algo 1 of [2015/1060](https://eprint.iacr.org/2015/1060)).
   * Cost: `12M + 0S + 3*a + 3*b3 + 23add`.
   */
  // prettier-ignore
  add(e) {
    const { px: r, py: n, pz: s } = this, { px: i, py: o, pz: u } = or(e), f = 0n, a = cn;
    let p = 0n, l = 0n, d = 0n;
    const b = m(a * 3n);
    let w = m(r * i), S = m(n * o), P = m(s * u), k = m(r + n), H = m(i + o);
    k = m(k * H), H = m(w + S), k = m(k - H), H = m(r + s);
    let _ = m(i + u);
    return H = m(H * _), _ = m(w + P), H = m(H - _), _ = m(n + s), p = m(o + u), _ = m(_ * p), p = m(S + P), _ = m(_ - p), d = m(f * H), p = m(b * P), d = m(p + d), p = m(S - d), d = m(S + d), l = m(p * d), S = m(w + w), S = m(S + w), P = m(f * P), H = m(b * H), S = m(S + P), P = m(w - P), P = m(f * P), H = m(H + P), w = m(S * H), l = m(l + w), w = m(_ * H), p = m(k * p), p = m(p - w), w = m(k * S), d = m(_ * d), d = m(d + w), new R(p, l, d);
  }
  /**
   * Point-by-scalar multiplication. Scalar must be in range 1 <= n < CURVE.n.
   * Uses {@link wNAF} for base point.
   * Uses fake point to mitigate side-channel leakage.
   * @param n scalar by which point is multiplied
   * @param safe safe mode guards against timing attacks; unsafe mode is faster
   */
  multiply(e, r = !0) {
    if (!r && e === 0n)
      return ut;
    if (Fs(e), e === 1n)
      return this;
    if (this.equals(wt))
      return Ds(e).p;
    let n = ut, s = wt;
    for (let i = this; e > 0n; i = i.double(), e >>= 1n)
      e & 1n ? n = n.add(i) : r && (s = s.add(i));
    return n;
  }
  /** Convert point to 2d xy affine point. (X, Y, Z) ∋ (x=X/Z, y=Y/Z) */
  toAffine() {
    const { px: e, py: r, pz: n } = this;
    if (this.equals(ut))
      return { x: 0n, y: 0n };
    if (n === 1n)
      return { x: e, y: r };
    const s = Rs(n, ct);
    return m(n * s) !== 1n && G("inverse invalid"), { x: m(e * s), y: m(r * s) };
  }
  /** Checks if the point is valid and on-curve. */
  assertValidity() {
    const { x: e, y: r } = this.toAffine();
    return jt(e), jt(r), m(r * r) === yn(e) ? this : G("bad point: not on curve");
  }
  /** Converts point to 33/65-byte Uint8Array. */
  toBytes(e = !0) {
    const { x: r, y: n } = this.assertValidity().toAffine(), s = cr(r);
    return e ? ir(Ls(n), s) : ir(mn(4), s, cr(n));
  }
  /** Create 3d xyz point from 2d xy. (0, 0) => (0, 1, 0), not (0, 0, 1) */
  static fromAffine(e) {
    const { x: r, y: n } = e;
    return r === 0n && n === 0n ? ut : new R(r, n, 1n);
  }
  toHex(e) {
    return dn(this.toBytes(e));
  }
  static fromPrivateKey(e) {
    return wt.multiply(Gs(e));
  }
  static fromHex(e) {
    return R.fromBytes(gn(e));
  }
  get x() {
    return this.toAffine().x;
  }
  get y() {
    return this.toAffine().y;
  }
  toRawBytes(e) {
    return this.toBytes(e);
  }
}
const wt = new R(ks, Ns, 1n), ut = new R(0n, 1n, 0n);
R.BASE = wt;
R.ZERO = ut;
const bn = (t) => Qt("0x" + (dn(t) || "0")), fr = (t, e, r) => bn(t.subarray(e, r)), Ws = 2n ** 256n, cr = (t) => wn(ln(Ft(t, 0n, Ws), Oe)), Gs = (t) => {
  const e = an(t) ? t : bn(gn(t, Xt));
  return Ft(e, 1n, fn, "private key invalid 3");
}, Jt = 8, Ks = 256, En = Math.ceil(Ks / Jt) + 1, Pe = 2 ** (Jt - 1), Ms = () => {
  const t = [];
  let e = wt, r = e;
  for (let n = 0; n < En; n++) {
    r = e, t.push(r);
    for (let s = 1; s < Pe; s++)
      r = r.add(e), t.push(r);
    e = r.double();
  }
  return t;
};
let ar;
const hr = (t, e) => {
  const r = e.negate();
  return t ? r : e;
}, Ds = (t) => {
  const e = ar || (ar = Ms());
  let r = ut, n = wt;
  const s = 2 ** Jt, i = s, o = Qt(s - 1), u = Qt(Jt);
  for (let f = 0; f < En; f++) {
    let a = Number(t & o);
    t >>= u, a > Pe && (a -= i, t += 1n);
    const p = f * Pe, l = p, d = p + Math.abs(a) - 1, b = f % 2 !== 0, w = a < 0;
    a === 0 ? n = n.add(hr(b, e[l])) : r = r.add(hr(w, e[d]));
  }
  return { p: r, f: n };
}, Ht = (t) => {
  const e = t.length === 32 ? t : t.subarray(1, 33);
  return h.isBuffer(e) ? e : h.from(e);
};
function pr(t) {
  let e = t.toString(16);
  return e = e.padStart(64, "0"), e.length > 64 && (e = e.slice(-64)), h.from(e, "hex");
}
function Ie(t) {
  if (t.length === 32)
    return;
  if (![33, 65].includes(t.length)) {
    console.warn(`Unsupported key length=${t.length}. Must be 33 (compressed) or 65 (uncompressed).`);
    return;
  }
  let e;
  try {
    e = R.fromHex(t);
  } catch {
    throw new Error("Invalid secp256k1 public key bytes. Cannot parse.");
  }
  const r = pr(e.x), n = pr(e.y), i = e.y % 2n === 0n ? 6 : 7, o = h.alloc(65);
  o[0] = i, r.copy(o, 1), n.copy(o, 33);
  const u = h.concat([h.from([4]), r, n]);
  return {
    hybrid: o,
    uncompressed: u
  };
}
function Gt(t, e) {
  if (t.equals(e))
    return !0;
  if (t.length === 65 && e.length === 65) {
    const r = h.from(t), n = h.from(e);
    return (r[0] === 6 || r[0] === 7) && (r[0] = 4), (n[0] === 6 || n[0] === 7) && (n[0] = 4), r.equals(n);
  }
  return !1;
}
const st = E;
function yt(t, e) {
  if (!t.address && !t.hash && !t.output && !t.pubkey && !t.input)
    throw new TypeError("Not enough data");
  e = Object.assign({ validate: !0 }, e || {}), c({
    network: c.maybe(c.Object),
    address: c.maybe(c.String),
    hash: c.maybe(c.BufferN(20)),
    output: c.maybe(c.BufferN(25)),
    pubkey: c.maybe(W),
    signature: c.maybe(Q),
    input: c.maybe(c.Buffer)
  }, t);
  const r = B(() => {
    const o = h.from(Ut.decode(t.address)), u = o.readUInt8(0), f = o.subarray(1);
    return { version: u, hash: f };
  }), n = B(() => A(t.input)), s = t.network || M, i = {
    name: K.P2PKH,
    network: s,
    hash: void 0
  };
  if (y(i, "address", () => {
    if (!i.hash)
      return;
    const o = h.allocUnsafe(21);
    return o.writeUInt8(s.pubKeyHash, 0), i.hash.copy(o, 1), Ut.encode(o);
  }), y(i, "hash", () => {
    if (t.output)
      return t.output.subarray(3, 23);
    if (t.address)
      return r().hash;
    if (t.pubkey || i.pubkey)
      return C(t.pubkey || i.pubkey);
  }), y(i, "output", () => {
    if (i.hash)
      return v([
        st.OP_DUP,
        st.OP_HASH160,
        i.hash,
        st.OP_EQUALVERIFY,
        st.OP_CHECKSIG
      ]);
  }), y(i, "pubkey", () => {
    if (t.input)
      return n()[1];
  }), y(i, "signature", () => {
    if (t.input)
      return n()[0];
  }), y(i, "input", () => {
    if (!t.pubkey || !t.signature)
      return;
    let o = t.pubkey;
    if (t.useHybrid || t.useUncompressed) {
      const u = Ie(t.pubkey);
      u && (t.useUncompressed ? o = u.uncompressed : o = u.hybrid);
    }
    return v([t.signature, o]);
  }), y(i, "witness", () => {
    if (i.input)
      return [];
  }), e.validate) {
    let o = h.from([]);
    if (t.address) {
      if (r().version !== s.pubKeyHash)
        throw new TypeError("Invalid version or Network mismatch");
      if (r().hash.length !== 20)
        throw new TypeError("Invalid address");
      o = r().hash;
    }
    if (t.hash) {
      if (o.length > 0 && !o.equals(t.hash))
        throw new TypeError("Hash mismatch");
      o = t.hash;
    }
    if (t.output) {
      if (t.output.length !== 25 || t.output[0] !== st.OP_DUP || t.output[1] !== st.OP_HASH160 || t.output[2] !== 20 || t.output[23] !== st.OP_EQUALVERIFY || t.output[24] !== st.OP_CHECKSIG)
        throw new TypeError("Output is invalid");
      const u = t.output.subarray(3, 23);
      if (o.length > 0 && !o.equals(u))
        throw new TypeError("Hash mismatch");
      o = u;
    }
    if (t.pubkey) {
      const u = C(t.pubkey);
      let f = o.length > 0 && !o.equals(u);
      if (f && (t.pubkey.length === 33 && (t.pubkey[0] === 2 || t.pubkey[0] === 3) || t.pubkey.length === 65 && t.pubkey[0] === 4)) {
        const a = Ie(t.pubkey);
        if (a) {
          const p = C(a.uncompressed);
          if (o.equals(p))
            f = !1, t.useUncompressed = !0;
          else {
            const l = C(a.hybrid);
            f = !o.equals(l), f || (t.useHybrid = !0);
          }
        }
      }
      if (f)
        throw new TypeError("Hash mismatch");
      o = u;
    }
    if (t.input) {
      const u = n();
      if (u.length !== 2)
        throw new TypeError("Input is invalid");
      if (!Q(u[0]))
        throw new TypeError("Input has invalid signature");
      if (!W(u[1]))
        throw new TypeError("Input has invalid pubkey");
      if (t.signature && !t.signature.equals(u[0]))
        throw new TypeError("Signature mismatch");
      if (t.pubkey && !t.pubkey.equals(u[1]))
        throw new TypeError("Pubkey mismatch");
      const f = C(u[1]);
      if (o.length > 0 && !o.equals(f))
        throw new TypeError("Hash mismatch (input)");
    }
  }
  return Object.assign(i, t);
}
const _t = E;
function mt(t, e) {
  if (!t.address && !t.hash && !t.output && !t.redeem && !t.input)
    throw new TypeError("Not enough data");
  e = Object.assign({ validate: !0 }, e || {}), c({
    network: c.maybe(c.Object),
    address: c.maybe(c.String),
    hash: c.maybe(c.BufferN(20)),
    output: c.maybe(c.BufferN(23)),
    redeem: c.maybe({
      network: c.maybe(c.Object),
      output: c.maybe(c.Buffer),
      input: c.maybe(c.Buffer),
      witness: c.maybe(c.arrayOf(c.Buffer))
    }),
    input: c.maybe(c.Buffer),
    witness: c.maybe(c.arrayOf(c.Buffer))
  }, t);
  let r = t.network;
  r || (r = t.redeem && t.redeem.network || M);
  const n = {
    network: r,
    name: K.P2SH
  }, s = B(() => {
    const u = h.from(Ut.decode(t.address)), f = u.readUInt8(0), a = u.subarray(1);
    return { version: f, hash: a };
  }), i = B(() => A(t.input)), o = B(() => {
    const u = i(), f = u[u.length - 1];
    return {
      network: r,
      output: f === _t.OP_FALSE ? h.from([]) : f,
      input: v(u.slice(0, -1)),
      witness: t.witness || []
    };
  });
  if (y(n, "address", () => {
    if (!n.hash)
      return;
    const u = h.allocUnsafe(21);
    return u.writeUInt8(n.network.scriptHash, 0), n.hash.copy(u, 1), Ut.encode(u);
  }), y(n, "hash", () => {
    if (t.output)
      return t.output.subarray(2, 22);
    if (t.address)
      return s().hash;
    if (n.redeem && n.redeem.output)
      return C(n.redeem.output);
  }), y(n, "output", () => {
    if (n.hash)
      return v([_t.OP_HASH160, n.hash, _t.OP_EQUAL]);
  }), y(n, "redeem", () => {
    if (t.input)
      return o();
  }), y(n, "input", () => {
    if (!(!t.redeem || !t.redeem.input || !t.redeem.output))
      return v([].concat(A(t.redeem.input), t.redeem.output));
  }), y(n, "witness", () => {
    if (n.redeem && n.redeem.witness)
      return n.redeem.witness;
    if (n.input)
      return [];
  }), y(n, "name", () => {
    const u = ["p2sh"];
    return n.redeem !== void 0 && n.redeem.name !== void 0 && u.push(n.redeem.name), u.join("-");
  }), e.validate) {
    let u = h.from([]);
    if (t.address) {
      if (s().version !== r.scriptHash)
        throw new TypeError("Invalid version or Network mismatch");
      if (s().hash.length !== 20)
        throw new TypeError("Invalid address");
      u = s().hash;
    }
    if (t.hash) {
      if (u.length > 0 && !u.equals(t.hash))
        throw new TypeError("Hash mismatch");
      u = t.hash;
    }
    if (t.output) {
      if (t.output.length !== 23 || t.output[0] !== _t.OP_HASH160 || t.output[1] !== 20 || t.output[22] !== _t.OP_EQUAL)
        throw new TypeError("Output is invalid");
      const a = t.output.subarray(2, 22);
      if (u.length > 0 && !u.equals(a))
        throw new TypeError("Hash mismatch");
      u = a;
    }
    const f = (a) => {
      if (a.output) {
        const p = A(a.output);
        if (!p || p.length < 1)
          throw new TypeError("Redeem.output too short");
        if (a.output.byteLength > 520)
          throw new TypeError("Redeem.output unspendable if larger than 520 bytes");
        if (Fe(p) > 201)
          throw new TypeError("Redeem.output unspendable with more than 201 non-push ops");
        const l = C(a.output);
        if (u.length > 0 && !u.equals(l))
          throw new TypeError("Hash mismatch");
        u = l;
      }
      if (a.input) {
        const p = a.input.length > 0, l = a.witness && a.witness.length > 0;
        if (!p && !l)
          throw new TypeError("Empty input");
        if (p && l)
          throw new TypeError("Input and witness provided");
        if (p) {
          const d = A(a.input);
          if (!ue(d))
            throw new TypeError("Non push-only scriptSig");
        }
      }
    };
    if (t.input) {
      const a = i();
      if (!a || a.length < 1)
        throw new TypeError("Input too short");
      if (!h.isBuffer(o().output))
        throw new TypeError("Input is invalid");
      f(o());
    }
    if (t.redeem) {
      if (t.redeem.network && t.redeem.network !== r)
        throw new TypeError("Network mismatch");
      if (t.input) {
        const a = o();
        if (t.redeem.output && !t.redeem.output.equals(a.output))
          throw new TypeError("Redeem.output mismatch");
        if (t.redeem.input && !t.redeem.input.equals(a.input))
          throw new TypeError("Redeem.input mismatch");
      }
      f(t.redeem);
    }
    if (t.witness && t.redeem && t.redeem.witness && !at(t.redeem.witness, t.witness))
      throw new TypeError("Witness and redeem.witness mismatch");
  }
  return Object.assign(n, t);
}
const vt = {};
function oo(t) {
  t ? t !== vt.eccLib && ($s(t), vt.eccLib = t) : vt.eccLib = t;
}
function He() {
  if (!vt.eccLib)
    throw new Error("No ECC Library provided. You must call initEccLib() with a valid TinySecp256k1Interface instance");
  return vt.eccLib;
}
const tt = (t) => h.from(t, "hex");
function $s(t) {
  L(typeof t.isXOnlyPoint == "function"), L(t.isXOnlyPoint(tt("79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798"))), L(t.isXOnlyPoint(tt("fffffffffffffffffffffffffffffffffffffffffffffffffffffffeeffffc2e"))), L(t.isXOnlyPoint(tt("f9308a019258c31049344f85f89d5229b531c845836f99b08601f113bce036f9"))), L(t.isXOnlyPoint(tt("0000000000000000000000000000000000000000000000000000000000000001"))), L(!t.isXOnlyPoint(tt("0000000000000000000000000000000000000000000000000000000000000000"))), L(!t.isXOnlyPoint(tt("fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f"))), L(typeof t.xOnlyPointAddTweak == "function"), qs.forEach((e) => {
    const r = t.xOnlyPointAddTweak(tt(e.pubkey), tt(e.tweak));
    e.result === null ? L(r === null) : (L(r !== null), L(r.parity === e.parity), L(h.from(r.xOnlyPubkey).equals(tt(e.result))));
  });
}
function L(t) {
  if (!t)
    throw new Error("ecc library invalid");
}
const qs = [
  {
    pubkey: "79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798",
    tweak: "fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364140",
    parity: -1,
    result: null
  },
  {
    pubkey: "1617d38ed8d8657da4d4761e8057bc396ea9e4b9d29776d4be096016dbd2509b",
    tweak: "a8397a935f0dfceba6ba9618f6451ef4d80637abf4e6af2669fbc9de6a8fd2ac",
    parity: 1,
    result: "e478f99dab91052ab39a33ea35fd5e6e4933f4d28023cd597c9a1f6760346adf"
  },
  {
    pubkey: "2c0b7cf95324a07d05398b240174dc0c2be444d96b159aa6c7f7b1e668680991",
    tweak: "823c3cd2142744b075a87eade7e1b8678ba308d566226a0056ca2b7a76f86b47",
    parity: 0,
    result: "9534f8dc8c6deda2dc007655981c78b49c5d96c778fbf363462a11ec9dfd948c"
  }
], { typeforce: Sn } = ie;
function _n(t, e) {
  if (typeof t != "number")
    throw new Error("cannot write a non-number as a number");
  if (t < 0)
    throw new Error("specified a negative value for writing an unsigned value");
  if (t > e)
    throw new Error("RangeError: value out of range");
  if (Math.floor(t) !== t)
    throw new Error("value has a fractional component");
}
function Xs(t, e) {
  const r = t.readUInt32LE(e);
  let n = t.readUInt32LE(e + 4);
  return n *= 4294967296, _n(n + r, 9007199254740991), n + r;
}
function js(t, e, r) {
  return _n(e, 9007199254740991), t.writeInt32LE(e & -1, r), t.writeUInt32LE(Math.floor(e / 4294967296), r + 4), r + 8;
}
function We(t) {
  if (t.length < 1)
    return t;
  let e = t.length - 1, r = 0;
  for (let n = 0; n < t.length / 2; n++)
    r = t[n], t[n] = t[e], t[e] = r, e--;
  return t;
}
function lr(t) {
  const e = h.allocUnsafe(t.length);
  return t.copy(e), e;
}
class x {
  constructor(e, r = 0) {
    this.buffer = e, this.offset = r, Sn(z(V, Y), [e, r]);
  }
  static withCapacity(e) {
    return new x(h.alloc(e));
  }
  writeUInt8(e) {
    this.offset = this.buffer.writeUInt8(e, this.offset);
  }
  writeInt32(e) {
    this.offset = this.buffer.writeInt32LE(e, this.offset);
  }
  writeUInt32(e) {
    this.offset = this.buffer.writeUInt32LE(e, this.offset);
  }
  writeUInt64(e) {
    this.offset = js(this.buffer, e, this.offset);
  }
  writeVarInt(e) {
    const r = Br(e, this.buffer, this.offset);
    this.offset += r.bytes;
  }
  writeSlice(e) {
    if (this.buffer.length < this.offset + e.length)
      throw new Error("Cannot write slice out of bounds");
    this.offset += e.copy(this.buffer, this.offset);
  }
  writeVarSlice(e) {
    this.writeVarInt(e.length), this.writeSlice(e);
  }
  writeVector(e) {
    this.writeVarInt(e.length), e.forEach((r) => this.writeVarSlice(r));
  }
  end() {
    if (this.buffer.length === this.offset)
      return this.buffer;
    throw new Error(`buffer size ${this.buffer.length}, offset ${this.offset}`);
  }
}
class zs {
  constructor(e, r = 0) {
    this.buffer = e, this.offset = r, Sn(z(V, Y), [e, r]);
  }
  readUInt8() {
    const e = this.buffer.readUInt8(this.offset);
    return this.offset++, e;
  }
  readInt32() {
    const e = this.buffer.readInt32LE(this.offset);
    return this.offset += 4, e;
  }
  readUInt32() {
    const e = this.buffer.readUInt32LE(this.offset);
    return this.offset += 4, e;
  }
  readUInt64() {
    const e = Xs(this.buffer, this.offset);
    return this.offset += 8, e;
  }
  readVarInt() {
    const e = Kn(this.buffer, this.offset);
    return this.offset += e.bytes, e.numberValue || 0;
  }
  readSlice(e) {
    if (this.buffer.length < this.offset + e)
      throw new Error("Cannot read slice out of bounds");
    const r = h.from(this.buffer.subarray(this.offset, this.offset + e));
    return this.offset += e, r;
  }
  readVarSlice() {
    return this.readSlice(this.readVarInt());
  }
  readVector() {
    const e = this.readVarInt(), r = [];
    for (let n = 0; n < e; n++)
      r.push(this.readVarSlice());
    return r;
  }
}
const Ge = 192, Tn = 128, Ys = (t) => "left" in t && "right" in t;
function te(t, e) {
  if (t.length < 33)
    throw new TypeError(`The control-block length is too small. Got ${t.length}, expected min 33.`);
  const r = (t.length - 33) / 32;
  let n = e;
  for (let s = 0; s < r; s++) {
    const i = t.subarray(33 + 32 * s, 65 + 32 * s);
    n.compare(i) < 0 ? n = ve(n, i) : n = ve(i, n);
  }
  return n;
}
function ee(t) {
  if (ne(t))
    return { hash: Z(t) };
  const e = [ee(t[0]), ee(t[1])];
  e.sort((s, i) => s.hash.compare(i.hash));
  const [r, n] = e;
  return {
    hash: ve(r.hash, n.hash),
    left: r,
    right: n
  };
}
function Bt(t, e) {
  if (Ys(t)) {
    const r = Bt(t.left, e);
    if (r !== void 0)
      return [...r, t.right.hash];
    const n = Bt(t.right, e);
    if (n !== void 0)
      return [...n, t.left.hash];
  } else if (t.hash.equals(e))
    return [];
}
function Z(t) {
  const e = t.version || Ge;
  return Rt("TapLeaf", U.concat([U.from([e]), Qs(t.output)]));
}
function Zs(t, e) {
  return Rt("TapTweak", U.concat(e ? [t, e] : [t]));
}
function At(t, e) {
  if (!U.isBuffer(t) || t.length !== 32 || e && e.length !== 32)
    return null;
  const r = Zs(t, e), n = He().xOnlyPointAddTweak(t, r);
  return !n || n.xOnlyPubkey === null ? null : {
    parity: n.parity,
    x: U.from(n.xOnlyPubkey)
  };
}
function ve(t, e) {
  return Rt("TapBranch", U.concat([t, e]));
}
function Qs(t) {
  const e = kt(t.length), r = U.allocUnsafe(e);
  return Br(t.length, r), U.concat([r, t]);
}
const dr = E, wr = 1, Js = 80;
function bt(t, e) {
  if (!t.address && !t.output && !t.pubkey && !t.internalPubkey && !(t.witness && t.witness.length > 1))
    throw new TypeError("Not enough data");
  e = Object.assign({ validate: !0 }, e || {}), c({
    address: c.maybe(c.String),
    input: c.maybe(c.BufferN(0)),
    network: c.maybe(c.Object),
    output: c.maybe(c.BufferN(34)),
    internalPubkey: c.maybe(c.BufferN(32)),
    hash: c.maybe(c.BufferN(32)),
    pubkey: c.maybe(c.BufferN(32)),
    signature: c.maybe(c.anyOf(c.BufferN(64), c.BufferN(65))),
    witness: c.maybe(c.arrayOf(c.Buffer)),
    scriptTree: c.maybe(Re),
    redeem: c.maybe({
      output: c.maybe(c.Buffer),
      redeemVersion: c.maybe(c.Number),
      witness: c.maybe(c.arrayOf(c.Buffer))
    }),
    redeemVersion: c.maybe(c.Number)
  }, t);
  const r = B(() => Be(t.address)), n = B(() => {
    if (!(!t.witness || !t.witness.length))
      return t.witness.length >= 2 && t.witness[t.witness.length - 1][0] === Js ? t.witness.slice(0, -1) : t.witness.slice();
  }), s = B(() => {
    if (t.scriptTree)
      return ee(t.scriptTree);
    if (t.hash)
      return { hash: t.hash };
  }), i = t.network || M, o = {
    name: K.P2TR,
    network: i
  };
  if (y(o, "address", () => {
    if (!o.pubkey)
      return;
    const u = I.bech32m.toWords(o.pubkey);
    return u.unshift(wr), I.bech32m.encode(i.bech32, u);
  }), y(o, "hash", () => {
    const u = s();
    if (u)
      return u.hash;
    const f = n();
    if (f && f.length > 1) {
      const a = f[f.length - 1], p = a[0] & Pt, l = f[f.length - 2], d = Z({
        output: l,
        version: p
      });
      return te(a, d);
    }
  }), y(o, "output", () => {
    if (o.pubkey)
      return v([dr.OP_1, o.pubkey]);
  }), y(o, "redeemVersion", () => t.redeemVersion ? t.redeemVersion : t.redeem && t.redeem.redeemVersion !== void 0 && t.redeem.redeemVersion !== null ? t.redeem.redeemVersion : Ge), y(o, "redeem", () => {
    const u = n();
    if (!(!u || u.length < 2))
      return {
        output: u[u.length - 2],
        witness: u.slice(0, -2),
        redeemVersion: u[u.length - 1][0] & Pt
      };
  }), y(o, "pubkey", () => {
    if (t.pubkey)
      return t.pubkey;
    if (t.output)
      return t.output.subarray(2);
    if (t.address)
      return r().data;
    if (o.internalPubkey) {
      const u = At(o.internalPubkey, o.hash);
      if (u)
        return u.x;
    }
  }), y(o, "internalPubkey", () => {
    if (t.internalPubkey)
      return t.internalPubkey;
    const u = n();
    if (u && u.length > 1)
      return u[u.length - 1].subarray(1, 33);
  }), y(o, "signature", () => {
    if (t.signature)
      return t.signature;
    const u = n();
    if (!(!u || u.length !== 1))
      return u[0];
  }), y(o, "witness", () => {
    if (t.witness)
      return t.witness;
    const u = s();
    if (u && t.redeem && t.redeem.output && t.internalPubkey) {
      const f = Z({
        output: t.redeem.output,
        version: o.redeemVersion
      }), a = Bt(u, f);
      if (!a)
        return;
      const p = At(t.internalPubkey, u.hash);
      if (!p)
        return;
      const l = U.concat([U.from([o.redeemVersion | p.parity]), t.internalPubkey].concat(a));
      return [t.redeem.output, l];
    }
    if (t.signature)
      return [t.signature];
  }), e.validate) {
    let u = U.from([]);
    if (t.address) {
      if (i && i.bech32 !== r().prefix)
        throw new TypeError("Invalid prefix or Network mismatch");
      if (r().version !== wr)
        throw new TypeError("Invalid address version");
      if (r().data.length !== 32)
        throw new TypeError("Invalid address data");
      u = r().data;
    }
    if (t.pubkey) {
      if (u.length > 0 && !u.equals(t.pubkey))
        throw new TypeError("Pubkey mismatch");
      u = t.pubkey;
    }
    if (t.output) {
      if (t.output.length !== 34 || t.output[0] !== dr.OP_1 || t.output[1] !== 32)
        throw new TypeError("Output is invalid");
      if (u.length > 0 && !u.equals(t.output.subarray(2)))
        throw new TypeError("Pubkey mismatch");
      u = t.output.subarray(2);
    }
    if (t.internalPubkey) {
      const p = At(t.internalPubkey, o.hash);
      if (u.length > 0 && !u.equals(p.x))
        throw new TypeError("Pubkey mismatch");
      u = p.x;
    }
    if (u && u.length && !He().isXOnlyPoint(u))
      throw new TypeError("Invalid pubkey for p2tr");
    const f = s();
    if (t.hash && f && !t.hash.equals(f.hash))
      throw new TypeError("Hash mismatch");
    if (t.redeem && t.redeem.output && f) {
      const p = Z({
        output: t.redeem.output,
        version: o.redeemVersion
      });
      if (!Bt(f, p))
        throw new TypeError("Redeem script not in tree");
    }
    const a = n();
    if (t.redeem && o.redeem) {
      if (t.redeem.redeemVersion && t.redeem.redeemVersion !== o.redeem.redeemVersion)
        throw new TypeError("Redeem.redeemVersion and witness mismatch");
      if (t.redeem.output) {
        if (A(t.redeem.output).length === 0)
          throw new TypeError("Redeem.output is invalid");
        if (o.redeem.output && !t.redeem.output.equals(o.redeem.output))
          throw new TypeError("Redeem.output and witness mismatch");
      }
      if (t.redeem.witness && o.redeem.witness && !at(t.redeem.witness, o.redeem.witness))
        throw new TypeError("Redeem.witness and witness mismatch");
    }
    if (a && a.length)
      if (a.length === 1) {
        if (t.signature && !t.signature.equals(a[0]))
          throw new TypeError("Signature mismatch");
      } else {
        const p = a[a.length - 1];
        if (p.length < 33)
          throw new TypeError(`The control-block length is too small. Got ${p.length}, expected min 33.`);
        if ((p.length - 33) % 32 !== 0)
          throw new TypeError(`The control-block length of ${p.length} is incorrect!`);
        const l = (p.length - 33) / 32;
        if (l > 128)
          throw new TypeError(`The script path is too long. Got ${l}, expected max 128.`);
        const d = p.subarray(1, 33);
        if (t.internalPubkey && !t.internalPubkey.equals(d))
          throw new TypeError("Internal pubkey mismatch");
        if (!He().isXOnlyPoint(d))
          throw new TypeError("Invalid internalPubkey for p2tr witness");
        const b = p[0] & Pt, w = a[a.length - 2], S = Z({
          output: w,
          version: b
        }), P = te(p, S), k = At(d, P);
        if (!k)
          throw new TypeError("Invalid outputKey for p2tr witness");
        if (u.length && !u.equals(k.x))
          throw new TypeError("Pubkey mismatch for p2tr witness");
        if (k.parity !== (p[0] & 1))
          throw new Error("Incorrect parity");
      }
  }
  return Object.assign(o, t);
}
const gr = E, ti = h.alloc(0);
function Lt(t, e) {
  if (!t.address && !t.hash && !t.output && !t.pubkey && !t.witness)
    throw new TypeError("Not enough data");
  e = Object.assign({ validate: !0 }, e || {}), c({
    address: c.maybe(c.String),
    hash: c.maybe(c.BufferN(20)),
    input: c.maybe(c.BufferN(0)),
    network: c.maybe(c.Object),
    output: c.maybe(c.BufferN(22)),
    pubkey: c.maybe(W),
    signature: c.maybe(Q),
    witness: c.maybe(c.arrayOf(c.Buffer))
  }, t);
  const r = B(() => {
    const i = I.bech32.decode(t.address), o = i.words.shift(), u = I.bech32.fromWords(i.words);
    return {
      version: o,
      prefix: i.prefix,
      data: h.from(u)
    };
  }), n = t.network || M, s = {
    name: K.P2WPKH,
    network: n
  };
  if (y(s, "address", () => {
    if (!s.hash)
      return;
    const i = I.bech32.toWords(s.hash);
    return i.unshift(0), I.bech32.encode(n.bech32, i);
  }), y(s, "hash", () => {
    if (t.output)
      return t.output.subarray(2, 22);
    if (t.address)
      return r().data;
    if (t.pubkey || s.pubkey)
      return C(t.pubkey || s.pubkey);
  }), y(s, "output", () => {
    if (s.hash)
      return v([gr.OP_0, s.hash]);
  }), y(s, "pubkey", () => {
    if (t.pubkey)
      return t.pubkey;
    if (t.witness)
      return t.witness[1];
  }), y(s, "signature", () => {
    if (t.witness)
      return t.witness[0];
  }), y(s, "input", () => {
    if (s.witness)
      return ti;
  }), y(s, "witness", () => {
    if (t.pubkey && t.signature)
      return [t.signature, t.pubkey];
  }), e.validate) {
    let i = h.from([]);
    if (t.address) {
      if (n && n.bech32 !== r().prefix)
        throw new TypeError("Invalid prefix or Network mismatch");
      if (r().version !== 0)
        throw new TypeError("Invalid address version");
      if (r().data.length !== 20)
        throw new TypeError("Invalid address data");
      i = r().data;
    }
    if (t.hash) {
      if (i.length > 0 && !i.equals(t.hash))
        throw new TypeError("Hash mismatch");
      i = t.hash;
    }
    if (t.output) {
      if (t.output.length !== 22 || t.output[0] !== gr.OP_0 || t.output[1] !== 20)
        throw new TypeError("Output is invalid");
      if (i.length > 0 && !i.equals(t.output.subarray(2)))
        throw new TypeError("Hash mismatch");
      i = t.output.subarray(2);
    }
    if (t.pubkey) {
      const o = C(t.pubkey);
      if (i.length > 0 && !i.equals(o))
        throw new TypeError("Hash mismatch");
      if (i = o, !W(t.pubkey) || t.pubkey.length !== 33)
        throw new TypeError("Invalid pubkey for p2wpkh");
    }
    if (t.witness) {
      if (t.witness.length !== 2)
        throw new TypeError("Witness is invalid");
      if (!Q(t.witness[0]))
        throw new TypeError("Witness has invalid signature");
      if (!W(t.witness[1]) || t.witness[1].length !== 33)
        throw new TypeError("Witness has invalid pubkey");
      if (t.signature && !t.signature.equals(t.witness[0]))
        throw new TypeError("Signature mismatch");
      if (t.pubkey && !t.pubkey.equals(t.witness[1]))
        throw new TypeError("Pubkey mismatch");
      const o = C(t.witness[1]);
      if (i.length > 0 && !i.equals(o))
        throw new TypeError("Hash mismatch");
    }
  }
  return Object.assign(s, t);
}
const yr = E, ge = h.alloc(0);
function Kt(t) {
  return !!(h.isBuffer(t) && t.length === 65 && t[0] === 4 && W(t));
}
function Et(t, e) {
  if (!t.address && !t.hash && !t.output && !t.redeem && !t.witness)
    throw new TypeError("Not enough data");
  e = Object.assign({ validate: !0 }, e || {}), c({
    network: c.maybe(c.Object),
    address: c.maybe(c.String),
    hash: c.maybe(c.BufferN(32)),
    output: c.maybe(c.BufferN(34)),
    redeem: c.maybe({
      input: c.maybe(c.Buffer),
      network: c.maybe(c.Object),
      output: c.maybe(c.Buffer),
      witness: c.maybe(c.arrayOf(c.Buffer))
    }),
    input: c.maybe(c.BufferN(0)),
    witness: c.maybe(c.arrayOf(c.Buffer))
  }, t);
  const r = B(() => {
    const o = I.bech32.decode(t.address), u = o.words.shift(), f = I.bech32.fromWords(o.words);
    return {
      version: u,
      prefix: o.prefix,
      data: h.from(f)
    };
  }), n = B(() => A(t.redeem.input));
  let s = t.network;
  s || (s = t.redeem && t.redeem.network || M);
  const i = {
    network: s,
    name: K.P2WSH
  };
  if (y(i, "address", () => {
    if (!i.hash)
      return;
    const o = I.bech32.toWords(i.hash);
    return o.unshift(0), I.bech32.encode(s.bech32, o);
  }), y(i, "hash", () => {
    if (t.output)
      return t.output.subarray(2);
    if (t.address)
      return r().data;
    if (i.redeem && i.redeem.output)
      return D(i.redeem.output);
  }), y(i, "output", () => {
    if (i.hash)
      return v([yr.OP_0, i.hash]);
  }), y(i, "redeem", () => {
    if (t.witness)
      return {
        output: t.witness[t.witness.length - 1],
        input: ge,
        witness: t.witness.slice(0, -1)
      };
  }), y(i, "input", () => {
    if (i.witness)
      return ge;
  }), y(i, "witness", () => {
    if (t.redeem && t.redeem.input && t.redeem.input.length > 0 && t.redeem.output && t.redeem.output.length > 0) {
      const o = jr(n());
      return i.redeem = Object.assign({ witness: o }, t.redeem), i.redeem.input = ge, [].concat(o, t.redeem.output);
    }
    if (t.redeem && t.redeem.output && t.redeem.witness)
      return [].concat(t.redeem.witness, t.redeem.output);
  }), y(i, "name", () => {
    const o = ["p2wsh"];
    return i.redeem !== void 0 && i.redeem.name !== void 0 && o.push(i.redeem.name), o.join("-");
  }), e.validate) {
    let o = h.from([]);
    if (t.address) {
      if (r().prefix !== s.bech32)
        throw new TypeError("Invalid prefix or Network mismatch");
      if (r().version !== 0)
        throw new TypeError("Invalid address version");
      if (r().data.length !== 32)
        throw new TypeError("Invalid address data");
      o = r().data;
    }
    if (t.hash) {
      if (o.length > 0 && !o.equals(t.hash))
        throw new TypeError("Hash mismatch");
      o = t.hash;
    }
    if (t.output) {
      if (t.output.length !== 34 || t.output[0] !== yr.OP_0 || t.output[1] !== 32)
        throw new TypeError("Output is invalid");
      const u = t.output.subarray(2);
      if (o.length > 0 && !o.equals(u))
        throw new TypeError("Hash mismatch");
      o = u;
    }
    if (t.redeem) {
      if (t.redeem.network && t.redeem.network !== s)
        throw new TypeError("Network mismatch");
      if (t.redeem.input && t.redeem.input.length > 0 && t.redeem.witness && t.redeem.witness.length > 0)
        throw new TypeError("Ambiguous witness source");
      if (t.redeem.output) {
        const u = A(t.redeem.output);
        if (!u || u.length < 1)
          throw new TypeError("Redeem.output is invalid");
        if (t.redeem.output.byteLength > 3600)
          throw new TypeError("Redeem.output unspendable if larger than 3600 bytes");
        if (Fe(u) > 201)
          throw new TypeError("Redeem.output unspendable with more than 201 non-push ops");
        const f = D(t.redeem.output);
        if (o.length > 0 && !o.equals(f))
          throw new TypeError("Hash mismatch");
        o = f;
      }
      if (t.redeem.input && !ue(n()))
        throw new TypeError("Non push-only scriptSig");
      if (t.witness && t.redeem.witness && !at(t.witness, t.redeem.witness))
        throw new TypeError("Witness and redeem.witness mismatch");
      if (t.redeem.input && n().some(Kt) || t.redeem.output && (A(t.redeem.output) || []).some(Kt))
        throw new TypeError("redeem.input or redeem.output contains uncompressed pubkey");
    }
    if (t.witness && t.witness.length > 0) {
      const u = t.witness[t.witness.length - 1];
      if (t.redeem && t.redeem.output && !t.redeem.output.equals(u))
        throw new TypeError("Witness and redeem.output mismatch");
      if (t.witness.some(Kt) || (A(u) || []).some(Kt))
        throw new TypeError("Witness contains uncompressed pubkey");
    }
  }
  return Object.assign(i, t);
}
const fe = 40, ce = 2, ae = 15, ei = 16, xe = 16, Ke = 2, Me = 80, ri = "WARNING: Sending to a future segwit version address can lead to loss of funds. End users MUST be warned carefully in the GUI and asked if they wish to proceed with caution. Wallets should verify the segwit version from the output of fromBech32, then decide when it is safe to use which version of segwit.", ni = (t) => {
  try {
    const e = h.from(t.subarray(2));
    if (e.length < ce || e.length > fe)
      throw new TypeError("Invalid program length for segwit address");
    const r = t[0] - Me;
    if (r < Ke || r > ae + 1)
      throw new TypeError("Invalid version for segwit address");
    if (r === 1)
      throw new TypeError("taproot");
    return !0;
  } catch {
  }
  return !1;
};
function si(t, e) {
  if (!h.isBuffer(t))
    throw new TypeError("output must be a Buffer");
  if (!e.bech32Opnet)
    throw new Error("Network does not support opnet");
  const r = t[0];
  let n = 1, s;
  if (t[1] < 76)
    s = t[1], n = 2;
  else if (t[1] === 76)
    s = t[2], n = 3;
  else
    throw new TypeError("Unsupported push opcode in script");
  const i = h.from(t.subarray(n, n + s));
  if (i.length < ce || i.length > fe)
    throw new TypeError("Invalid program length for segwit address");
  const o = r === E.OP_0 ? 0 : r >= E.OP_1 && r <= E.OP_16 ? r - (E.OP_1 - 1) : -1;
  if (o < ae || o > ei)
    throw new TypeError(`Invalid segwit version ${o}`);
  const u = [o, ...I.bech32m.toWords(i)];
  return I.bech32m.encode(e.bech32Opnet, u);
}
function ii(t, e) {
  const r = h.from(t.subarray(2));
  if (r.length < ce || r.length > fe)
    throw new TypeError("Invalid program length for segwit address");
  const n = t[0] - Me;
  if (n < Ke || n > ae)
    throw new TypeError("Invalid version for segwit address");
  if (t[1] !== r.length)
    throw new TypeError(`Invalid script for segwit address ${t[1]} !== ${r.length}`);
  return ui(r, n, e.bech32, e.bech32Opnet);
}
function oi(t) {
  const e = h.from(Ut.decode(t));
  if (e.length < 21)
    throw new TypeError(t + " is too short");
  if (e.length > 21)
    throw new TypeError(t + " is too long");
  const r = e.readUInt8(0), n = h.from(e.subarray(1));
  return { version: r, hash: n };
}
function ui(t, e, r, n) {
  const s = I.bech32.toWords(t);
  return s.unshift(e), e === xe && n ? I.bech32m.encode(n, s) : e === 0 ? I.bech32.encode(r, s) : I.bech32m.encode(r, s);
}
function fi(t, e) {
  e = e || M;
  try {
    return yt({ output: t, network: e }).address;
  } catch {
  }
  try {
    return mt({ output: t, network: e }).address;
  } catch {
  }
  try {
    return Lt({ output: t, network: e }).address;
  } catch {
  }
  try {
    return Et({ output: t, network: e }).address;
  } catch {
  }
  try {
    return bt({ output: t, network: e }).address;
  } catch {
  }
  try {
    return si(t, e);
  } catch {
  }
  try {
    return ii(t, e);
  } catch {
  }
  throw new Error(Xr(t) + " has no matching Address");
}
function ci(t, e) {
  e = e || M;
  let r, n;
  try {
    r = oi(t);
  } catch {
  }
  if (r) {
    if (r.version === e.pubKeyHash)
      return yt({ hash: r.hash }).output;
    if (r.version === e.scriptHash)
      return mt({ hash: r.hash }).output;
  } else {
    try {
      n = Be(t);
    } catch {
    }
    if (n) {
      if (n.prefix !== e.bech32 && e.bech32Opnet && n.prefix !== e.bech32Opnet)
        throw new Error(t + " has an invalid prefix");
      if (n.version === 0) {
        if (n.data.length === 20)
          return Lt({ hash: n.data }).output;
        if (n.data.length === 32)
          return Et({ hash: n.data }).output;
      } else if (n.version === 1) {
        if (n.data.length === 32)
          return bt({ pubkey: n.data }).output;
      } else if (n.version === xe) {
        if (!e.bech32Opnet)
          throw new Error(t + " has an invalid prefix");
        return Yr({
          program: n.data,
          network: e
        }).output;
      } else if (n.version >= Ke && n.version <= ae && n.data.length >= ce && n.data.length <= fe)
        return n.version !== xe && console.warn(ri), v([
          n.version + Me,
          n.data
        ]);
    }
  }
  throw new TypeError(t + " has no matching Script");
}
const mr = E;
function ai(t, e) {
  if (!t.data && !t.output)
    throw new TypeError("Not enough data");
  e = Object.assign({ validate: !0 }, e || {}), c({
    network: c.maybe(c.Object),
    output: c.maybe(c.Buffer),
    data: c.maybe(c.arrayOf(c.Buffer))
  }, t);
  const r = t.network || M, n = { name: K.Embed, network: r, data: [] };
  if (y(n, "output", () => {
    if (t.data)
      return v([mr.OP_RETURN].concat(t.data));
  }), y(n, "data", () => {
    if (!t.output)
      return;
    const s = A(t.output);
    if (s != null)
      return s.slice(1);
  }), e.validate && t.output) {
    const s = A(t.output);
    if (s[0] !== mr.OP_RETURN)
      throw new TypeError("Output is invalid");
    if (!s.slice(1).every(c.Buffer))
      throw new TypeError("Output is invalid");
    if (t.data && !at(t.data, n.data))
      throw new TypeError("Data mismatch");
  }
  return Object.assign(n, t);
}
const lt = E, Mt = lt.OP_RESERVED;
function Vt(t, e) {
  if (!t.input && !t.output && !(t.pubkeys && t.m !== void 0) && !t.signatures)
    throw new TypeError("Not enough data");
  e = Object.assign({ validate: !0 }, e || {});
  function r(f) {
    return Q(f) || (e.allowIncomplete && f === lt.OP_0) !== void 0;
  }
  c({
    network: c.maybe(c.Object),
    m: c.maybe(c.Number),
    n: c.maybe(c.Number),
    output: c.maybe(c.Buffer),
    pubkeys: c.maybe(c.arrayOf(W)),
    signatures: c.maybe(c.arrayOf(r)),
    input: c.maybe(c.Buffer)
  }, t);
  const s = {
    network: t.network || M,
    name: K.P2MS
  };
  let i = [], o = !1;
  function u(f) {
    o || (o = !0, i = A(f), s.m = i[0] - Mt, s.n = i[i.length - 2] - Mt, s.pubkeys = i.slice(1, -2));
  }
  if (y(s, "output", () => {
    if (t.m && s.n && t.pubkeys)
      return v([].concat(Mt + t.m, t.pubkeys, Mt + s.n, lt.OP_CHECKMULTISIG));
  }), y(s, "m", () => {
    if (s.output)
      return u(s.output), s.m;
  }), y(s, "n", () => {
    if (s.pubkeys)
      return s.pubkeys.length;
  }), y(s, "pubkeys", () => {
    if (t.output)
      return u(t.output), s.pubkeys;
  }), y(s, "signatures", () => {
    if (!t.input)
      return;
    const f = A(t.input);
    if (f != null)
      return f.slice(1);
  }), y(s, "input", () => {
    if (t.signatures)
      return v([lt.OP_0].concat(t.signatures));
  }), y(s, "witness", () => {
    if (s.input)
      return [];
  }), y(s, "name", () => {
    if (!(!s.m || !s.n))
      return `p2ms(${s.m} of ${s.n})`;
  }), e.validate) {
    if (t.output) {
      if (u(t.output), !c.Number(i[0]))
        throw new TypeError("Output is invalid");
      if (!c.Number(i[i.length - 2]))
        throw new TypeError("Output is invalid");
      if (i[i.length - 1] !== lt.OP_CHECKMULTISIG)
        throw new TypeError("Output is invalid");
      if (s.m <= 0 || s.n > 16 || s.m > s.n || s.n !== i.length - 3)
        throw new TypeError("Output is invalid");
      if (!s.pubkeys.every((f) => W(f)))
        throw new TypeError("Output is invalid");
      if (t.m !== void 0 && t.m !== s.m)
        throw new TypeError("m mismatch");
      if (t.n !== void 0 && t.n !== s.n)
        throw new TypeError("n mismatch");
      if (t.pubkeys && !at(t.pubkeys, s.pubkeys))
        throw new TypeError("Pubkeys mismatch");
    }
    if (t.pubkeys) {
      if (t.n !== void 0 && t.n !== t.pubkeys.length)
        throw new TypeError("Pubkey count mismatch");
      if (s.n = t.pubkeys.length, s.n < s.m)
        throw new TypeError("Pubkey count cannot be less than m");
    }
    if (t.signatures) {
      if (t.signatures.length < s.m)
        throw new TypeError("Not enough signatures provided");
      if (t.signatures.length > s.m)
        throw new TypeError("Too many signatures provided");
    }
    if (t.input) {
      if (t.input[0] !== lt.OP_0)
        throw new TypeError("Input is invalid");
      if (s.signatures.length === 0 || !s.signatures.every(r))
        throw new TypeError("Input has invalid signature(s)");
      if (t.signatures && !at(t.signatures, s.signatures))
        throw new TypeError("Signature mismatch");
      if (t.m !== void 0 && t.m !== t.signatures.length)
        throw new TypeError("Signature count mismatch");
    }
  }
  return Object.assign(s, t);
}
const br = E;
function De(t, e) {
  if (!t.input && !t.output && !t.pubkey && !t.input && !t.signature)
    throw new TypeError("Not enough data");
  e = Object.assign({ validate: !0 }, e || {}), c({
    network: c.maybe(c.Object),
    output: c.maybe(c.Buffer),
    pubkey: c.maybe(W),
    signature: c.maybe(Q),
    input: c.maybe(c.Buffer)
  }, t);
  const r = B(() => A(t.input)), n = t.network || M, s = {
    name: K.P2PK,
    network: n,
    pubkey: void 0
  };
  if (y(s, "output", () => {
    if (t.pubkey)
      return v([t.pubkey, br.OP_CHECKSIG]);
  }), y(s, "pubkey", () => {
    if (t.output)
      return t.output.subarray(1, -1);
  }), y(s, "signature", () => {
    if (t.input)
      return r()[0];
  }), y(s, "input", () => {
    if (t.signature)
      return v([t.signature]);
  }), y(s, "witness", () => {
    if (s.input)
      return [];
  }), e.validate) {
    if (t.output) {
      if (t.output[t.output.length - 1] !== br.OP_CHECKSIG)
        throw new TypeError("Output is invalid");
      if (!W(s.pubkey))
        throw new TypeError("Output pubkey is invalid");
      if (t.pubkey && !t.pubkey.equals(s.pubkey))
        throw new TypeError("Pubkey mismatch");
    }
    if (t.signature && t.input && !t.input.equals(s.input))
      throw new TypeError("Signature mismatch");
    if (t.input) {
      if (r().length !== 1)
        throw new TypeError("Input is invalid");
      if (!Q(s.signature))
        throw new TypeError("Input has invalid signature");
    }
  }
  return Object.assign(s, t);
}
const hi = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  LEAF_VERSION_TAPSCRIPT: Ge,
  MAX_TAPTREE_DEPTH: Tn,
  get PaymentType() {
    return K;
  },
  findScriptPath: Bt,
  p2data: ai,
  p2ms: Vt,
  p2op: Yr,
  p2pk: De,
  p2pkh: yt,
  p2sh: mt,
  p2tr: bt,
  p2wpkh: Lt,
  p2wsh: Et,
  prop: y,
  rootHashFromPath: te,
  tapleafHash: Z,
  toHashTree: ee,
  tweakKey: At,
  value: B
}, Symbol.toStringTag, { value: "Module" })), { typeforce: X } = ie;
function j(t) {
  const e = t.length;
  return kt(e) + e;
}
function pi(t) {
  const e = t.length;
  return kt(e) + t.reduce((r, n) => r + j(n), 0);
}
const ot = h.allocUnsafe(0), Er = [], ye = h.from("0000000000000000000000000000000000000000000000000000000000000000", "hex"), Sr = h.from("0000000000000000000000000000000000000000000000000000000000000001", "hex"), li = h.from("ffffffffffffffff", "hex"), di = {
  script: ot,
  valueBuffer: li
};
function wi(t) {
  return "value" in t;
}
class g {
  constructor() {
    this.version = 1, this.locktime = 0, this.ins = [], this.outs = [];
  }
  static fromBuffer(e, r) {
    const n = new zs(e), s = new g();
    s.version = n.readInt32();
    const i = n.readUInt8(), o = n.readUInt8();
    let u = !1;
    i === g.ADVANCED_TRANSACTION_MARKER && o === g.ADVANCED_TRANSACTION_FLAG ? u = !0 : n.offset -= 2;
    const f = n.readVarInt();
    for (let p = 0; p < f; ++p) {
      const l = n.readSlice(32), d = n.readUInt32(), b = n.readVarSlice(), w = n.readUInt32();
      s.ins.push({
        hash: l,
        index: d,
        script: b,
        sequence: w,
        witness: Er
      });
    }
    const a = n.readVarInt();
    for (let p = 0; p < a; ++p)
      s.outs.push({
        value: n.readUInt64(),
        script: n.readVarSlice()
      });
    if (u) {
      for (let p = 0; p < f; ++p)
        s.ins[p].witness = n.readVector();
      if (!s.hasWitnesses())
        throw new Error("Transaction has superfluous witness data");
    }
    if (s.locktime = n.readUInt32(), r)
      return s;
    if (n.offset !== e.length)
      throw new Error("Transaction has unexpected data");
    return s;
  }
  static fromHex(e) {
    return g.fromBuffer(h.from(e, "hex"), !1);
  }
  static isCoinbaseHash(e) {
    X(_e, e);
    for (let r = 0; r < 32; ++r)
      if (e[r] !== 0)
        return !1;
    return !0;
  }
  isCoinbase() {
    return this.ins.length === 1 && g.isCoinbaseHash(this.ins[0].hash);
  }
  addInput(e, r, n, s) {
    return X(z(_e, Y, Te(Y), Te(V)), [e, r, n, s]), Mr(n) && (n = g.DEFAULT_SEQUENCE), this.ins.push({
      hash: e,
      index: r,
      script: s || ot,
      sequence: n,
      witness: Er
    }) - 1;
  }
  addOutput(e, r) {
    return X(z(V, qt), [e, r]), this.outs.push({
      script: e,
      value: r
    }) - 1;
  }
  hasWitnesses() {
    return this.ins.some((e) => e.witness.length !== 0);
  }
  weight() {
    const e = this.byteLength(!1), r = this.byteLength(!0);
    return e * 3 + r;
  }
  virtualSize() {
    return Math.ceil(this.weight() / 4);
  }
  byteLength(e = !0) {
    const r = e && this.hasWitnesses();
    return (r ? 10 : 8) + kt(this.ins.length) + kt(this.outs.length) + this.ins.reduce((n, s) => n + 40 + j(s.script), 0) + this.outs.reduce((n, s) => n + 8 + j(s.script), 0) + (r ? this.ins.reduce((n, s) => n + pi(s.witness), 0) : 0);
  }
  clone() {
    const e = new g();
    return e.version = this.version, e.locktime = this.locktime, e.ins = this.ins.map((r) => ({
      hash: r.hash,
      index: r.index,
      script: r.script,
      sequence: r.sequence,
      witness: r.witness
    })), e.outs = this.outs.map((r) => ({
      script: r.script,
      value: r.value
    })), e;
  }
  hashForSignature(e, r, n) {
    if (X(z(Y, V, It), [
      e,
      r,
      n
    ]), e >= this.ins.length)
      return Sr;
    const s = A(r);
    if (!s)
      throw new Error("Could not decompile prevOutScript");
    const i = v(s.filter((f) => f !== E.OP_CODESEPARATOR)), o = this.clone();
    if ((n & 31) === g.SIGHASH_NONE)
      o.outs = [], o.ins.forEach((f, a) => {
        a !== e && (f.sequence = 0);
      });
    else if ((n & 31) === g.SIGHASH_SINGLE) {
      if (e >= this.outs.length)
        return Sr;
      o.outs.length = e + 1;
      for (let f = 0; f < e; f++)
        o.outs[f] = di;
      o.ins.forEach((f, a) => {
        a !== e && (f.sequence = 0);
      });
    }
    n & g.SIGHASH_ANYONECANPAY ? (o.ins = [o.ins[e]], o.ins[0].script = i) : (o.ins.forEach((f) => {
      f.script = ot;
    }), o.ins[e].script = i);
    const u = h.allocUnsafe(o.byteLength(!1) + 4);
    return u.writeInt32LE(n, u.length - 4), o.__toBuffer(u, 0, !1), it(u);
  }
  hashForWitnessV1(e, r, n, s, i, o) {
    if (X(z(Y, X.arrayOf(V), X.arrayOf(qt), Y), [e, r, n, s]), n.length !== this.ins.length || r.length !== this.ins.length)
      throw new Error("Must supply prevout script and value for all inputs");
    const u = s === g.SIGHASH_DEFAULT ? g.SIGHASH_ALL : s & g.SIGHASH_OUTPUT_MASK, a = (s & g.SIGHASH_INPUT_MASK) === g.SIGHASH_ANYONECANPAY, p = u === g.SIGHASH_NONE, l = u === g.SIGHASH_SINGLE;
    let d = ot, b = ot, w = ot, S = ot, P = ot;
    if (!a) {
      let O = x.withCapacity(36 * this.ins.length);
      this.ins.forEach((T) => {
        O.writeSlice(T.hash), O.writeUInt32(T.index);
      }), d = D(O.end()), O = x.withCapacity(8 * this.ins.length), n.forEach((T) => O.writeUInt64(T)), b = D(O.end()), O = x.withCapacity(r.map(j).reduce((T, F) => T + F)), r.forEach((T) => O.writeVarSlice(T)), w = D(O.end()), O = x.withCapacity(4 * this.ins.length), this.ins.forEach((T) => O.writeUInt32(T.sequence)), S = D(O.end());
    }
    if (p || l) {
      if (l && e < this.outs.length) {
        const O = this.outs[e], T = x.withCapacity(8 + j(O.script));
        T.writeUInt64(O.value), T.writeVarSlice(O.script), P = D(T.end());
      }
    } else {
      if (!this.outs.length)
        throw new Error("Add outputs to the transaction before signing.");
      const O = this.outs.map((F) => 8 + j(F.script)).reduce((F, Gn) => F + Gn), T = x.withCapacity(O);
      this.outs.forEach((F) => {
        T.writeUInt64(F.value), T.writeVarSlice(F.script);
      }), P = D(T.end());
    }
    const k = (i ? 2 : 0) + (o ? 1 : 0), H = 174 - (a ? 49 : 0) - (p ? 32 : 0) + (o ? 32 : 0) + (i ? 37 : 0), _ = x.withCapacity(H);
    if (_.writeUInt8(s), _.writeInt32(this.version), _.writeUInt32(this.locktime), _.writeSlice(d), _.writeSlice(b), _.writeSlice(w), _.writeSlice(S), p || l || _.writeSlice(P), _.writeUInt8(k), a) {
      const O = this.ins[e];
      _.writeSlice(O.hash), _.writeUInt32(O.index), _.writeUInt64(n[e]), _.writeVarSlice(r[e]), _.writeUInt32(O.sequence);
    } else
      _.writeUInt32(e);
    if (o) {
      const O = x.withCapacity(j(o));
      O.writeVarSlice(o), _.writeSlice(D(O.end()));
    }
    return l && _.writeSlice(P), i && (_.writeSlice(i), _.writeUInt8(0), _.writeUInt32(4294967295)), Rt("TapSighash", h.concat([h.from([0]), _.end()]));
  }
  hashForWitnessV0(e, r, n, s) {
    X(z(Y, V, qt, Y), [
      e,
      r,
      n,
      s
    ]);
    let i = h.from([]), o, u = ye, f = ye, a = ye;
    if (s & g.SIGHASH_ANYONECANPAY || (i = h.allocUnsafe(36 * this.ins.length), o = new x(i, 0), this.ins.forEach((l) => {
      o.writeSlice(l.hash), o.writeUInt32(l.index);
    }), f = it(i)), !(s & g.SIGHASH_ANYONECANPAY) && (s & 31) !== g.SIGHASH_SINGLE && (s & 31) !== g.SIGHASH_NONE && (i = h.allocUnsafe(4 * this.ins.length), o = new x(i, 0), this.ins.forEach((l) => {
      o.writeUInt32(l.sequence);
    }), a = it(i)), (s & 31) !== g.SIGHASH_SINGLE && (s & 31) !== g.SIGHASH_NONE) {
      const l = this.outs.reduce((d, b) => d + 8 + j(b.script), 0);
      i = h.allocUnsafe(l), o = new x(i, 0), this.outs.forEach((d) => {
        o.writeUInt64(d.value), o.writeVarSlice(d.script);
      }), u = it(i);
    } else if ((s & 31) === g.SIGHASH_SINGLE && e < this.outs.length) {
      const l = this.outs[e];
      i = h.allocUnsafe(8 + j(l.script)), o = new x(i, 0), o.writeUInt64(l.value), o.writeVarSlice(l.script), u = it(i);
    }
    i = h.allocUnsafe(156 + j(r)), o = new x(i, 0);
    const p = this.ins[e];
    return o.writeInt32(this.version), o.writeSlice(f), o.writeSlice(a), o.writeSlice(p.hash), o.writeUInt32(p.index), o.writeVarSlice(r), o.writeUInt64(n), o.writeUInt32(p.sequence), o.writeSlice(u), o.writeUInt32(this.locktime), o.writeUInt32(s), it(i);
  }
  getHash(e) {
    return e && this.isCoinbase() ? h.alloc(32, 0) : it(this.__toBuffer(void 0, void 0, e));
  }
  getId() {
    return We(this.getHash(!1)).toString("hex");
  }
  toBuffer(e, r) {
    return this.__toBuffer(e, r, !0);
  }
  toHex() {
    return this.toBuffer(void 0, void 0).toString("hex");
  }
  setInputScript(e, r) {
    X(z(It, V), [e, r]), this.ins[e].script = r;
  }
  setWitness(e, r) {
    X(z(It, [V]), [e, r]), this.ins[e].witness = r;
  }
  __toBuffer(e, r, n = !1) {
    e || (e = h.allocUnsafe(this.byteLength(n)));
    const s = new x(e, r || 0);
    s.writeInt32(this.version);
    const i = n && this.hasWitnesses();
    return i && (s.writeUInt8(g.ADVANCED_TRANSACTION_MARKER), s.writeUInt8(g.ADVANCED_TRANSACTION_FLAG)), s.writeVarInt(this.ins.length), this.ins.forEach((o) => {
      s.writeSlice(o.hash), s.writeUInt32(o.index), s.writeVarSlice(o.script), s.writeUInt32(o.sequence);
    }), s.writeVarInt(this.outs.length), this.outs.forEach((o) => {
      wi(o) ? s.writeUInt64(o.value) : s.writeSlice(o.valueBuffer), s.writeVarSlice(o.script);
    }), i && this.ins.forEach((o) => {
      s.writeVector(o.witness);
    }), s.writeUInt32(this.locktime), r !== void 0 ? e.subarray(r, s.offset) : e;
  }
}
g.DEFAULT_SEQUENCE = 4294967295;
g.SIGHASH_DEFAULT = 0;
g.SIGHASH_ALL = 1;
g.SIGHASH_NONE = 2;
g.SIGHASH_SINGLE = 3;
g.SIGHASH_ANYONECANPAY = 128;
g.SIGHASH_OUTPUT_MASK = 3;
g.SIGHASH_INPUT_MASK = 128;
g.ADVANCED_TRANSACTION_MARKER = 0;
g.ADVANCED_TRANSACTION_FLAG = 1;
g.TRUC_VERSION = 3;
g.TRUC_MAX_VSIZE = 1e4;
g.TRUC_CHILD_MAX_VSIZE = 1e3;
function pt(t) {
  return (e) => {
    try {
      return t({ output: e }), !0;
    } catch {
      return !1;
    }
  };
}
const gi = pt(Vt), yi = pt(De), mi = pt(yt), he = pt(Lt), _r = pt(Et), On = pt(mt), An = pt(bt), uo = (t) => t.length === 4 && t[0] === 81 && t[1] === 2 && t[2] === 78 && t[3] === 115;
function xt(t) {
  let e = h.allocUnsafe(0);
  function r(o) {
    e = h.concat([e, h.from(o)]);
  }
  function n(o) {
    const u = e.length, f = Se.encodingLength(o);
    e = h.concat([e, h.allocUnsafe(f)]), Se.encode(o, e, u);
  }
  function s(o) {
    n(o.length), r(o);
  }
  function i(o) {
    n(o.length), o.forEach(s);
  }
  return i(t), e;
}
function Pn(t, e) {
  const r = A(e);
  if (r === null)
    throw new Error("Unknown script error");
  const n = C(t), s = Ht(t), i = Ie(t), o = i?.hybrid ? C(i.hybrid) : void 0, u = i?.uncompressed ? C(i.uncompressed) : void 0;
  return r.findIndex((f) => {
    if (typeof f == "number")
      return !1;
    if (Gt(f, t) || Gt(f, s) || f.equals(n) || i && (Gt(f, i.uncompressed) || Gt(f, i.hybrid) || o && f.equals(o) || u && f.equals(u)))
      return !0;
  });
}
function pe(t, e) {
  return Pn(t, e) !== -1;
}
function bi(t, e) {
  return Ei(t).some((n) => In(n, dt.decode, e));
}
function In(t, e, r) {
  const { hashType: n } = e(t), s = [];
  switch (n & g.SIGHASH_ANYONECANPAY && s.push("addInput"), n & 31) {
    case g.SIGHASH_ALL:
      break;
    case g.SIGHASH_SINGLE:
    case g.SIGHASH_NONE:
      s.push("addOutput"), s.push("setInputSequence");
      break;
  }
  return s.indexOf(r) === -1;
}
function Ei(t) {
  const { partialSig: e } = t;
  let r;
  if (!e || e.length === 0) {
    if (!t.finalScriptSig && !t.finalScriptWitness)
      return [];
    r = Si(t);
  } else
    r = e;
  return r.map((n) => n.signature);
}
function Si(t) {
  const e = t.finalScriptSig ? A(t.finalScriptSig) || [] : [], r = t.finalScriptWitness ? A(t.finalScriptWitness) || [] : [];
  return e.concat(r).filter((n) => h.isBuffer(n) && Q(n)).map((n) => ({ signature: n }));
}
function Tr(t, e, r) {
  const n = Bi(e, t, r);
  try {
    const i = Ni(e, n).concat(n.script).concat(n.controlBlock);
    return { finalScriptWitness: xt(i) };
  } catch (s) {
    throw new Error(`Can not finalize taproot input #${t}: ${s}`);
  }
}
function Dt(t, e) {
  const r = e ? h.from([e]) : h.from([]);
  return h.concat([t, r]);
}
function $(t) {
  return t && !!(t.tapInternalKey || t.tapMerkleRoot || t.tapLeafScript && t.tapLeafScript.length || t.tapBip32Derivation && t.tapBip32Derivation.length || t.witnessUtxo && An(t.witnessUtxo.script));
}
function me(t, e) {
  return t && !!(t.tapInternalKey || t.tapTree || t.tapBip32Derivation && t.tapBip32Derivation.length || e);
}
function Or(t, e, r) {
  xi(t, e, r), ki(t, e, r);
}
function Ar(t, e, r) {
  Ui(t, e, r), _i(t, e);
}
function _i(t, e) {
  if (!e.tapTree && !e.tapInternalKey)
    return;
  const r = e.tapInternalKey || t.tapInternalKey, n = e.tapTree || t.tapTree;
  if (r) {
    const s = t.script, i = Ti(r, n);
    if (s && !s.equals(i))
      throw new Error("Error adding output. Script or address missmatch.");
  }
}
function Ti(t, e) {
  const r = e && Oi(e.leaves), { output: n } = bt({
    internalPubkey: t,
    scriptTree: r
  });
  if (!n)
    throw new Error("Failed to generate taproot script pubkey");
  return n;
}
function Oi(t = []) {
  return t.length === 1 && t[0].depth === 0 ? {
    output: t[0].script,
    version: t[0].leafVersion
  } : vi(t);
}
function Ai(t, e) {
  return Ii(t).some((n) => In(n, Pi, e));
}
function Pi(t) {
  return {
    signature: t.subarray(0, 64),
    hashType: t.subarray(64)[0] || g.SIGHASH_DEFAULT
  };
}
function Ii(t) {
  const e = [];
  if (t.tapKeySig && e.push(t.tapKeySig), t.tapScriptSig && e.push(...t.tapScriptSig.map((r) => r.signature)), !e.length) {
    const r = Hi(t.finalScriptWitness);
    r && e.push(r);
  }
  return e;
}
function Hi(t) {
  if (!t)
    return;
  const e = t.subarray(2);
  if (e.length === 64 || e.length === 65)
    return e;
}
function vi(t) {
  let e;
  for (const r of t)
    if (e = Ue(r, e), !e)
      throw new Error("No room left to insert tapleaf in tree");
  return e;
}
function Ue(t, e, r = 0) {
  if (r > Tn)
    throw new Error("Max taptree depth exceeded.");
  if (t.depth === r)
    return e ? void 0 : {
      output: t.script,
      version: t.leafVersion
    };
  if (ne(e))
    return;
  const n = Ue(t, e && e[0], r + 1);
  if (n)
    return [n, e && e[1]];
  const s = Ue(t, e && e[1], r + 1);
  if (s)
    return [e && e[0], s];
}
function xi(t, e, r) {
  const n = $(t) && gt(e), s = gt(t) && $(e), i = t === e && $(e) && gt(e);
  if (n || s || i)
    throw new Error(`Invalid arguments for Psbt.${r}. Cannot use both taproot and non-taproot fields.`);
}
function Ui(t, e, r) {
  const n = me(t) && gt(e), s = gt(t) && me(e), i = t === e && me(e) && gt(e);
  if (n || s || i)
    throw new Error(`Invalid arguments for Psbt.${r}. Cannot use both taproot and non-taproot fields.`);
}
function ki(t, e, r) {
  if (e.tapMerkleRoot) {
    const n = (e.tapLeafScript || []).every((i) => be(i, e.tapMerkleRoot)), s = (t.tapLeafScript || []).every((i) => be(i, e.tapMerkleRoot));
    if (!n || !s)
      throw new Error(`Invalid arguments for Psbt.${r}. Tapleaf not part of taptree.`);
  } else if (t.tapMerkleRoot && !(e.tapLeafScript || []).every((s) => be(s, t.tapMerkleRoot)))
    throw new Error(`Invalid arguments for Psbt.${r}. Tapleaf not part of taptree.`);
}
function be(t, e) {
  if (!e)
    return !0;
  const r = Z({
    output: t.script,
    version: t.leafVersion
  });
  return te(t.controlBlock, r).equals(e);
}
function Ni(t, e) {
  const r = Z({
    output: e.script,
    version: e.leafVersion
  });
  return (t.tapScriptSig || []).filter((n) => n.leafHash.equals(r)).map((n) => Ci(e.script, n)).sort((n, s) => s.positionInScript - n.positionInScript).map((n) => n.signature);
}
function Ci(t, e) {
  return Object.assign({
    positionInScript: Pn(e.pubkey, t)
  }, e);
}
function Bi(t, e, r) {
  const { tapScriptSig: n } = t;
  if (!n || !n.length)
    throw new Error(`Can not finalize taproot input #${e}. No tapleaf script signature provided.`);
  const s = (t.tapLeafScript || []).sort((i, o) => i.controlBlock.length - o.controlBlock.length).find((i) => Ri(i, n, r));
  if (!s)
    throw new Error(`Can not finalize taproot input #${e}. Signature for tapleaf script not found.`);
  return s;
}
function Ri(t, e, r) {
  const n = Z({
    output: t.script,
    version: t.leafVersion
  });
  return (!r || r.equals(n)) && e.find((i) => i.leafHash.equals(n)) !== void 0;
}
function gt(t) {
  return t && !!(t.redeemScript || t.witnessScript || t.bip32Derivation && t.bip32Derivation.length);
}
const Pr = Se.decode, Fi = {
  network: M,
  maximumFeeRate: 5e3
};
class ke {
  constructor(e = {}, r = new qe.Psbt(new Hn())) {
    this.data = r, this.opts = Object.assign({}, Fi, e), this.__CACHE = {
      __NON_WITNESS_UTXO_TX_CACHE: [],
      __NON_WITNESS_UTXO_BUF_CACHE: [],
      __TX_IN_CACHE: {},
      __TX: this.data.globalMap.unsignedTx.tx,
      __UNSAFE_SIGN_NONSEGWIT: !1
    }, e.version === 3 ? this.setVersionTRUC() : this.data.inputs.length === 0 && this.setVersion(2);
    const n = (s, i, o, u) => {
      Object.defineProperty(s, i, {
        enumerable: o,
        writable: u
      });
    };
    n(this, "__CACHE", !1, !0), n(this, "opts", !1, !0);
  }
  get inputCount() {
    return this.data.inputs.length;
  }
  get version() {
    return this.__CACHE.__TX.version;
  }
  set version(e) {
    this.setVersion(e);
  }
  get locktime() {
    return this.__CACHE.__TX.locktime;
  }
  set locktime(e) {
    this.setLocktime(e);
  }
  get txInputs() {
    return this.__CACHE.__TX.ins.map((e) => ({
      hash: lr(e.hash),
      index: e.index,
      sequence: e.sequence
    }));
  }
  get txOutputs() {
    return this.__CACHE.__TX.outs.map((e) => {
      let r;
      try {
        r = fi(e.script, this.opts.network);
      } catch {
      }
      return {
        script: lr(e.script),
        value: e.value,
        address: r
      };
    });
  }
  static fromBase64(e, r = {}) {
    const n = h.from(e, "base64");
    return this.fromBuffer(n, r);
  }
  static fromHex(e, r = {}) {
    const n = h.from(e, "hex");
    return this.fromBuffer(n, r);
  }
  static fromBuffer(e, r = {}) {
    const n = qe.Psbt.fromBuffer(e, Li), s = new ke(r, n);
    return Mi(s.__CACHE.__TX, s.__CACHE), s;
  }
  combine(...e) {
    return this.data.combine(...e.map((r) => r.data)), this;
  }
  clone() {
    const e = JSON.parse(JSON.stringify(this.opts));
    return ke.fromBuffer(this.data.toBuffer(), e);
  }
  setMaximumFeeRate(e) {
    $t(e), this.opts.maximumFeeRate = e;
  }
  setVersion(e) {
    $t(e), Tt(this.data.inputs, "setVersion");
    const r = this.__CACHE;
    return r.__TX.version = e, r.__EXTRACTED_TX = void 0, this;
  }
  setVersionTRUC() {
    return this.setVersion(g.TRUC_VERSION);
  }
  setLocktime(e) {
    $t(e), Tt(this.data.inputs, "setLocktime");
    const r = this.__CACHE;
    return r.__TX.locktime = e, r.__EXTRACTED_TX = void 0, this;
  }
  setInputSequence(e, r) {
    $t(r), Tt(this.data.inputs, "setInputSequence");
    const n = this.__CACHE;
    if (n.__TX.ins.length <= e)
      throw new Error("Input index too high");
    return n.__TX.ins[e].sequence = r, n.__EXTRACTED_TX = void 0, this;
  }
  addInputs(e, r = !0) {
    return e.forEach((n) => this.addInput(n, r)), this;
  }
  addInput(e, r = !0) {
    if (!e || e.hash === void 0 || e.index === void 0)
      throw new Error("Invalid arguments for Psbt.addInput. Requires single object with at least [hash] and [index]");
    Or(e, e, "addInput"), r && Tt(this.data.inputs, "addInput"), e.witnessScript && re(e.witnessScript);
    const n = this.__CACHE;
    this.data.addInput(e);
    const s = n.__TX.ins[n.__TX.ins.length - 1];
    Un(n, s);
    const i = this.data.inputs.length - 1, o = this.data.inputs[i];
    return o.nonWitnessUtxo && Ce(this.__CACHE, o, i), n.__FEE = void 0, n.__FEE_RATE = void 0, n.__EXTRACTED_TX = void 0, this;
  }
  addOutputs(e) {
    return e.forEach((r) => this.addOutput(r)), this;
  }
  addOutput(e) {
    const r = "address" in e, n = "script" in e;
    if (arguments.length > 1 || !e || e.value === void 0 || !r && !n)
      throw new Error("Invalid arguments for Psbt.addOutput. Requires single object with at least [script or address] and [value]");
    if (Tt(this.data.inputs, "addOutput"), r) {
      const { address: i } = e, { network: o } = this.opts, u = ci(i, o);
      e = Object.assign({}, e, { script: u });
    }
    Ar(e, e, "addOutput");
    const s = this.__CACHE;
    return this.data.addOutput(e), s.__FEE = void 0, s.__FEE_RATE = void 0, s.__EXTRACTED_TX = void 0, this;
  }
  extractTransaction(e, r) {
    if (r && (this.data.inputs = this.data.inputs.filter((i) => !i.partialSig)), !this.data.inputs.every(vn))
      throw new Error("Not finalized");
    const n = this.__CACHE;
    if (e || Wi(this, n, this.opts), n.__EXTRACTED_TX)
      return n.__EXTRACTED_TX;
    const s = n.__TX.clone();
    return Fn(this.data.inputs, s, n, !0, r), s;
  }
  getFeeRate(e = !1) {
    return Ur("__FEE_RATE", "fee rate", this.data.inputs, this.__CACHE, e);
  }
  getFee(e = !1) {
    return Ur("__FEE", "fee", this.data.inputs, this.__CACHE, e);
  }
  finalizeAllInputs() {
    return N.checkForInput(this.data.inputs, 0), Ot(this.data.inputs.length).forEach((e) => this.finalizeInput(e)), this;
  }
  finalizeInput(e, r, n) {
    const s = N.checkForInput(this.data.inputs, e);
    return $(s) ? this._finalizeTaprootInput(e, s, void 0, r) : this._finalizeInput(e, s, r, n ?? !0);
  }
  finalizeTaprootInput(e, r, n = Tr) {
    const s = N.checkForInput(this.data.inputs, e);
    if ($(s))
      return this._finalizeTaprootInput(e, s, r, n);
    throw new Error(`Cannot finalize input #${e}. Not Taproot.`);
  }
  getInputType(e) {
    const r = N.checkForInput(this.data.inputs, e), n = Ln(e, r, this.__CACHE), s = de(n, e, "input", r.redeemScript || Ji(r.finalScriptSig), r.witnessScript || to(r.finalScriptWitness)), i = s.type === "raw" ? "" : s.type + "-", o = Wn(s.meaningfulScript);
    return i + o;
  }
  inputHasPubkey(e, r) {
    const n = N.checkForInput(this.data.inputs, e);
    return Zi(r, n, e, this.__CACHE);
  }
  inputHasHDKey(e, r) {
    const n = N.checkForInput(this.data.inputs, e), s = Hr(r);
    return !!n.bip32Derivation && n.bip32Derivation.some(s);
  }
  outputHasPubkey(e, r) {
    const n = N.checkForOutput(this.data.outputs, e);
    return Qi(r, n, e, this.__CACHE);
  }
  outputHasHDKey(e, r) {
    const n = N.checkForOutput(this.data.outputs, e), s = Hr(r);
    return !!n.bip32Derivation && n.bip32Derivation.some(s);
  }
  validateSignaturesOfAllInputs(e) {
    return N.checkForInput(this.data.inputs, 0), Ot(this.data.inputs.length).map((n) => this.validateSignaturesOfInput(n, e)).reduce((n, s) => s && n, !0);
  }
  validateSignaturesOfInput(e, r, n) {
    const s = this.data.inputs[e];
    return $(s) ? this.validateSignaturesOfTaprootInput(e, r, n) : this._validateSignaturesOfInput(e, r, n);
  }
  signAllInputsHD(e, r = [g.SIGHASH_ALL]) {
    if (!e || !e.publicKey || !e.fingerprint)
      throw new Error("Need HDSigner to sign input");
    const n = [];
    for (const s of Ot(this.data.inputs.length))
      try {
        this.signInputHD(s, e, r), n.push(!0);
      } catch {
        n.push(!1);
      }
    if (n.every((s) => !s))
      throw new Error("No inputs were signed");
    return this;
  }
  signAllInputsHDAsync(e, r = [g.SIGHASH_ALL]) {
    return new Promise((n, s) => {
      if (!e || !e.publicKey || !e.fingerprint)
        return s(new Error("Need HDSigner to sign input"));
      const i = [], o = [];
      for (const u of Ot(this.data.inputs.length))
        o.push(this.signInputHDAsync(u, e, r).then(() => {
          i.push(!0);
        }, () => {
          i.push(!1);
        }));
      return Promise.all(o).then(() => {
        if (i.every((u) => !u))
          return s(new Error("No inputs were signed"));
        n();
      });
    });
  }
  signInputHD(e, r, n = [g.SIGHASH_ALL]) {
    if (!r || !r.publicKey || !r.fingerprint)
      throw new Error("Need HDSigner to sign input");
    return Cr(e, this.data.inputs, r).forEach((i) => this.signInput(e, i, n)), this;
  }
  signInputHDAsync(e, r, n = [g.SIGHASH_ALL]) {
    return new Promise((s, i) => {
      if (!r || !r.publicKey || !r.fingerprint)
        return i(new Error("Need HDSigner to sign input"));
      const u = Cr(e, this.data.inputs, r).map((f) => this.signInputAsync(e, f, n));
      return Promise.all(u).then(() => {
        s();
      }).catch(i);
    });
  }
  signAllInputs(e, r) {
    if (!e || !e.publicKey)
      throw new Error("Need Signer to sign input");
    const n = [];
    for (const s of Ot(this.data.inputs.length))
      try {
        this.signInput(s, e, r), n.push(!0);
      } catch {
        n.push(!1);
      }
    if (n.every((s) => !s))
      throw new Error("No inputs were signed");
    return this;
  }
  signAllInputsAsync(e, r) {
    return new Promise((n, s) => {
      if (!e || !e.publicKey)
        return s(new Error("Need Signer to sign input"));
      const i = [], o = [];
      for (const [u] of this.data.inputs.entries())
        o.push(this.signInputAsync(u, e, r).then(() => {
          i.push(!0);
        }, () => {
          i.push(!1);
        }));
      return Promise.all(o).then(() => {
        if (i.every((u) => !u))
          return s(new Error("No inputs were signed"));
        n();
      });
    });
  }
  signInput(e, r, n) {
    if (!r || !r.publicKey)
      throw new Error("Need Signer to sign input");
    const s = N.checkForInput(this.data.inputs, e);
    return $(s) ? this._signTaprootInput(e, s, r, void 0, n) : this._signInput(e, r, n);
  }
  signTaprootInput(e, r, n, s) {
    if (!r || !r.publicKey)
      throw new Error("Need Signer to sign input");
    const i = N.checkForInput(this.data.inputs, e);
    if ($(i))
      return this._signTaprootInput(e, i, r, n, s);
    throw new Error(`Input #${e} is not of type Taproot.`);
  }
  signInputAsync(e, r, n) {
    return Promise.resolve().then(() => {
      if (!r || !r.publicKey)
        throw new Error("Need Signer to sign input");
      const s = N.checkForInput(this.data.inputs, e);
      return $(s) ? this._signTaprootInputAsync(e, s, r, void 0, n) : this._signInputAsync(e, r, n);
    });
  }
  signTaprootInputAsync(e, r, n, s) {
    return Promise.resolve().then(() => {
      if (!r || !r.publicKey)
        throw new Error("Need Signer to sign input");
      const i = N.checkForInput(this.data.inputs, e);
      if ($(i))
        return this._signTaprootInputAsync(e, i, r, n, s);
      throw new Error(`Input #${e} is not of type Taproot.`);
    });
  }
  toBuffer() {
    return Ee(this.__CACHE), this.data.toBuffer();
  }
  toHex() {
    return Ee(this.__CACHE), this.data.toHex();
  }
  toBase64() {
    return Ee(this.__CACHE), this.data.toBase64();
  }
  updateGlobal(e) {
    return this.data.updateGlobal(e), this;
  }
  updateInput(e, r) {
    return r.witnessScript && re(r.witnessScript), Or(this.data.inputs[e], r, "updateInput"), this.data.updateInput(e, r), r.nonWitnessUtxo && Ce(this.__CACHE, this.data.inputs[e], e), this;
  }
  updateOutput(e, r) {
    const n = this.data.outputs[e];
    return Ar(n, r, "updateOutput"), this.data.updateOutput(e, r), this;
  }
  addUnknownKeyValToGlobal(e) {
    return this.data.addUnknownKeyValToGlobal(e), this;
  }
  addUnknownKeyValToInput(e, r) {
    return this.data.addUnknownKeyValToInput(e, r), this;
  }
  addUnknownKeyValToOutput(e, r) {
    return this.data.addUnknownKeyValToOutput(e, r), this;
  }
  clearFinalizedInput(e) {
    return this.data.clearFinalizedInput(e), this;
  }
  checkTaprootHashesForSig(e, r, n, s, i) {
    if (typeof n.signSchnorr != "function")
      throw new Error(`Need Schnorr Signer to sign taproot input #${e}.`);
    const o = h.isBuffer(n.publicKey) ? n.publicKey : h.from(n.publicKey), u = Ne(e, r, this.data.inputs, o, this.__CACHE, s, i);
    if (!u || !u.length)
      throw new Error(`Can not sign for input #${e} with the key ${o.toString("hex")}`);
    return u;
  }
  _finalizeInput(e, r, n = Di, s = !0) {
    const { script: i, isP2SH: o, isP2WSH: u, isSegwit: f } = ji(e, r, this.__CACHE);
    if (!i)
      throw new Error(`No script found for input #${e}`);
    Gi(r);
    const { finalScriptSig: a, finalScriptWitness: p } = n(e, r, i, f, o, u, s);
    if (a && this.data.updateInput(e, { finalScriptSig: a }), p && this.data.updateInput(e, { finalScriptWitness: p }), !a && !p)
      throw new Error(`Unknown error finalizing input #${e}`);
    return this.data.clearFinalizedInput(e), this;
  }
  _finalizeTaprootInput(e, r, n, s = Tr) {
    if (!r.witnessUtxo)
      throw new Error(`Cannot finalize input #${e}. Missing witness utxo.`);
    if (r.tapKeySig) {
      const i = bt({
        output: r.witnessUtxo.script,
        signature: r.tapKeySig
      });
      if (!i.witness)
        throw new Error("Cannot finalize taproot key spend");
      const o = xt(i.witness);
      this.data.updateInput(e, { finalScriptWitness: o });
    } else {
      const { finalScriptWitness: i } = s(e, r, n);
      this.data.updateInput(e, { finalScriptWitness: i });
    }
    return this.data.clearFinalizedInput(e), this;
  }
  _validateSignaturesOfInput(e, r, n) {
    const s = this.data.inputs[e], i = (s || {}).partialSig;
    if (!s || !i || i.length < 1)
      throw new Error("No signatures to validate");
    if (typeof r != "function")
      throw new Error("Need validator function to validate signatures");
    const o = n ? i.filter((l) => l.pubkey.equals(n)) : i;
    if (o.length < 1)
      throw new Error("No signatures for this pubkey");
    const u = [];
    let f, a, p;
    for (const l of o) {
      const d = dt.decode(l.signature), { hash: b, script: w } = p !== d.hashType || !f || !a ? Nn(e, Object.assign({}, s, {
        sighashType: d.hashType
      }), this.__CACHE, !0) : { hash: f, script: a };
      p = d.hashType, f = b, a = w, xn(l.pubkey, w, "verify"), u.push(r(l.pubkey, b, d.signature));
    }
    return u.every((l) => l);
  }
  validateSignaturesOfTaprootInput(e, r, n) {
    const s = this.data.inputs[e], i = (s || {}).tapKeySig, o = (s || {}).tapScriptSig;
    if (!s && !i && !(o && !o.length))
      throw new Error("No signatures to validate");
    if (typeof r != "function")
      throw new Error("Need validator function to validate signatures");
    n = n && Ht(n);
    const u = n ? Ne(e, s, this.data.inputs, n, this.__CACHE) : qi(e, s, this.data.inputs, this.__CACHE);
    if (!u.length)
      throw new Error("No signatures for this pubkey");
    const f = u.find((p) => !p.leafHash);
    let a = 0;
    if (i && f) {
      if (!r(f.pubkey, f.hash, Nr(i)))
        return !1;
      a++;
    }
    if (o)
      for (const p of o) {
        const l = u.find((d) => p.pubkey.equals(d.pubkey));
        if (l) {
          if (!r(p.pubkey, l.hash, Nr(p.signature)))
            return !1;
          a++;
        }
      }
    return a > 0;
  }
  _signInput(e, r, n = [g.SIGHASH_ALL]) {
    const s = h.isBuffer(r.publicKey) ? r.publicKey : h.from(r.publicKey), { hash: i, sighashType: o } = kr(this.data.inputs, e, s, this.__CACHE, n), u = r.sign(i), f = [
      {
        pubkey: s,
        signature: dt.encode(h.isBuffer(u) ? u : h.from(u), o)
      }
    ];
    return this.data.updateInput(e, { partialSig: f }), this;
  }
  _signTaprootInput(e, r, n, s, i = [g.SIGHASH_DEFAULT]) {
    const o = h.isBuffer(n.publicKey) ? n.publicKey : h.from(n.publicKey), u = this.checkTaprootHashesForSig(e, r, n, s, i), f = n.signSchnorr.bind(n), a = (d) => h.isBuffer(d) ? d : h.from(d), p = u.filter((d) => !d.leafHash).map((d) => Dt(a(f(d.hash)), r.sighashType))[0], l = u.filter((d) => !!d.leafHash).map((d) => ({
      pubkey: Ht(o),
      signature: Dt(a(f(d.hash)), r.sighashType),
      leafHash: d.leafHash
    }));
    return p && this.data.updateInput(e, { tapKeySig: p }), l.length && this.data.updateInput(e, { tapScriptSig: l }), this;
  }
  _signInputAsync(e, r, n = [g.SIGHASH_ALL]) {
    const s = h.isBuffer(r.publicKey) ? r.publicKey : h.from(r.publicKey), { hash: i, sighashType: o } = kr(this.data.inputs, e, s, this.__CACHE, n);
    return Promise.resolve(r.sign(i)).then((u) => {
      const f = h.isBuffer(u) ? u : h.from(u), a = [
        {
          pubkey: s,
          signature: dt.encode(f, o)
        }
      ];
      this.data.updateInput(e, { partialSig: a });
    });
  }
  async _signTaprootInputAsync(e, r, n, s, i = [g.SIGHASH_DEFAULT]) {
    const o = h.isBuffer(n.publicKey) ? n.publicKey : h.from(n.publicKey), u = this.checkTaprootHashesForSig(e, r, n, s, i), f = n.signSchnorr.bind(n), a = (w) => h.isBuffer(w) ? w : h.from(w), p = [], l = u.filter((w) => !w.leafHash)[0];
    if (l) {
      const w = Promise.resolve(f(l.hash)).then((S) => ({
        tapKeySig: Dt(a(S), r.sighashType)
      }));
      p.push(w);
    }
    const d = u.filter((w) => !!w.leafHash);
    if (d.length) {
      const w = d.map(async (S) => {
        const P = await f(S.hash);
        return { tapScriptSig: [
          {
            pubkey: Ht(o),
            signature: Dt(a(P), r.sighashType),
            leafHash: S.leafHash
          }
        ] };
      });
      p.push(...w);
    }
    const b = await Promise.all(p);
    for (const w of b)
      this.data.updateInput(e, w);
  }
}
const Li = (t) => new Hn(t);
class Hn {
  constructor(e = h.from([2, 0, 0, 0, 0, 0, 0, 0, 0, 0])) {
    this.tx = g.fromBuffer(e), Ki(this.tx), Object.defineProperty(this, "tx", {
      enumerable: !1,
      writable: !0
    });
  }
  getInputOutputCounts() {
    return {
      inputCount: this.tx.ins.length,
      outputCount: this.tx.outs.length
    };
  }
  addInput(e) {
    if (e.hash === void 0 || e.index === void 0 || !h.isBuffer(e.hash) && typeof e.hash != "string" || typeof e.index != "number")
      throw new Error("Error adding input.");
    const r = typeof e.hash == "string" ? We(h.from(e.hash, "hex")) : e.hash;
    this.tx.addInput(r, e.index, e.sequence);
  }
  addOutput(e) {
    if (e.script === void 0 || e.value === void 0 || !h.isBuffer(e.script) || typeof e.value != "number")
      throw new Error("Error adding output.");
    this.tx.addOutput(e.script, e.value);
  }
  toBuffer() {
    return this.tx.toBuffer();
  }
}
function Vi(t, e, r) {
  switch (r) {
    case "pubkey":
    case "pubkeyhash":
    case "witnesspubkeyhash":
      return Ir(1, t.partialSig);
    case "multisig": {
      const n = Vt({
        output: e
      });
      return Ir(n.m, t.partialSig, n.pubkeys);
    }
    case "nonstandard":
      return !0;
    default:
      return !1;
  }
}
function Ee(t) {
  if (t.__UNSAFE_SIGN_NONSEGWIT)
    throw new Error("Not BIP174 compliant, can not export");
}
function Ir(t, e, r) {
  if (!e)
    return !1;
  let n;
  if (r ? n = r.map((s) => {
    const i = eo(s);
    return e.find((o) => o.pubkey.equals(i));
  }).filter((s) => !!s) : n = e, n.length > t)
    throw new Error("Too many signatures");
  return n.length === t;
}
function vn(t) {
  return !!t.finalScriptSig || !!t.finalScriptWitness;
}
function Hr(t) {
  return (e) => {
    const r = h.isBuffer(t.fingerprint) ? t.fingerprint : h.from(t.fingerprint);
    if (!e.masterFingerprint.equals(r))
      return !1;
    const n = t.derivePath(e.path).publicKey;
    return !!(h.isBuffer(n) ? n : h.from(n)).equals(e.pubkey);
  };
}
function $t(t) {
  if (typeof t != "number" || t !== Math.floor(t) || t > 4294967295 || t < 0)
    throw new Error("Invalid 32 bit integer");
}
function Wi(t, e, r) {
  const n = e.__FEE_RATE || t.getFeeRate(), s = e.__EXTRACTED_TX.virtualSize(), i = n * s;
  if (n >= r.maximumFeeRate)
    throw new Error(`Warning: You are paying around ${(i / 1e8).toFixed(8)} in fees, which is ${n} satoshi per byte for a transaction with a VSize of ${s} bytes (segwit counted as 0.25 byte per byte). Use setMaximumFeeRate method to raise your threshold, or pass true to the first arg of extractTransaction.`);
}
function Tt(t, e) {
  t.forEach((r) => {
    if ($(r) ? Ai(r, e) : bi(r, e))
      throw new Error("Can not modify transaction, signatures exist.");
  });
}
function Gi(t) {
  if (!t.sighashType || !t.partialSig)
    return;
  const { partialSig: e, sighashType: r } = t;
  e.forEach((n) => {
    const { hashType: s } = dt.decode(n.signature);
    if (r !== s)
      throw new Error("Signature sighash does not match input sighash type");
  });
}
function xn(t, e, r) {
  if (!pe(t, e))
    throw new Error(`Can not ${r} for this input with the key ${t.toString("hex")}`);
}
function Ki(t) {
  if (!t.ins.every((r) => r.script && r.script.length === 0 && r.witness && r.witness.length === 0))
    throw new Error("Format Error: Transaction ScriptSigs are not empty");
}
function Mi(t, e) {
  t.ins.forEach((r) => {
    Un(e, r);
  });
}
function Un(t, e) {
  const r = `${We(h.from(e.hash)).toString("hex")}:${e.index}`;
  if (t.__TX_IN_CACHE[r])
    throw new Error("Duplicate input detected.");
  t.__TX_IN_CACHE[r] = 1;
}
function kn(t, e) {
  return (r, n, s, i) => {
    const o = t({
      redeem: { output: s }
    }).output;
    if (!n.equals(o))
      throw new Error(`${e} for ${i} #${r} doesn't match the scriptPubKey in the prevout`);
  };
}
const vr = kn(mt, "Redeem script"), xr = kn(Et, "Witness script");
function Ur(t, e, r, n, s = !1) {
  if (!r.every(vn))
    throw new Error(`PSBT must be finalized to calculate ${e}`);
  if (t === "__FEE_RATE" && n.__FEE_RATE)
    return n.__FEE_RATE;
  if (t === "__FEE" && n.__FEE)
    return n.__FEE;
  let i, o = !0;
  n.__EXTRACTED_TX ? (i = n.__EXTRACTED_TX, o = !1) : i = n.__TX.clone(), Fn(r, i, n, o, s);
  const u = t === "__FEE_RATE" ? n.__FEE_RATE : n.__FEE;
  if (u === void 0)
    throw new Error(`Failed to calculate ${e}`);
  return u;
}
function Di(t, e, r, n, s, i, o = !0, u) {
  const f = Wn(r);
  if (!Vi(e, r, f) && o)
    throw new Error(`Can not finalize input #${t}`);
  return $i(r, f, e.partialSig, n, s, i, u);
}
function $i(t, e, r, n, s, i, o) {
  let u, f;
  const a = Xi(t, e, r), p = i ? Et({ redeem: a }) : null, l = s ? mt({ redeem: p || a }) : null;
  return n ? (p ? f = xt(p.witness) : a ? f = xt(a.witness) : f = xt(o ?? [h.from([0])]), l && (u = l?.input)) : l ? u = l?.input : a ? u = a.input : u = Array.isArray(o) && o[0] ? o[0] : h.from([1]), {
    finalScriptSig: u,
    finalScriptWitness: f
  };
}
function kr(t, e, r, n, s) {
  const i = N.checkForInput(t, e), { hash: o, sighashType: u, script: f } = Nn(e, i, n, !1, s);
  return xn(r, f, "sign"), {
    hash: o,
    sighashType: u
  };
}
function Nn(t, e, r, n, s) {
  const i = r.__TX, o = e.sighashType || g.SIGHASH_ALL;
  Bn(o, s);
  let u, f;
  if (e.nonWitnessUtxo) {
    const l = le(r, e, t), d = i.ins[t].hash, b = l.getHash();
    if (!d.equals(b))
      throw new Error(`Non-witness UTXO hash for input #${t} doesn't match the hash specified in the prevout`);
    const w = i.ins[t].index;
    f = l.outs[w];
  } else if (e.witnessUtxo)
    f = e.witnessUtxo;
  else
    throw new Error("Need a Utxo input item for signing");
  const { meaningfulScript: a, type: p } = de(f.script, t, "input", e.redeemScript, e.witnessScript);
  if (["p2sh-p2wsh", "p2wsh"].indexOf(p) >= 0)
    u = i.hashForWitnessV0(t, a, f.value, o);
  else if (he(a)) {
    const l = yt({
      hash: a.subarray(2)
    }).output;
    u = i.hashForWitnessV0(t, l, f.value, o);
  } else {
    if (e.nonWitnessUtxo === void 0 && !r.__UNSAFE_SIGN_NONSEGWIT)
      throw new Error(`Input #${t} has witnessUtxo but non-segwit script: ` + a.toString("hex"));
    !n && r.__UNSAFE_SIGN_NONSEGWIT && console.warn(`Warning: Signing non-segwit inputs without the full parent transaction means there is a chance that a miner could feed you incorrect information to trick you into paying large fees. This behavior is the same as Psbt's predecessor (TransactionBuilder - now removed) when signing non-segwit scripts. You are not able to export this Psbt with toBuffer|toBase64|toHex since it is not BIP174 compliant.
*********************
PROCEED WITH CAUTION!
*********************`), u = i.hashForSignature(t, a, o);
  }
  return {
    script: a,
    sighashType: o,
    hash: u
  };
}
function qi(t, e, r, n) {
  const s = [];
  if (e.tapInternalKey) {
    const o = Cn(t, e, n);
    o && s.push(o);
  }
  if (e.tapScriptSig) {
    const o = e.tapScriptSig.map((u) => u.pubkey);
    s.push(...o);
  }
  return s.map((o) => Ne(t, e, r, o, n)).flat();
}
function Cn(t, e, r) {
  const { script: n } = $e(t, e, r);
  return An(n) ? h.from(n.subarray(2, 34)) : null;
}
function Nr(t) {
  return t.length === 64 ? t : h.from(t.subarray(0, 64));
}
function Ne(t, e, r, n, s, i, o) {
  const u = s.__TX, f = e.sighashType || g.SIGHASH_DEFAULT;
  Bn(f, o);
  const a = r.map((w, S) => $e(S, w, s)), p = a.map((w) => w.script), l = a.map((w) => w.value), d = [];
  if (e.tapInternalKey && !i) {
    const w = Cn(t, e, s) || h.from([]);
    if (Ht(n).equals(w)) {
      const S = u.hashForWitnessV1(t, p, l, f);
      d.push({ pubkey: n, hash: S });
    }
  }
  const b = (e.tapLeafScript || []).filter((w) => pe(n, w.script)).map((w) => {
    const S = Z({
      output: w.script,
      version: w.leafVersion
    });
    return Object.assign({ hash: S }, w);
  }).filter((w) => !i || i.equals(w.hash)).map((w) => {
    const S = u.hashForWitnessV1(t, p, l, f, w.hash);
    return {
      pubkey: n,
      hash: S,
      leafHash: w.hash
    };
  });
  return d.concat(b);
}
function Bn(t, e) {
  if (e && e.indexOf(t) < 0) {
    const r = Yi(t);
    throw new Error(`Sighash type is not allowed. Retry the sign method passing the sighashTypes array of whitelisted types. Sighash type: ${r}`);
  }
}
function Xi(t, e, r) {
  let n;
  switch (e) {
    case "multisig": {
      const s = zi(t, r);
      n = Vt({
        output: t,
        signatures: s
      });
      break;
    }
    case "pubkey":
      n = De({
        output: t,
        signature: r[0].signature
      });
      break;
    case "pubkeyhash":
      n = yt({
        output: t,
        pubkey: r[0].pubkey,
        signature: r[0].signature
      });
      break;
    case "witnesspubkeyhash":
      n = Lt({
        output: t,
        pubkey: r[0].pubkey,
        signature: r[0].signature
      });
      break;
  }
  return n;
}
function ji(t, e, r) {
  const n = r.__TX, s = {
    script: null,
    isSegwit: !1,
    isP2SH: !1,
    isP2WSH: !1
  };
  if (s.isP2SH = !!e.redeemScript, s.isP2WSH = !!e.witnessScript, e.witnessScript)
    s.script = e.witnessScript;
  else if (e.redeemScript)
    s.script = e.redeemScript;
  else if (e.nonWitnessUtxo) {
    const i = le(r, e, t), o = n.ins[t].index;
    s.script = i.outs[o].script;
  } else e.witnessUtxo && (s.script = e.witnessUtxo.script);
  if (e.witnessScript || he(s.script))
    s.isSegwit = !0;
  else
    try {
      const i = s.script;
      if (!i)
        throw new TypeError("Invalid script for segwit address");
      s.isSegwit = ni(i);
    } catch {
    }
  return s;
}
function Cr(t, e, r) {
  const n = N.checkForInput(e, t);
  if (!n.bip32Derivation || n.bip32Derivation.length === 0)
    throw new Error("Need bip32Derivation to sign with HD");
  const s = n.bip32Derivation.map((i) => {
    if (i.masterFingerprint.equals(r.fingerprint))
      return i;
  }).filter((i) => !!i);
  if (s.length === 0)
    throw new Error("Need one bip32Derivation masterFingerprint to match the HDSigner fingerprint");
  return s.map((i) => {
    const o = r.derivePath(i.path);
    if (!i.pubkey.equals(o.publicKey))
      throw new Error("pubkey did not match bip32Derivation");
    return o;
  });
}
function zi(t, e) {
  return Vt({ output: t }).pubkeys.map((n) => (e.filter((s) => s.pubkey.equals(n))[0] || {}).signature).filter((n) => !!n);
}
function Rn(t) {
  let e = 0;
  function r(o) {
    return e += o, t.subarray(e - o, e);
  }
  function n() {
    const o = Pr(t, e);
    return e += Pr.bytes, o;
  }
  function s() {
    return r(n());
  }
  function i() {
    const o = n(), u = [];
    for (let f = 0; f < o; f++)
      u.push(s());
    return u;
  }
  return i();
}
function Yi(t) {
  let e = t & g.SIGHASH_ANYONECANPAY ? "SIGHASH_ANYONECANPAY | " : "";
  switch (t & 31) {
    case g.SIGHASH_ALL:
      e += "SIGHASH_ALL";
      break;
    case g.SIGHASH_SINGLE:
      e += "SIGHASH_SINGLE";
      break;
    case g.SIGHASH_NONE:
      e += "SIGHASH_NONE";
      break;
  }
  return e;
}
function Ce(t, e, r) {
  t.__NON_WITNESS_UTXO_BUF_CACHE[r] = e.nonWitnessUtxo, t.__NON_WITNESS_UTXO_TX_CACHE[r] = g.fromBuffer(e.nonWitnessUtxo);
  const n = t, s = r;
  delete e.nonWitnessUtxo, Object.defineProperty(e, "nonWitnessUtxo", {
    enumerable: !0,
    get() {
      const i = n.__NON_WITNESS_UTXO_BUF_CACHE[s], o = n.__NON_WITNESS_UTXO_TX_CACHE[s];
      if (i !== void 0)
        return i;
      {
        const u = o.toBuffer();
        return n.__NON_WITNESS_UTXO_BUF_CACHE[s] = u, u;
      }
    },
    set(i) {
      n.__NON_WITNESS_UTXO_BUF_CACHE[s] = i;
    }
  });
}
function Fn(t, e, r, n, s) {
  let i = 0;
  t.forEach((a, p) => {
    if (n && a.finalScriptSig && (e.ins[p].script = a.finalScriptSig), n && a.finalScriptWitness && (e.ins[p].witness = Rn(a.finalScriptWitness)), a.witnessUtxo)
      i += a.witnessUtxo.value;
    else if (a.nonWitnessUtxo) {
      const l = le(r, a, p), d = e.ins[p].index, b = l.outs[d];
      i += b.value;
    }
  });
  const o = e.outs.reduce((a, p) => a + p.value, 0), u = i - o;
  if (!s && u < 0)
    throw new Error(`Outputs are spending more than Inputs ${i} < ${o}`);
  const f = e.virtualSize();
  r.__FEE = u, r.__EXTRACTED_TX = e, r.__FEE_RATE = Math.floor(u / f);
}
function le(t, e, r) {
  const n = t.__NON_WITNESS_UTXO_TX_CACHE;
  return n[r] || Ce(t, e, r), n[r];
}
function Ln(t, e, r) {
  const { script: n } = $e(t, e, r);
  return n;
}
function $e(t, e, r) {
  if (e.witnessUtxo !== void 0)
    return {
      script: e.witnessUtxo.script,
      value: e.witnessUtxo.value
    };
  if (e.nonWitnessUtxo !== void 0) {
    const s = le(r, e, t).outs[r.__TX.ins[t].index];
    return { script: s.script, value: s.value };
  } else
    throw new Error("Can't find pubkey in input without Utxo data");
}
function Zi(t, e, r, n) {
  const s = Ln(r, e, n), { meaningfulScript: i } = de(s, r, "input", e.redeemScript, e.witnessScript);
  return pe(t, i);
}
function Qi(t, e, r, n) {
  const s = n.__TX.outs[r].script, { meaningfulScript: i } = de(s, r, "output", e.redeemScript, e.witnessScript);
  return pe(t, i);
}
function Ji(t) {
  if (!t)
    return;
  const e = A(t);
  if (!e)
    return;
  const r = e[e.length - 1];
  if (!(!h.isBuffer(r) || Vn(r) || ro(r) || !A(r)))
    return r;
}
function to(t) {
  if (!t)
    return;
  const e = Rn(t), r = e[e.length - 1];
  if (!(Vn(r) || !A(r)))
    return r;
}
function eo(t) {
  if (t.length === 65) {
    const e = t[64] & 1, r = h.from(t.subarray(0, 33));
    return r[0] = 2 | e, r;
  }
  return h.from(t);
}
function Vn(t) {
  return t.length === 33 && zr(t);
}
function ro(t) {
  return Q(t);
}
function de(t, e, r, n, s) {
  const i = On(t), o = i && n && _r(n), u = _r(t);
  if (i && n === void 0)
    throw new Error("scriptPubkey is P2SH but redeemScript missing");
  if ((u || o) && s === void 0)
    throw new Error("scriptPubkey or redeemScript is P2WSH but witnessScript missing");
  let f;
  return o ? (f = s, vr(e, t, n, r), xr(e, n, s, r), re(f)) : u ? (f = s, xr(e, t, s, r), re(f)) : i ? (f = n, vr(e, t, n, r)) : f = t, {
    meaningfulScript: f,
    type: o ? "p2sh-p2wsh" : i ? "p2sh" : u ? "p2wsh" : "raw"
  };
}
function re(t) {
  if (he(t) || On(t))
    throw new Error("P2WPKH or P2SH can not be contained within P2WSH");
}
function Wn(t) {
  return he(t) ? "witnesspubkeyhash" : mi(t) ? "pubkeyhash" : gi(t) ? "multisig" : yi(t) ? "pubkey" : "nonstandard";
}
function Ot(t) {
  return [...Array(t).keys()];
}
const fo = {
  payments: hi,
  script: hs,
  crypto: xs
};
export {
  so as A,
  An as B,
  Pn as C,
  Di as D,
  mi as E,
  he as F,
  ni as G,
  _r as H,
  On as I,
  uo as J,
  yi as K,
  gi as L,
  ke as M,
  Hs as N,
  K as P,
  g as T,
  Et as a,
  M as b,
  Lt as c,
  bt as d,
  fo as e,
  v as f,
  fi as g,
  ci as h,
  oo as i,
  mt as j,
  yt as k,
  De as l,
  A as m,
  Be as n,
  E as o,
  Vt as p,
  oi as q,
  as as r,
  Ie as s,
  Ht as t,
  it as u,
  C as v,
  Is as w,
  Zs as x,
  D as y,
  io as z
};
