// @btc-vision/visitor-as/dist/examples/toString.js
import { ASTTransformVisitor } from "../index.js";
import { SimpleParser } from "../simpleParser.js";
import { not, isStdlib, isMethodNamed, getName } from '../utils.js';
class ToStringCallTransform extends ASTTransformVisitor {
    currentClass;
    fields;
    visitFieldDeclaration(node) {
        const name = getName(node);
        let rhs = `this.${name}.toString()`;
        this.fields.push(`sb.push(\`${name}: \${${rhs}}\`)`);
        super.visitFieldDeclaration(node);
    }
    visitClassDeclaration(node) {
        if (!node.members || node.members.some(isMethodNamed("toString"))) {
            super.visitClassDeclaration(node);
            return;
        }
        this.currentClass = node;
        this.fields = [];
        this.visit(node.members); // will visit fields and methods
        const method = `
  toString(): string {
    const sb = new Array<string>();
    ${this.fields.join(";\n\t")};
    return \`${getName(node)}:\\n\\t\${sb.join("\\n\\t")}\`
  }
    `;
        let member = SimpleParser.parseClassMember(method, node);
        node.members.push(member);
        super.visitClassDeclaration(node);
    }
    afterParse(_) {
        let sources = _.sources.filter(not(isStdlib));
        this.visit(sources);
    }
}
export default ToStringCallTransform;
