// @btc-vision/transaction/build/consensus/Consensus.js
export var Consensus;
(function (Consensus) {
    Consensus[Consensus["Roswell"] = 0] = "Roswell";
    Consensus[Consensus["Rachel"] = 1] = "Rachel";
    Consensus[Consensus["Kecksburg"] = 2] = "Kecksburg";
    Consensus[Consensus["Phoenix"] = 3] = "Phoenix";
    Consensus[Consensus["Aurora"] = 4] = "Aurora";
    Consensus[Consensus["Rendlesham"] = 5] = "Rendlesham";
    Consensus[Consensus["Lazar"] = 6] = "Lazar";
    Consensus[Consensus["ShagHarbor"] = 7] = "ShagHarbor";
    Consensus[Consensus["Exeter"] = 8] = "Exeter";
    Consensus[Consensus["Stephenville"] = 9] = "Stephenville";
    Consensus[Consensus["Valensole"] = 10] = "Valensole";
    Consensus[Consensus["Socorro"] = 11] = "Socorro";
    Consensus[Consensus["Pascagoula"] = 12] = "Pascagoula";
    Consensus[Consensus["Tehran"] = 13] = "Tehran";
    Consensus[Consensus["Westall"] = 14] = "Westall";
    Consensus[Consensus["Hopkinsville"] = 15] = "Hopkinsville";
    Consensus[Consensus["Belgium"] = 16] = "Belgium";
    Consensus[Consensus["Breeze"] = 17] = "Breeze";
    Consensus[Consensus["Flatwoods"] = 18] = "Flatwoods";
    Consensus[Consensus["Maury"] = 20] = "Maury";
    Consensus[Consensus["Varginha"] = 21] = "Varginha";
    Consensus[Consensus["Trindade"] = 22] = "Trindade";
    Consensus[Consensus["Levelland"] = 23] = "Levelland";
    Consensus[Consensus["Wanaque"] = 24] = "Wanaque";
    Consensus[Consensus["Coyame"] = 25] = "Coyame";
    Consensus[Consensus["Delphos"] = 26] = "Delphos";
})(Consensus || (Consensus = {}));
