// @btc-vision/transaction/build/transaction/builders/InteractionTransaction.js
import { TransactionType } from '../enums/TransactionType.js';
import { SharedInteractionTransaction } from './SharedInteractionTransaction.js';
import { FeaturePriority, Features } from '../../generators/Features.js';
export class InteractionTransaction extends SharedInteractionTransaction {
    constructor(parameters) {
        super(parameters);
        this.type = TransactionType.INTERACTION;
        this.tapLeafScript = null;
        if (!parameters.contract) {
            throw new Error('parameters.contract is required for interaction transaction.');
        }
        this.contractSecret = Buffer.from(parameters.contract.replace('0x', ''), 'hex');
        if (this.contractSecret.length !== 32) {
            throw new Error('Invalid contract secret length. Expected 32 bytes.');
        }
        if (parameters.compiledTargetScript) {
            if (Buffer.isBuffer(parameters.compiledTargetScript)) {
                this.compiledTargetScript = parameters.compiledTargetScript;
            }
            else if (typeof parameters.compiledTargetScript === 'string') {
                this.compiledTargetScript = Buffer.from(parameters.compiledTargetScript, 'hex');
            }
            else {
                throw new Error('Invalid compiled target script format.');
            }
        }
        else {
            this.compiledTargetScript = this.calldataGenerator.compile(this.calldata, this.contractSecret, this.challenge, this.priorityFee, this.generateFeatures(parameters));
        }
        this.scriptTree = this.getScriptTree();
        this.internalInit();
    }
    generateFeatures(parameters) {
        const features = [];
        if (parameters.loadedStorage) {
            features.push({
                priority: FeaturePriority.ACCESS_LIST,
                opcode: Features.ACCESS_LIST,
                data: parameters.loadedStorage,
            });
        }
        const submission = parameters.challenge.getSubmission();
        if (submission) {
            features.push({
                priority: FeaturePriority.EPOCH_SUBMISSION,
                opcode: Features.EPOCH_SUBMISSION,
                data: submission,
            });
        }
        if (parameters.revealMLDSAPublicKey && !parameters.linkMLDSAPublicKeyToAddress) {
            throw new Error('To reveal the MLDSA public key, you must set linkMLDSAPublicKeyToAddress to true.');
        }
        if (parameters.linkMLDSAPublicKeyToAddress) {
            this.generateMLDSALinkRequest(parameters, features);
        }
        return features;
    }
}
