// @btc-vision/bitcoin/build/hooks/AdvancedSignatureManager.js
export class AdvancedSignatureManager {
    constructor() {
        this.cacheBySigner = new Map();
        this.registry = new FinalizationRegistry((entry) => {
            const set = this.cacheBySigner.get(entry.pubKey);
            if (set) {
                set.delete(entry);
                if (set.size === 0) {
                    this.cacheBySigner.delete(entry.pubKey);
                }
            }
        });
    }
    static getInstance() {
        if (!AdvancedSignatureManager.instance) {
            AdvancedSignatureManager.instance = new AdvancedSignatureManager();
        }
        return AdvancedSignatureManager.instance;
    }
    addSignature(pubKey, data, signature) {
        const entry = {
            pubKey,
            dataRef: new WeakRef(data),
            signature,
        };
        if (!this.cacheBySigner.has(pubKey)) {
            this.cacheBySigner.set(pubKey, new Set());
        }
        const set = this.cacheBySigner.get(pubKey);
        set.add(entry);
        this.registry.register(data, entry);
        return signature;
    }
    getSignature(pubKey, data) {
        const set = this.cacheBySigner.get(pubKey);
        if (!set)
            return undefined;
        for (const entry of set) {
            const cachedData = entry.dataRef.deref();
            if (cachedData && cachedData.equals(data)) {
                return entry.signature;
            }
        }
        return undefined;
    }
    clearCache() {
        this.cacheBySigner.clear();
    }
    clearCacheForSigner(pubKey) {
        this.cacheBySigner.delete(pubKey);
    }
}
