// @btc-vision/bitcoin/build/hooks/SignatureManager.js
export class SignatureManager {
    constructor() {
        this.cacheByLength = new Map();
        this.registry = new FinalizationRegistry((entry) => {
            const set = this.cacheByLength.get(entry.length);
            if (set) {
                set.delete(entry);
                if (set.size === 0) {
                    this.cacheByLength.delete(entry.length);
                }
            }
        });
    }
    addSignature(data, signature) {
        const length = data.length;
        const entry = {
            length,
            dataRef: new WeakRef(data),
            signature,
        };
        if (!this.cacheByLength.has(length)) {
            this.cacheByLength.set(length, new Set());
        }
        const set = this.cacheByLength.get(length);
        set.add(entry);
        this.registry.register(data, entry);
        return signature;
    }
    getSignature(data) {
        const length = data.length;
        const set = this.cacheByLength.get(length);
        if (!set)
            return undefined;
        for (const entry of set) {
            const cachedData = entry.dataRef.deref();
            if (cachedData && cachedData.equals(data)) {
                return entry.signature;
            }
        }
        return undefined;
    }
    clearCache() {
        this.cacheByLength.clear();
    }
}
