// opnet/build/threading/JSONThreader.js
import { BaseThreader } from './SharedThreader.js';
import { jsonWorkerScript } from './worker-scripts/JSONWorker.js';
let _isServiceWorker = null;
function checkIsServiceWorker() {
    if (_isServiceWorker !== null)
        return _isServiceWorker;
    if (typeof __IS_SERVICE_WORKER__ !== 'undefined' && __IS_SERVICE_WORKER__) {
        _isServiceWorker = true;
        return true;
    }
    if (typeof ServiceWorkerGlobalScope !== 'undefined' &&
        self instanceof ServiceWorkerGlobalScope) {
        _isServiceWorker = true;
        return true;
    }
    _isServiceWorker = false;
    return false;
}
const isServiceWorker = checkIsServiceWorker();
export class JsonThreader extends BaseThreader {
    workerScript = jsonWorkerScript;
    threadingThreshold;
    constructor(options = {}) {
        super(options);
        this.threadingThreshold = options.threadingThreshold ?? 16_384;
    }
    async parse(json) {
        if (isServiceWorker || json.length < this.threadingThreshold) {
            return JSON.parse(json);
        }
        const result = await this.run('parse', json);
        return result;
    }
    async parseBuffer(buffer) {
        if (isServiceWorker || buffer.byteLength <= this.threadingThreshold) {
            const text = new TextDecoder().decode(buffer);
            return JSON.parse(text);
        }
        const result = await this.runWithTransfer('parse', buffer, [buffer]);
        return result;
    }
    async stringify(data) {
        const result = JSON.stringify(data);
        if (isServiceWorker || result.length < this.threadingThreshold) {
            return result;
        }
        return (await this.run('stringify', data));
    }
}
const GLOBAL_KEY = Symbol.for('opnet.jsonThreader');
const globalObj = (typeof globalThis !== 'undefined' ? globalThis : global);
if (!globalObj[GLOBAL_KEY]) {
    globalObj[GLOBAL_KEY] = new JsonThreader();
}
export const jsonThreader = globalObj[GLOBAL_KEY];
