// @btc-vision/logger/build/logger/shared/SharedLogger.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SharedLogger_pink, _SharedLogger_lightPink, _SharedLogger_purple, _SharedLogger_lightPurple, _SharedLogger_lighterPurple, _SharedLogger_green, _SharedLogger_lightGreen, _SharedLogger_moca, _SharedLogger_lightMoca, _SharedLogger_orange, _SharedLogger_lightOrange, _SharedLogger_red, _SharedLogger_lightRed, _SharedLogger_white, _SharedLogger_lightWhite, _SharedLogger_darkred, _SharedLogger_lightdarkred;
import { lightenColor } from './SharedFunctions.js';
import { Chalk } from 'chalk';
export class SharedLogger {
    constructor(chalk = new Chalk()) {
        this.chalk = chalk;
        this.moduleName = '';
        this.logColor = '#00bfff';
        this.enableLogs = true;
        this.hideLogs = false;
        _SharedLogger_pink.set(this, '#ff00ff');
        _SharedLogger_lightPink.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_pink, "f"), 75));
        _SharedLogger_purple.set(this, '#9400d3');
        _SharedLogger_lightPurple.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_purple, "f"), 15));
        _SharedLogger_lighterPurple.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_lightPurple, "f"), 15));
        _SharedLogger_green.set(this, '#7cfc00');
        _SharedLogger_lightGreen.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_green, "f"), 15));
        _SharedLogger_moca.set(this, '#ffdead');
        _SharedLogger_lightMoca.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_moca, "f"), 15));
        _SharedLogger_orange.set(this, '#ff8c00');
        _SharedLogger_lightOrange.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_orange, "f"), 15));
        _SharedLogger_red.set(this, '#ff4500');
        _SharedLogger_lightRed.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_red, "f"), 15));
        _SharedLogger_white.set(this, '#ffffff');
        _SharedLogger_lightWhite.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_white, "f"), 15));
        _SharedLogger_darkred.set(this, '#8b0000');
        _SharedLogger_lightdarkred.set(this, lightenColor(__classPrivateFieldGet(this, _SharedLogger_darkred, "f"), 15));
        this.prefix = '';
        this.moduleName = this.constructor.name;
    }
    setLogPrefix(prefix) {
        this.prefix = prefix;
    }
    getStartPrefix() {
        return this.prefix;
    }
    disable() {
        this.enableLogs = false;
    }
    enable() {
        this.enableLogs = true;
    }
    fancyLog(msg1, highlight1, msg2, highlight2, msg3) {
        if (!this.enableLogs)
            return;
        console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_pink, "f"))(`${this.getStartPrefix()}[${this.moduleName} INFO]: `) +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_white, "f"))(msg1) +
            ' ' +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightOrange, "f"))(highlight1) +
            ' ' +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_white, "f"))(msg2) +
            ' ' +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lighterPurple, "f"))(highlight2) +
            ' ' +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_white, "f"))(msg3));
    }
    log(...args) {
        if (!this.enableLogs)
            return;
        if (!this.hideLogs) {
            const light = lightenColor(this.logColor, 15);
            console.log(this.chalk.hex(this.logColor)(`${this.getStartPrefix()}[${this.moduleName} LOG]: `) + this.chalk.hex(light)(...args));
        }
    }
    lightOrangeLog(...args) {
        if (!this.enableLogs)
            return;
        if (!this.hideLogs) {
            console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightOrange, "f"))(`${this.getStartPrefix()}[${this.moduleName} LOG]: `) + this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_white, "f"))(...args));
        }
    }
    error(...args) {
        if (!this.enableLogs)
            return;
        console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_red, "f"))(`${this.getStartPrefix()}[${this.moduleName} ERROR]: `) +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightRed, "f"))(...args));
    }
    warn(...args) {
        if (!this.enableLogs)
            return;
        console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_orange, "f"))(`${this.getStartPrefix()}[${this.moduleName} WARN]: `) +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightOrange, "f"))(...args));
    }
    debug(...args) {
        if (!this.enableLogs)
            return;
        if (!this.hideLogs) {
            console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_moca, "f"))(`${this.getStartPrefix()}[${this.moduleName} DEBUG]: `) +
                this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightMoca, "f"))(...args));
        }
    }
    success(...args) {
        if (!this.enableLogs)
            return;
        if (!this.hideLogs) {
            console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_green, "f"))(`${this.getStartPrefix()}[${this.moduleName} SUCCESS]: `) + this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightGreen, "f"))(...args));
        }
    }
    fail(...args) {
        if (!this.enableLogs)
            return;
        if (!this.hideLogs) {
            console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_red, "f"))(`${this.getStartPrefix()}[${this.moduleName} FAIL]: `) +
                this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightRed, "f"))(...args));
        }
    }
    debugBright(...args) {
        if (!this.enableLogs)
            return;
        if (!this.hideLogs) {
            console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_purple, "f"))(`${this.getStartPrefix()}[${this.moduleName} DEBUG]: `) + this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightPurple, "f"))(...args));
        }
    }
    important(...args) {
        if (!this.enableLogs)
            return;
        console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_pink, "f"))(`${this.getStartPrefix()}[${this.moduleName} IMPORTANT]: `) +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightPink, "f"))(...args));
    }
    panic(...args) {
        if (!this.enableLogs)
            return;
        console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_darkred, "f"))(`${this.getStartPrefix()}[${this.moduleName} HELP PANIC]: `) + this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightdarkred, "f"))(...args));
    }
    info(...args) {
        if (!this.enableLogs)
            return;
        console.log(this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_pink, "f"))(`${this.getStartPrefix()}[${this.moduleName} INFO]: `) +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_white, "f"))(...args));
    }
    securityNotice(...args) {
        if (!this.enableLogs)
            return;
        console.log(this.chalk.hex('#22d8e6')(`${this.getStartPrefix()}[${this.moduleName} SECURITY NOTICE]: `) + this.chalk.hex('#22e3e6')(...args));
    }
    traceLog(...args) {
        if (!this.enableLogs)
            return;
        console.log(this.chalk.hex('#ffffff')(`${this.getStartPrefix()}[${this.moduleName} TRACE LOG]: `) +
            this.chalk.hex(__classPrivateFieldGet(this, _SharedLogger_lightWhite, "f"))(...args));
    }
}
_SharedLogger_pink = new WeakMap(), _SharedLogger_lightPink = new WeakMap(), _SharedLogger_purple = new WeakMap(), _SharedLogger_lightPurple = new WeakMap(), _SharedLogger_lighterPurple = new WeakMap(), _SharedLogger_green = new WeakMap(), _SharedLogger_lightGreen = new WeakMap(), _SharedLogger_moca = new WeakMap(), _SharedLogger_lightMoca = new WeakMap(), _SharedLogger_orange = new WeakMap(), _SharedLogger_lightOrange = new WeakMap(), _SharedLogger_red = new WeakMap(), _SharedLogger_lightRed = new WeakMap(), _SharedLogger_white = new WeakMap(), _SharedLogger_lightWhite = new WeakMap(), _SharedLogger_darkred = new WeakMap(), _SharedLogger_lightdarkred = new WeakMap();
