// @btc-vision/bitcoin/build/networks.js
export const bitcoin = {
    messagePrefix: '\x18Bitcoin Signed Message:\n',
    bech32: 'bc',
    bech32Opnet: 'op',
    bip32: {
        public: 0x0488b21e,
        private: 0x0488ade4,
    },
    pubKeyHash: 0x00,
    scriptHash: 0x05,
    wif: 0x80,
};
export const regtest = {
    messagePrefix: '\x18Bitcoin Signed Message:\n',
    bech32: 'bcrt',
    bech32Opnet: 'opr',
    bip32: {
        public: 0x043587cf,
        private: 0x04358394,
    },
    pubKeyHash: 0x6f,
    scriptHash: 0xc4,
    wif: 0xef,
};
export const testnet = {
    messagePrefix: '\x18Bitcoin Signed Message:\n',
    bech32: 'tb',
    bech32Opnet: 'opt',
    bip32: {
        public: 0x043587cf,
        private: 0x04358394,
    },
    pubKeyHash: 0x6f,
    scriptHash: 0xc4,
    wif: 0xef,
};
export const dogecoin = {
    messagePrefix: '\x19Dogecoin Signed Message:\n',
    bech32: '',
    bech32Opnet: '',
    bip32: {
        public: 0x02facafd,
        private: 0x02fac398,
    },
    pubKeyHash: 0x1e,
    scriptHash: 0x16,
    wif: 0x9e,
};
export const dogecoinTestnet = {
    messagePrefix: '\x19Dogecoin Signed Message:\n',
    bech32: '',
    bech32Opnet: '',
    bip32: {
        public: 0x0432a9a8,
        private: 0x0432a243,
    },
    pubKeyHash: 0x71,
    scriptHash: 0xc4,
    wif: 0xf1,
};
export const litecoin = {
    messagePrefix: '\x19Litecoin Signed Message:\n',
    bech32: 'ltc',
    bech32Opnet: 'opl',
    bip32: {
        public: 0x019da462,
        private: 0x019d9cfe,
    },
    pubKeyHash: 0x30,
    scriptHash: 0x32,
    wif: 0xb0,
};
export const litecoinTestnet = {
    messagePrefix: '\x19Litecoin Signed Message:\n',
    bech32: 'tltc',
    bech32Opnet: 'oplt',
    bip32: {
        public: 0x0436ef7d,
        private: 0x0436f6e1,
    },
    pubKeyHash: 0x6f,
    scriptHash: 0x3a,
    wif: 0xef,
};
export const bitcoinCash = {
    messagePrefix: '\x18Bitcoin Signed Message:\n',
    bech32: 'bitcoincash',
    bech32Opnet: 'opbch',
    bip32: {
        public: 0x0488b21e,
        private: 0x0488ade4,
    },
    pubKeyHash: 0x00,
    scriptHash: 0x05,
    wif: 0x80,
};
export const bitcoinCashTestnet = {
    messagePrefix: '\x18Bitcoin Signed Message:\n',
    bech32: 'bchtest',
    bech32Opnet: 'opbcht',
    bip32: {
        public: 0x043587cf,
        private: 0x04358394,
    },
    pubKeyHash: 0x6f,
    scriptHash: 0xc4,
    wif: 0xef,
};
export const dash = {
    messagePrefix: '\x19DarkCoin Signed Message:\n',
    bech32: '',
    bech32Opnet: '',
    bip32: {
        public: 0x02fe52cc,
        private: 0x02fe52f8,
    },
    pubKeyHash: 0x4c,
    scriptHash: 0x10,
    wif: 0xcc,
};
export const dashTestnet = {
    messagePrefix: '\x19DarkCoin Signed Message:\n',
    bech32: '',
    bech32Opnet: '',
    bip32: {
        public: 0x3a8061a0,
        private: 0x3a805837,
    },
    pubKeyHash: 0x8c,
    scriptHash: 0x13,
    wif: 0xef,
};
