// @btc-vision/bitcoin/build/script.js
import * as bip66 from './bip66.js';
import { opcodes, REVERSE_OPS } from './opcodes.js';
import * as pushdata from './push_data.js';
import * as scriptNumber from './script_number.js';
import * as scriptSignature from './script_signature.js';
import * as types from './types.js';
const { typeforce } = types;
const OP_INT_BASE = opcodes.OP_RESERVED;
export { opcodes };
function isOPInt(value) {
    return (types.Number(value) &&
        (value === opcodes.OP_0 ||
            (value >= opcodes.OP_1 && value <= opcodes.OP_16) ||
            value === opcodes.OP_1NEGATE));
}
function isPushOnlyChunk(value) {
    return types.Buffer(value) || isOPInt(value);
}
export function isPushOnly(value) {
    return types.Array(value) && value.every(isPushOnlyChunk);
}
export function countNonPushOnlyOPs(value) {
    return value.length - value.filter(isPushOnlyChunk).length;
}
function asMinimalOP(buffer) {
    if (buffer.length === 0)
        return opcodes.OP_0;
    if (buffer.length !== 1)
        return;
    if (buffer[0] >= 1 && buffer[0] <= 16)
        return OP_INT_BASE + buffer[0];
    if (buffer[0] === 0x81)
        return opcodes.OP_1NEGATE;
}
function chunksIsBuffer(buf) {
    return Buffer.isBuffer(buf);
}
function chunksIsArray(buf) {
    return types.Array(buf);
}
function singleChunkIsBuffer(buf) {
    return Buffer.isBuffer(buf);
}
export function compile(chunks) {
    if (chunksIsBuffer(chunks))
        return chunks;
    typeforce(types.Array, chunks);
    const bufferSize = chunks.reduce((accum, chunk) => {
        if (singleChunkIsBuffer(chunk)) {
            if (chunk.length === 1 && asMinimalOP(chunk) !== undefined) {
                return accum + 1;
            }
            return accum + pushdata.encodingLength(chunk.length) + chunk.length;
        }
        return accum + 1;
    }, 0.0);
    const buffer = Buffer.allocUnsafe(bufferSize);
    let offset = 0;
    chunks.forEach((chunk) => {
        if (singleChunkIsBuffer(chunk)) {
            const opcode = asMinimalOP(chunk);
            if (opcode !== undefined) {
                buffer.writeUInt8(opcode, offset);
                offset += 1;
                return;
            }
            offset += pushdata.encode(buffer, chunk.length, offset);
            chunk.copy(buffer, offset);
            offset += chunk.length;
        }
        else {
            buffer.writeUInt8(chunk, offset);
            offset += 1;
        }
    });
    if (offset !== buffer.length)
        throw new Error('Could not decode chunks');
    return buffer;
}
export function decompile(buffer) {
    if (chunksIsArray(buffer))
        return buffer;
    typeforce(types.Buffer, buffer);
    const chunks = [];
    let i = 0;
    while (i < buffer.length) {
        const opcode = buffer[i];
        if (opcode > opcodes.OP_0 && opcode <= opcodes.OP_PUSHDATA4) {
            const d = pushdata.decode(buffer, i);
            if (d === null)
                return null;
            i += d.size;
            if (i + d.number > buffer.length)
                return null;
            const data = buffer.subarray(i, i + d.number);
            i += d.number;
            const op = asMinimalOP(data);
            if (op !== undefined) {
                chunks.push(op);
            }
            else {
                chunks.push(data);
            }
        }
        else {
            chunks.push(opcode);
            i += 1;
        }
    }
    return chunks;
}
export function toASM(chunks) {
    if (chunksIsBuffer(chunks)) {
        chunks = decompile(chunks);
    }
    if (!chunks) {
        throw new Error('Could not convert invalid chunks to ASM');
    }
    return chunks
        .map((chunk) => {
        if (singleChunkIsBuffer(chunk)) {
            const op = asMinimalOP(chunk);
            if (op === undefined)
                return chunk.toString('hex');
            chunk = op;
        }
        return REVERSE_OPS[chunk];
    })
        .join(' ');
}
export function fromASM(asm) {
    typeforce(types.String, asm);
    return compile(asm.split(' ').map((chunkStr) => {
        if (opcodes[chunkStr] !== undefined) {
            return opcodes[chunkStr];
        }
        typeforce(types.Hex, chunkStr);
        return Buffer.from(chunkStr, 'hex');
    }));
}
export function toStack(chunks) {
    chunks = decompile(chunks);
    typeforce(isPushOnly, chunks);
    return chunks.map((op) => {
        if (singleChunkIsBuffer(op))
            return op;
        if (op === opcodes.OP_0)
            return Buffer.allocUnsafe(0);
        return scriptNumber.encode(op - OP_INT_BASE);
    });
}
export function isCanonicalPubKey(buffer) {
    return types.isPoint(buffer);
}
import { isDefinedHashType } from './script_signature.js';
export function isCanonicalScriptSignature(buffer) {
    if (!Buffer.isBuffer(buffer))
        return false;
    if (!isDefinedHashType(buffer[buffer.length - 1]))
        return false;
    return bip66.check(buffer.subarray(0, -1));
}
export const number = scriptNumber;
export const signature = scriptSignature;
