// @btc-vision/bitcoin/build/types.js
import { Buffer as NBuffer } from 'buffer';
import typeforce from 'typeforce';
export { typeforce };
const ZERO32 = NBuffer.alloc(32, 0);
const EC_P = NBuffer.from('fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f', 'hex');
export function stacksEqual(a, b) {
    if (a.length !== b.length)
        return false;
    return a.every((x, i) => {
        return x.equals(b[i]);
    });
}
export function isPoint(p) {
    if (!NBuffer.isBuffer(p))
        return false;
    if (p.length < 33)
        return false;
    const t = p[0];
    const x = p.subarray(1, 33);
    if (x.compare(ZERO32) === 0)
        return false;
    if (x.compare(EC_P) >= 0)
        return false;
    if ((t === 0x02 || t === 0x03) && p.length === 33) {
        return true;
    }
    if (p.length !== 65)
        return false;
    const y = p.subarray(33);
    if (y.compare(ZERO32) === 0)
        return false;
    if (y.compare(EC_P) >= 0)
        return false;
    return t === 0x04 || t === 0x06 || t === 0x07;
}
const SATOSHI_MAX = 21 * 1e14;
export function Satoshi(value) {
    return typeforce.UInt53(value) && value <= SATOSHI_MAX;
}
export const TAPLEAF_VERSION_MASK = 0xfe;
export function isTapleaf(o) {
    if (!o || typeof o !== 'object' || !('output' in o))
        return false;
    const obj = o;
    if (!NBuffer.isBuffer(obj.output))
        return false;
    if (obj.version !== undefined)
        return (obj.version & TAPLEAF_VERSION_MASK) === obj.version;
    return true;
}
export function isTaptree(scriptTree) {
    if (!globalThis.Array.isArray(scriptTree))
        return isTapleaf(scriptTree);
    if (scriptTree.length !== 2)
        return false;
    return scriptTree.every((t) => isTaptree(t));
}
export const Buffer256bit = typeforce.BufferN(32);
export const Hash160bit = typeforce.BufferN(20);
export const Hash256bit = typeforce.BufferN(32);
export const Number = typeforce.Number;
export const Array = typeforce.Array;
export const Boolean = typeforce.Boolean;
export const String = typeforce.String;
export const Buffer = typeforce.Buffer;
export const Hex = typeforce.Hex;
export const maybe = typeforce.maybe;
export const tuple = typeforce.tuple;
export const UInt8 = typeforce.UInt8;
export const UInt32 = typeforce.UInt32;
export const Function = typeforce.Function;
export const BufferN = typeforce.BufferN;
export const Null = typeforce.Null;
export const oneOf = typeforce.oneOf;
