// opnet/build/providers/interfaces/JSONRpcMethods.js
export var JSONRpcMethods;
(function (JSONRpcMethods) {
    JSONRpcMethods["BLOCK_BY_NUMBER"] = "btc_blockNumber";
    JSONRpcMethods["CHAIN_ID"] = "btc_chainId";
    JSONRpcMethods["REORG"] = "btc_reorg";
    JSONRpcMethods["GET_BLOCK_BY_HASH"] = "btc_getBlockByHash";
    JSONRpcMethods["GET_BLOCK_BY_CHECKSUM"] = "btc_getBlockByChecksum";
    JSONRpcMethods["GET_BLOCK_BY_NUMBER"] = "btc_getBlockByNumber";
    JSONRpcMethods["GAS"] = "btc_gas";
    JSONRpcMethods["GET_TRANSACTION_BY_HASH"] = "btc_getTransactionByHash";
    JSONRpcMethods["BROADCAST_TRANSACTION"] = "btc_sendRawTransaction";
    JSONRpcMethods["TRANSACTION_PREIMAGE"] = "btc_preimage";
    JSONRpcMethods["PUBLIC_KEY_INFO"] = "btc_publicKeyInfo";
    JSONRpcMethods["GET_UTXOS"] = "btc_getUTXOs";
    JSONRpcMethods["GET_BALANCE"] = "btc_getBalance";
    JSONRpcMethods["BLOCK_WITNESS"] = "btc_blockWitness";
    JSONRpcMethods["GET_TRANSACTION_RECEIPT"] = "btc_getTransactionReceipt";
    JSONRpcMethods["GET_CODE"] = "btc_getCode";
    JSONRpcMethods["GET_STORAGE_AT"] = "btc_getStorageAt";
    JSONRpcMethods["LATEST_EPOCH"] = "btc_latestEpoch";
    JSONRpcMethods["GET_EPOCH_BY_NUMBER"] = "btc_getEpochByNumber";
    JSONRpcMethods["GET_EPOCH_BY_HASH"] = "btc_getEpochByHash";
    JSONRpcMethods["GET_EPOCH_TEMPLATE"] = "btc_getEpochTemplate";
    JSONRpcMethods["SUBMIT_EPOCH"] = "btc_submitEpoch";
    JSONRpcMethods["CALL"] = "btc_call";
})(JSONRpcMethods || (JSONRpcMethods = {}));
