// @btc-vision/transaction/build/deterministic/AddressSet.js
export class AddressSet {
    constructor(keys = []) {
        this.items = new Set();
        this.keys = [];
        for (const key of keys) {
            this.add(key);
        }
    }
    get size() {
        return this.keys.length;
    }
    add(address) {
        const addressBigInt = address.toBigInt();
        if (!this.items.has(addressBigInt)) {
            this.items.add(addressBigInt);
            this.keys.push(address);
        }
    }
    has(address) {
        return this.items.has(address.toBigInt());
    }
    remove(address) {
        const addressBigInt = address.toBigInt();
        if (this.items.delete(addressBigInt)) {
            this.keys = this.keys.filter((k) => k.toBigInt() !== addressBigInt);
        }
    }
    clone() {
        return new AddressSet(this.keys);
    }
    clear() {
        this.items.clear();
        this.keys = [];
    }
    combine(set) {
        const clone = this.clone();
        for (const key of set.keys) {
            clone.add(key);
        }
        return clone;
    }
    *[Symbol.iterator]() {
        yield* this.keys;
    }
}
