// @btc-vision/transaction/build/deterministic/DeterministicMap.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _DeterministicMap_keys;
import { FastMap } from './FastMap.js';
export class DeterministicMap {
    constructor(compareFn) {
        this.compareFn = compareFn;
        _DeterministicMap_keys.set(this, void 0);
        this.map = new FastMap();
        __classPrivateFieldSet(this, _DeterministicMap_keys, [], "f");
    }
    get size() {
        return this.map.size;
    }
    static fromMap(map, compareFn) {
        const deterministicMap = new DeterministicMap(compareFn);
        for (const [key, value] of map) {
            deterministicMap.set(key, value);
        }
        return deterministicMap;
    }
    set(key, value) {
        if (!this.map.has(key)) {
            let left = 0, right = __classPrivateFieldGet(this, _DeterministicMap_keys, "f").length;
            while (left < right) {
                const mid = Math.floor((left + right) / 2);
                if (this.compareFn(__classPrivateFieldGet(this, _DeterministicMap_keys, "f")[mid], key) < 0) {
                    left = mid + 1;
                }
                else {
                    right = mid;
                }
            }
            __classPrivateFieldGet(this, _DeterministicMap_keys, "f").splice(left, 0, key);
        }
        this.map.set(key, value);
    }
    get(key) {
        return this.map.get(key);
    }
    *entries() {
        for (const key of __classPrivateFieldGet(this, _DeterministicMap_keys, "f")) {
            yield [key, this.map.get(key)];
        }
    }
    *keys() {
        yield* __classPrivateFieldGet(this, _DeterministicMap_keys, "f");
    }
    *values() {
        for (const key of __classPrivateFieldGet(this, _DeterministicMap_keys, "f")) {
            const value = this.map.get(key);
            if (value === undefined && !this.map.has(key)) {
                throw new Error('Value not found');
            }
            yield value;
        }
    }
    has(key) {
        return this.map.has(key);
    }
    delete(key) {
        if (this.map.has(key)) {
            this.map.delete(key);
            let left = 0, right = __classPrivateFieldGet(this, _DeterministicMap_keys, "f").length - 1;
            while (left <= right) {
                const mid = Math.floor((left + right) / 2);
                const cmp = this.compareFn(__classPrivateFieldGet(this, _DeterministicMap_keys, "f")[mid], key);
                if (cmp === 0) {
                    __classPrivateFieldGet(this, _DeterministicMap_keys, "f").splice(mid, 1);
                    return true;
                }
                else if (cmp < 0) {
                    left = mid + 1;
                }
                else {
                    right = mid - 1;
                }
            }
        }
        return false;
    }
    clear() {
        this.map.clear();
        __classPrivateFieldSet(this, _DeterministicMap_keys, [], "f");
    }
    forEach(callback) {
        for (const key of __classPrivateFieldGet(this, _DeterministicMap_keys, "f")) {
            const value = this.map.get(key);
            callback(value, key, this);
        }
    }
    *[(_DeterministicMap_keys = new WeakMap(), Symbol.iterator)]() {
        for (const key of __classPrivateFieldGet(this, _DeterministicMap_keys, "f")) {
            yield [key, this.map.get(key)];
        }
    }
}
