// @btc-vision/bitcoin/browser/chunks/psbt-BCNk7JUx.js
import { B as l, h as et } from "./crypto-BhCpKpek.js";
import { p as we, a as Ee, b as Xe, c as be, d as Te, e as On, f as Pn, g as De, t as Tr, h as qe, M as Nn, L as _r, r as mr } from "./payments-B1wlSccx.js";
import { o as ve, t as Ir, c as vr, d as ue, i as Un, s as de, a as kr } from "./script-DyPItFEl.js";
import { t as Ar, a as xr, H as Cr, U as Or, i as Pr, b as Hn, c as At, r as Bn } from "./utils-DNZi-T5W.js";
import { T as v } from "./transaction-C_UbhMGn.js";
const Ve = {
  /**
   * The message prefix used for signing Bitcoin messages.
   */
  messagePrefix: `Bitcoin Signed Message:
`,
  /**
   * The Bech32 prefix used for Bitcoin addresses.
   */
  bech32: "bc",
  bech32Opnet: "op",
  /**
   * The BIP32 key prefixes for Bitcoin.
   */
  bip32: {
    /**
     * The public key prefix for BIP32 extended public keys.
     */
    public: 76067358,
    /**
     * The private key prefix for BIP32 extended private keys.
     */
    private: 76066276
  },
  /**
   * The prefix for Bitcoin public key hashes.
   */
  pubKeyHash: 0,
  /**
   * The prefix for Bitcoin script hashes.
   */
  scriptHash: 5,
  /**
   * The prefix for Bitcoin Wallet Import Format (WIF) private keys.
   */
  wif: 128
}, Nr = {
  messagePrefix: `Bitcoin Signed Message:
`,
  bech32: "bcrt",
  bech32Opnet: "opr",
  bip32: {
    public: 70617039,
    private: 70615956
  },
  pubKeyHash: 111,
  scriptHash: 196,
  wif: 239
}, Ur = {
  messagePrefix: `Bitcoin Signed Message:
`,
  bech32: "tb",
  bech32Opnet: "opt",
  bip32: {
    public: 70617039,
    private: 70615956
  },
  pubKeyHash: 111,
  scriptHash: 196,
  wif: 239
}, Hr = {
  messagePrefix: `Dogecoin Signed Message:
`,
  bech32: "",
  // Dogecoin does not currently use Bech32
  bech32Opnet: "",
  bip32: {
    public: 49990397,
    private: 49988504
  },
  pubKeyHash: 30,
  scriptHash: 22,
  wif: 158
}, Br = {
  messagePrefix: `Dogecoin Signed Message:
`,
  bech32: "",
  // Dogecoin testnet does not currently use Bech32
  bech32Opnet: "",
  bip32: {
    public: 70429096,
    private: 70427203
  },
  pubKeyHash: 113,
  scriptHash: 196,
  wif: 241
}, Rr = {
  messagePrefix: `Litecoin Signed Message:
`,
  bech32: "ltc",
  bech32Opnet: "opl",
  bip32: {
    public: 27108450,
    private: 27106558
  },
  pubKeyHash: 48,
  scriptHash: 50,
  wif: 176
}, Fr = {
  messagePrefix: `Litecoin Signed Message:
`,
  bech32: "tltc",
  bech32Opnet: "oplt",
  bip32: {
    public: 70709117,
    private: 70711009
  },
  pubKeyHash: 111,
  scriptHash: 58,
  wif: 239
}, Kr = {
  messagePrefix: `Bitcoin Signed Message:
`,
  // Cashaddr prefix differs from bech32 for general usage, but we can set it similarly.
  // Actual cashaddr prefix is "bitcoincash", but this field is for bech32 which BCH doesn't fully use for segwit (it doesn't have segwit).
  bech32: "bitcoincash",
  bech32Opnet: "opbch",
  bip32: {
    public: 76067358,
    private: 76066276
  },
  pubKeyHash: 0,
  scriptHash: 5,
  wif: 128
}, Lr = {
  messagePrefix: `Bitcoin Signed Message:
`,
  bech32: "bchtest",
  bech32Opnet: "opbcht",
  bip32: {
    public: 70617039,
    private: 70615956
  },
  pubKeyHash: 111,
  scriptHash: 196,
  wif: 239
}, Mr = {
  // Historically Dash used DarkCoin message prefix, and most implementations use this:
  // As of Dash Core 0.17, this has not changed in code.
  messagePrefix: `DarkCoin Signed Message:
`,
  bech32: "",
  // Dash does not use Bech32
  bech32Opnet: "",
  bip32: {
    public: 50221772,
    private: 50221816
  },
  pubKeyHash: 76,
  scriptHash: 16,
  wif: 204
}, Wr = {
  messagePrefix: `DarkCoin Signed Message:
`,
  bech32: "",
  // Dash testnet does not use Bech32
  bech32Opnet: "",
  bip32: {
    public: 981492128,
    private: 981489719
  },
  pubKeyHash: 140,
  scriptHash: 19,
  wif: 239
}, Wo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  bitcoin: Ve,
  bitcoinCash: Kr,
  bitcoinCashTestnet: Lr,
  dash: Mr,
  dashTestnet: Wr,
  dogecoin: Hr,
  dogecoinTestnet: Br,
  litecoin: Rr,
  litecoinTestnet: Fr,
  regtest: Nr,
  testnet: Ur
}, Symbol.toStringTag, { value: "Module" }));
var ne = {}, xt;
function Gr() {
  if (xt) return ne;
  xt = 1, Object.defineProperty(ne, "__esModule", { value: !0 }), ne.bech32m = ne.bech32 = void 0;
  const t = "qpzry9x8gf2tvdw0s3jn54khce6mua7l", e = {};
  for (let u = 0; u < t.length; u++) {
    const f = t.charAt(u);
    e[f] = u;
  }
  function n(u) {
    const f = u >> 25;
    return (u & 33554431) << 5 ^ -(f >> 0 & 1) & 996825010 ^ -(f >> 1 & 1) & 642813549 ^ -(f >> 2 & 1) & 513874426 ^ -(f >> 3 & 1) & 1027748829 ^ -(f >> 4 & 1) & 705979059;
  }
  function r(u) {
    let f = 1;
    for (let h = 0; h < u.length; ++h) {
      const g = u.charCodeAt(h);
      if (g < 33 || g > 126)
        return "Invalid prefix (" + u + ")";
      f = n(f) ^ g >> 5;
    }
    f = n(f);
    for (let h = 0; h < u.length; ++h) {
      const g = u.charCodeAt(h);
      f = n(f) ^ g & 31;
    }
    return f;
  }
  function c(u, f, h, g) {
    let y = 0, p = 0;
    const d = (1 << h) - 1, E = [];
    for (let b = 0; b < u.length; ++b)
      for (y = y << f | u[b], p += f; p >= h; )
        p -= h, E.push(y >> p & d);
    if (g)
      p > 0 && E.push(y << h - p & d);
    else {
      if (p >= f)
        return "Excess padding";
      if (y << h - p & d)
        return "Non-zero padding";
    }
    return E;
  }
  function s(u) {
    return c(u, 8, 5, !0);
  }
  function o(u) {
    const f = c(u, 5, 8, !1);
    if (Array.isArray(f))
      return f;
  }
  function i(u) {
    const f = c(u, 5, 8, !1);
    if (Array.isArray(f))
      return f;
    throw new Error(f);
  }
  function a(u) {
    let f;
    u === "bech32" ? f = 1 : f = 734539939;
    function h(d, E, b) {
      if (b = b || 90, d.length + 7 + E.length > b)
        throw new TypeError("Exceeds length limit");
      d = d.toLowerCase();
      let S = r(d);
      if (typeof S == "string")
        throw new Error(S);
      let w = d + "1";
      for (let _ = 0; _ < E.length; ++_) {
        const I = E[_];
        if (I >> 5 !== 0)
          throw new Error("Non 5-bit word");
        S = n(S) ^ I, w += t.charAt(I);
      }
      for (let _ = 0; _ < 6; ++_)
        S = n(S);
      S ^= f;
      for (let _ = 0; _ < 6; ++_) {
        const I = S >> (5 - _) * 5 & 31;
        w += t.charAt(I);
      }
      return w;
    }
    function g(d, E) {
      if (E = E || 90, d.length < 8)
        return d + " too short";
      if (d.length > E)
        return "Exceeds length limit";
      const b = d.toLowerCase(), S = d.toUpperCase();
      if (d !== b && d !== S)
        return "Mixed-case string " + d;
      d = b;
      const w = d.lastIndexOf("1");
      if (w === -1)
        return "No separator character for " + d;
      if (w === 0)
        return "Missing prefix for " + d;
      const _ = d.slice(0, w), I = d.slice(w + 1);
      if (I.length < 6)
        return "Data too short";
      let P = r(_);
      if (typeof P == "string")
        return P;
      const m = [];
      for (let k = 0; k < I.length; ++k) {
        const U = I.charAt(k), B = e[U];
        if (B === void 0)
          return "Unknown character " + U;
        P = n(P) ^ B, !(k + 6 >= I.length) && m.push(B);
      }
      return P !== f ? "Invalid checksum for " + d : { prefix: _, words: m };
    }
    function y(d, E) {
      const b = g(d, E);
      if (typeof b == "object")
        return b;
    }
    function p(d, E) {
      const b = g(d, E);
      if (typeof b == "object")
        return b;
      throw new Error(b);
    }
    return {
      decodeUnsafe: y,
      decode: p,
      encode: h,
      toWords: s,
      fromWordsUnsafe: o,
      fromWords: i
    };
  }
  return ne.bech32 = a("bech32"), ne.bech32m = a("bech32m"), ne;
}
var K = Gr();
function Xr(t) {
  return t instanceof Uint8Array || ArrayBuffer.isView(t) && t.constructor.name === "Uint8Array";
}
function bt(t, ...e) {
  if (!Xr(t))
    throw new Error("Uint8Array expected");
  if (e.length > 0 && !e.includes(t.length))
    throw new Error("Uint8Array expected of length " + e + ", got length=" + t.length);
}
function Ct(t, e = !0) {
  if (t.destroyed)
    throw new Error("Hash instance has been destroyed");
  if (e && t.finished)
    throw new Error("Hash#digest() has already been called");
}
function Dr(t, e) {
  bt(t);
  const n = e.outputLen;
  if (t.length < n)
    throw new Error("digestInto() expects output buffer of length at least " + n);
}
function ft(...t) {
  for (let e = 0; e < t.length; e++)
    t[e].fill(0);
}
function tt(t) {
  return new DataView(t.buffer, t.byteOffset, t.byteLength);
}
function H(t, e) {
  return t << 32 - e | t >>> e;
}
function qr(t) {
  if (typeof t != "string")
    throw new Error("string expected");
  return new Uint8Array(new TextEncoder().encode(t));
}
function Rn(t) {
  return typeof t == "string" && (t = qr(t)), bt(t), t;
}
class Vr {
}
function jr(t) {
  const e = (r) => t().update(Rn(r)).digest(), n = t();
  return e.outputLen = n.outputLen, e.blockLen = n.blockLen, e.create = () => t(), e;
}
function zr(t, e, n, r) {
  if (typeof t.setBigUint64 == "function")
    return t.setBigUint64(e, n, r);
  const c = BigInt(32), s = BigInt(4294967295), o = Number(n >> c & s), i = Number(n & s), a = r ? 4 : 0, u = r ? 0 : 4;
  t.setUint32(e + a, o, r), t.setUint32(e + u, i, r);
}
function Yr(t, e, n) {
  return t & e ^ ~t & n;
}
function $r(t, e, n) {
  return t & e ^ t & n ^ e & n;
}
class Zr extends Vr {
  constructor(e, n, r, c) {
    super(), this.finished = !1, this.length = 0, this.pos = 0, this.destroyed = !1, this.blockLen = e, this.outputLen = n, this.padOffset = r, this.isLE = c, this.buffer = new Uint8Array(e), this.view = tt(this.buffer);
  }
  update(e) {
    Ct(this), e = Rn(e), bt(e);
    const { view: n, buffer: r, blockLen: c } = this, s = e.length;
    for (let o = 0; o < s; ) {
      const i = Math.min(c - this.pos, s - o);
      if (i === c) {
        const a = tt(e);
        for (; c <= s - o; o += c)
          this.process(a, o);
        continue;
      }
      r.set(e.subarray(o, o + i), this.pos), this.pos += i, o += i, this.pos === c && (this.process(n, 0), this.pos = 0);
    }
    return this.length += e.length, this.roundClean(), this;
  }
  digestInto(e) {
    Ct(this), Dr(e, this), this.finished = !0;
    const { buffer: n, view: r, blockLen: c, isLE: s } = this;
    let { pos: o } = this;
    n[o++] = 128, ft(this.buffer.subarray(o)), this.padOffset > c - o && (this.process(r, 0), o = 0);
    for (let h = o; h < c; h++)
      n[h] = 0;
    zr(r, c - 8, BigInt(this.length * 8), s), this.process(r, 0);
    const i = tt(e), a = this.outputLen;
    if (a % 4)
      throw new Error("_sha2: outputLen should be aligned to 32bit");
    const u = a / 4, f = this.get();
    if (u > f.length)
      throw new Error("_sha2: outputLen bigger than state");
    for (let h = 0; h < u; h++)
      i.setUint32(4 * h, f[h], s);
  }
  digest() {
    const { buffer: e, outputLen: n } = this;
    this.digestInto(e);
    const r = e.slice(0, n);
    return this.destroy(), r;
  }
  _cloneInto(e) {
    e || (e = new this.constructor()), e.set(...this.get());
    const { blockLen: n, buffer: r, length: c, finished: s, destroyed: o, pos: i } = this;
    return e.destroyed = o, e.finished = s, e.length = c, e.pos = i, c % n && e.buffer.set(r), e;
  }
  clone() {
    return this._cloneInto();
  }
}
const L = /* @__PURE__ */ Uint32Array.from([
  1779033703,
  3144134277,
  1013904242,
  2773480762,
  1359893119,
  2600822924,
  528734635,
  1541459225
]), Jr = /* @__PURE__ */ Uint32Array.from([
  1116352408,
  1899447441,
  3049323471,
  3921009573,
  961987163,
  1508970993,
  2453635748,
  2870763221,
  3624381080,
  310598401,
  607225278,
  1426881987,
  1925078388,
  2162078206,
  2614888103,
  3248222580,
  3835390401,
  4022224774,
  264347078,
  604807628,
  770255983,
  1249150122,
  1555081692,
  1996064986,
  2554220882,
  2821834349,
  2952996808,
  3210313671,
  3336571891,
  3584528711,
  113926993,
  338241895,
  666307205,
  773529912,
  1294757372,
  1396182291,
  1695183700,
  1986661051,
  2177026350,
  2456956037,
  2730485921,
  2820302411,
  3259730800,
  3345764771,
  3516065817,
  3600352804,
  4094571909,
  275423344,
  430227734,
  506948616,
  659060556,
  883997877,
  958139571,
  1322822218,
  1537002063,
  1747873779,
  1955562222,
  2024104815,
  2227730452,
  2361852424,
  2428436474,
  2756734187,
  3204031479,
  3329325298
]), M = /* @__PURE__ */ new Uint32Array(64);
class Qr extends Zr {
  constructor(e = 32) {
    super(64, e, 8, !1), this.A = L[0] | 0, this.B = L[1] | 0, this.C = L[2] | 0, this.D = L[3] | 0, this.E = L[4] | 0, this.F = L[5] | 0, this.G = L[6] | 0, this.H = L[7] | 0;
  }
  get() {
    const { A: e, B: n, C: r, D: c, E: s, F: o, G: i, H: a } = this;
    return [e, n, r, c, s, o, i, a];
  }
  // prettier-ignore
  set(e, n, r, c, s, o, i, a) {
    this.A = e | 0, this.B = n | 0, this.C = r | 0, this.D = c | 0, this.E = s | 0, this.F = o | 0, this.G = i | 0, this.H = a | 0;
  }
  process(e, n) {
    for (let h = 0; h < 16; h++, n += 4)
      M[h] = e.getUint32(n, !1);
    for (let h = 16; h < 64; h++) {
      const g = M[h - 15], y = M[h - 2], p = H(g, 7) ^ H(g, 18) ^ g >>> 3, d = H(y, 17) ^ H(y, 19) ^ y >>> 10;
      M[h] = d + M[h - 7] + p + M[h - 16] | 0;
    }
    let { A: r, B: c, C: s, D: o, E: i, F: a, G: u, H: f } = this;
    for (let h = 0; h < 64; h++) {
      const g = H(i, 6) ^ H(i, 11) ^ H(i, 25), y = f + g + Yr(i, a, u) + Jr[h] + M[h] | 0, d = (H(r, 2) ^ H(r, 13) ^ H(r, 22)) + $r(r, c, s) | 0;
      f = u, u = a, a = i, i = o + y | 0, o = s, s = c, c = r, r = y + d | 0;
    }
    r = r + this.A | 0, c = c + this.B | 0, s = s + this.C | 0, o = o + this.D | 0, i = i + this.E | 0, a = a + this.F | 0, u = u + this.G | 0, f = f + this.H | 0, this.set(r, c, s, o, i, a, u, f);
  }
  roundClean() {
    ft(M);
  }
  destroy() {
    this.set(0, 0, 0, 0, 0, 0, 0, 0), ft(this.buffer);
  }
}
const es = /* @__PURE__ */ jr(() => new Qr()), Ot = es;
function ts(t) {
  if (t.length >= 255)
    throw new TypeError("Alphabet too long");
  const e = new Uint8Array(256);
  for (let u = 0; u < e.length; u++)
    e[u] = 255;
  for (let u = 0; u < t.length; u++) {
    const f = t.charAt(u), h = f.charCodeAt(0);
    if (e[h] !== 255)
      throw new TypeError(f + " is ambiguous");
    e[h] = u;
  }
  const n = t.length, r = t.charAt(0), c = Math.log(n) / Math.log(256), s = Math.log(256) / Math.log(n);
  function o(u) {
    if (u instanceof Uint8Array || (ArrayBuffer.isView(u) ? u = new Uint8Array(u.buffer, u.byteOffset, u.byteLength) : Array.isArray(u) && (u = Uint8Array.from(u))), !(u instanceof Uint8Array))
      throw new TypeError("Expected Uint8Array");
    if (u.length === 0)
      return "";
    let f = 0, h = 0, g = 0;
    const y = u.length;
    for (; g !== y && u[g] === 0; )
      g++, f++;
    const p = (y - g) * s + 1 >>> 0, d = new Uint8Array(p);
    for (; g !== y; ) {
      let S = u[g], w = 0;
      for (let _ = p - 1; (S !== 0 || w < h) && _ !== -1; _--, w++)
        S += 256 * d[_] >>> 0, d[_] = S % n >>> 0, S = S / n >>> 0;
      if (S !== 0)
        throw new Error("Non-zero carry");
      h = w, g++;
    }
    let E = p - h;
    for (; E !== p && d[E] === 0; )
      E++;
    let b = r.repeat(f);
    for (; E < p; ++E)
      b += t.charAt(d[E]);
    return b;
  }
  function i(u) {
    if (typeof u != "string")
      throw new TypeError("Expected String");
    if (u.length === 0)
      return new Uint8Array();
    let f = 0, h = 0, g = 0;
    for (; u[f] === r; )
      h++, f++;
    const y = (u.length - f) * c + 1 >>> 0, p = new Uint8Array(y);
    for (; f < u.length; ) {
      const S = u.charCodeAt(f);
      if (S > 255)
        return;
      let w = e[S];
      if (w === 255)
        return;
      let _ = 0;
      for (let I = y - 1; (w !== 0 || _ < g) && I !== -1; I--, _++)
        w += n * p[I] >>> 0, p[I] = w % 256 >>> 0, w = w / 256 >>> 0;
      if (w !== 0)
        throw new Error("Non-zero carry");
      g = _, f++;
    }
    let d = y - g;
    for (; d !== y && p[d] === 0; )
      d++;
    const E = new Uint8Array(h + (y - d));
    let b = h;
    for (; d !== y; )
      E[b++] = p[d++];
    return E;
  }
  function a(u) {
    const f = i(u);
    if (f)
      return f;
    throw new Error("Non-base" + n + " character");
  }
  return {
    encode: o,
    decodeUnsafe: i,
    decode: a
  };
}
var ns = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
const nt = ts(ns);
function rs(t) {
  function e(s) {
    var o = Uint8Array.from(s), i = t(o), a = o.length + 4, u = new Uint8Array(a);
    return u.set(o, 0), u.set(i.subarray(0, 4), o.length), nt.encode(u);
  }
  function n(s) {
    var o = s.slice(0, -4), i = s.slice(-4), a = t(o);
    if (!(i[0] ^ a[0] | i[1] ^ a[1] | i[2] ^ a[2] | i[3] ^ a[3]))
      return o;
  }
  function r(s) {
    var o = nt.decodeUnsafe(s);
    if (o != null)
      return n(o);
  }
  function c(s) {
    var o = nt.decode(s), i = n(o);
    if (i == null)
      throw new Error("Invalid checksum");
    return i;
  }
  return {
    encode: e,
    decode: c,
    decodeUnsafe: r
  };
}
function ss(t) {
  return Ot(Ot(t));
}
const Fn = rs(ss);
function Kn(t) {
  let e, n;
  try {
    e = K.bech32.decode(t);
  } catch {
  }
  if (e) {
    if (n = e.words[0], n !== 0) throw new TypeError(t + " uses wrong encoding");
  } else if (e = K.bech32m.decode(t), n = e.words[0], n === 0) throw new TypeError(t + " uses wrong encoding");
  const r = K.bech32.fromWords(e.words.slice(1));
  return {
    version: n,
    prefix: e.prefix,
    data: l.from(r)
  };
}
const Ln = {
  p: 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2fn,
  n: 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141n,
  h: 1n,
  a: 0n,
  b: 7n,
  Gx: 0x79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798n,
  Gy: 0x483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8n
}, { p: se, n: os, Gx: is, Gy: cs, b: Mn } = Ln, ht = 32, Tt = 64, as = {
  publicKey: ht + 1,
  publicKeyUncompressed: Tt + 1
}, us = (...t) => {
  "captureStackTrace" in Error && typeof Error.captureStackTrace == "function" && Error.captureStackTrace(...t);
}, C = (t = "") => {
  const e = new Error(t);
  throw us(e, C), e;
}, fs = (t) => typeof t == "bigint", hs = (t) => typeof t == "string", ps = (t) => t instanceof Uint8Array || ArrayBuffer.isView(t) && t.constructor.name === "Uint8Array", _t = (t, e, n = "") => {
  const r = ps(t), c = t?.length, s = e !== void 0;
  if (!r || s && c !== e) {
    const o = n && `"${n}" `, i = s ? ` of length ${e}` : "", a = r ? `length=${c}` : `type=${typeof t}`;
    C(o + "expected Uint8Array" + i + ", got " + a);
  }
  return t;
}, Wn = (t) => new Uint8Array(t), Gn = (t, e) => t.toString(16).padStart(e, "0"), Xn = (t) => Array.from(_t(t)).map((e) => Gn(e, 2)).join(""), R = { _0: 48, _9: 57, A: 65, F: 70, a: 97, f: 102 }, Pt = (t) => {
  if (t >= R._0 && t <= R._9)
    return t - R._0;
  if (t >= R.A && t <= R.F)
    return t - (R.A - 10);
  if (t >= R.a && t <= R.f)
    return t - (R.a - 10);
}, Dn = (t) => {
  const e = "hex invalid";
  if (!hs(t))
    return C(e);
  const n = t.length, r = n / 2;
  if (n % 2)
    return C(e);
  const c = Wn(r);
  for (let s = 0, o = 0; s < r; s++, o += 2) {
    const i = Pt(t.charCodeAt(o)), a = Pt(t.charCodeAt(o + 1));
    if (i === void 0 || a === void 0)
      return C(e);
    c[s] = i * 16 + a;
  }
  return c;
}, Nt = (...t) => {
  const e = Wn(t.reduce((r, c) => r + _t(c).length, 0));
  let n = 0;
  return t.forEach((r) => {
    e.set(r, n), n += r.length;
  }), e;
}, Le = BigInt, je = (t, e, n, r = "bad number: out of range") => fs(t) && e <= t && t < n ? t : C(r), T = (t, e = se) => {
  const n = t % e;
  return n >= 0n ? n : e + n;
}, ls = (t, e) => {
  (t === 0n || e <= 0n) && C("no inverse n=" + t + " mod=" + e);
  let n = T(t, e), r = e, c = 0n, s = 1n;
  for (; n !== 0n; ) {
    const o = r / n, i = r % n, a = c - s * o;
    r = n, n = i, c = s, s = a;
  }
  return r === 1n ? T(c, e) : C("no inverse");
}, rt = (t) => t instanceof O ? t : C("Point expected"), qn = (t) => T(T(t * t) * t + Mn), Ut = (t) => je(t, 0n, se), Ke = (t) => je(t, 1n, se), ds = (t) => je(t, 1n, os), pt = (t) => (t & 1n) === 0n, Vn = (t) => Uint8Array.of(t), gs = (t) => Vn(pt(t) ? 2 : 3), ys = (t) => {
  const e = qn(Ke(t));
  let n = 1n;
  for (let r = e, c = (se + 1n) / 4n; c > 0n; c >>= 1n)
    c & 1n && (n = n * r % se), r = r * r % se;
  return T(n * n) === e ? n : C("sqrt invalid");
};
class O {
  static BASE;
  static ZERO;
  X;
  Y;
  Z;
  constructor(e, n, r) {
    this.X = Ut(e), this.Y = Ke(n), this.Z = Ut(r), Object.freeze(this);
  }
  static CURVE() {
    return Ln;
  }
  /** Create 3d xyz point from 2d xy. (0, 0) => (0, 1, 0), not (0, 0, 1) */
  static fromAffine(e) {
    const { x: n, y: r } = e;
    return n === 0n && r === 0n ? re : new O(n, r, 1n);
  }
  /** Convert Uint8Array or hex string to Point. */
  static fromBytes(e) {
    _t(e);
    const { publicKey: n, publicKeyUncompressed: r } = as;
    let c;
    const s = e.length, o = e[0], i = e.subarray(1), a = Ht(i, 0, ht);
    if (s === n && (o === 2 || o === 3)) {
      let u = ys(a);
      const f = pt(u);
      pt(Le(o)) !== f && (u = T(-u)), c = new O(a, u, 1n);
    }
    return s === r && o === 4 && (c = new O(a, Ht(i, ht, Tt), 1n)), c ? c.assertValidity() : C("bad point: not on curve");
  }
  static fromHex(e) {
    return O.fromBytes(Dn(e));
  }
  get x() {
    return this.toAffine().x;
  }
  get y() {
    return this.toAffine().y;
  }
  /** Equality check: compare points P&Q. */
  equals(e) {
    const { X: n, Y: r, Z: c } = this, { X: s, Y: o, Z: i } = rt(e), a = T(n * i), u = T(s * c), f = T(r * i), h = T(o * c);
    return a === u && f === h;
  }
  is0() {
    return this.equals(re);
  }
  /** Flip point over y coordinate. */
  negate() {
    return new O(this.X, T(-this.Y), this.Z);
  }
  /** Point doubling: P+P, complete formula. */
  double() {
    return this.add(this);
  }
  /**
   * Point addition: P+Q, complete, exception-free formula
   * (Renes-Costello-Batina, algo 1 of [2015/1060](https://eprint.iacr.org/2015/1060)).
   * Cost: `12M + 0S + 3*a + 3*b3 + 23add`.
   */
  // prettier-ignore
  add(e) {
    const { X: n, Y: r, Z: c } = this, { X: s, Y: o, Z: i } = rt(e), a = 0n, u = Mn;
    let f = 0n, h = 0n, g = 0n;
    const y = T(u * 3n);
    let p = T(n * s), d = T(r * o), E = T(c * i), b = T(n + r), S = T(s + o);
    b = T(b * S), S = T(p + d), b = T(b - S), S = T(n + c);
    let w = T(s + i);
    return S = T(S * w), w = T(p + E), S = T(S - w), w = T(r + c), f = T(o + i), w = T(w * f), f = T(d + E), w = T(w - f), g = T(a * S), f = T(y * E), g = T(f + g), f = T(d - g), g = T(d + g), h = T(f * g), d = T(p + p), d = T(d + p), E = T(a * E), S = T(y * S), d = T(d + E), E = T(p - E), E = T(a * E), S = T(S + E), p = T(d * S), h = T(h + p), p = T(w * S), f = T(b * f), f = T(f - p), p = T(b * d), g = T(w * g), g = T(g + p), new O(f, h, g);
  }
  subtract(e) {
    return this.add(rt(e).negate());
  }
  /**
   * Point-by-scalar multiplication. Scalar must be in range 1 <= n < CURVE.n.
   * Uses {@link wNAF} for base point.
   * Uses fake point to mitigate side-channel leakage.
   * @param n scalar by which point is multiplied
   * @param safe safe mode guards against timing attacks; unsafe mode is faster
   */
  multiply(e, n = !0) {
    if (!n && e === 0n)
      return re;
    if (ds(e), e === 1n)
      return this;
    if (this.equals(Se))
      return Ts(e).p;
    let r = re, c = Se;
    for (let s = this; e > 0n; s = s.double(), e >>= 1n)
      e & 1n ? r = r.add(s) : n && (c = c.add(s));
    return r;
  }
  multiplyUnsafe(e) {
    return this.multiply(e, !1);
  }
  /** Convert point to 2d xy affine point. (X, Y, Z) ∋ (x=X/Z, y=Y/Z) */
  toAffine() {
    const { X: e, Y: n, Z: r } = this;
    if (this.equals(re))
      return { x: 0n, y: 0n };
    if (r === 1n)
      return { x: e, y: n };
    const c = ls(r, se);
    return T(r * c) !== 1n && C("inverse invalid"), { x: T(e * c), y: T(n * c) };
  }
  /** Checks if the point is valid and on-curve. */
  assertValidity() {
    const { x: e, y: n } = this.toAffine();
    return Ke(e), Ke(n), T(n * n) === qn(e) ? this : C("bad point: not on curve");
  }
  /** Converts point to 33/65-byte Uint8Array. */
  toBytes(e = !0) {
    const { x: n, y: r } = this.assertValidity().toAffine(), c = Bt(n);
    return e ? Nt(gs(r), c) : Nt(Vn(4), c, Bt(r));
  }
  toHex(e) {
    return Xn(this.toBytes(e));
  }
}
const Se = new O(is, cs, 1n), re = new O(0n, 1n, 0n);
O.BASE = Se;
O.ZERO = re;
const Ss = (t) => Le("0x" + (Xn(t) || "0")), Ht = (t, e, n) => Ss(t.subarray(e, n)), ws = 2n ** 256n, Bt = (t) => Dn(Gn(je(t, 0n, ws), Tt)), Me = 8, Es = 256, jn = Math.ceil(Es / Me) + 1, lt = 2 ** (Me - 1), bs = () => {
  const t = [];
  let e = Se, n = e;
  for (let r = 0; r < jn; r++) {
    n = e, t.push(n);
    for (let c = 1; c < lt; c++)
      n = n.add(e), t.push(n);
    e = n.double();
  }
  return t;
};
let Rt;
const Ft = (t, e) => {
  const n = e.negate();
  return t ? n : e;
}, Ts = (t) => {
  const e = Rt || (Rt = bs());
  let n = re, r = Se;
  const c = 2 ** Me, s = c, o = Le(c - 1), i = Le(Me);
  for (let a = 0; a < jn; a++) {
    let u = Number(t & o);
    t >>= i, u > lt && (u -= s, t += 1n);
    const f = a * lt, h = f, g = f + Math.abs(u) - 1, y = a % 2 !== 0, p = u < 0;
    u === 0 ? r = r.add(Ft(y, e[h])) : n = n.add(Ft(p, e[g]));
  }
  return t !== 0n && C("invalid wnaf"), { p: n, f: r };
}, ge = (t) => {
  const e = t.length === 32 ? t : t.subarray(1, 33);
  return l.isBuffer(e) ? e : l.from(e);
};
function Kt(t) {
  let e = t.toString(16);
  return e = e.padStart(64, "0"), e.length > 64 && (e = e.slice(-64)), l.from(e, "hex");
}
function _s(t) {
  if (t.length === 32)
    return;
  if (![33, 65].includes(t.length)) {
    console.warn(
      `Unsupported key length=${t.length}. Must be 33 (compressed) or 65 (uncompressed).`
    );
    return;
  }
  let e;
  try {
    e = O.fromHex(l.from(t).toString("hex"));
  } catch {
    throw new Error("Invalid secp256k1 public key bytes. Cannot parse.");
  }
  const n = Kt(e.x), r = Kt(e.y), s = e.y % 2n === 0n ? 6 : 7, o = l.alloc(65);
  o[0] = s, n.copy(o, 1), r.copy(o, 33);
  const i = l.concat([l.from([4]), n, r]);
  return {
    hybrid: o,
    uncompressed: i
  };
}
function ke(t, e) {
  if (t.equals(e)) return !0;
  if (t.length === 65 && e.length === 65) {
    const n = l.from(t), r = l.from(e);
    return (n[0] === 6 || n[0] === 7) && (n[0] = 4), (r[0] === 6 || r[0] === 7) && (r[0] = 4), n.equals(r);
  }
  return !1;
}
const _e = 40, me = 2, Ie = 15, zn = 16, We = 16, ze = 2, Ye = 80, ms = "WARNING: Sending to a future segwit version address can lead to loss of funds. End users MUST be warned carefully in the GUI and asked if they wish to proceed with caution. Wallets should verify the segwit version from the output of fromBech32, then decide when it is safe to use which version of segwit.", Yn = (t) => {
  try {
    const e = l.from(t.subarray(2));
    if (e.length < me || e.length > _e)
      throw new TypeError("Invalid program length for segwit address");
    const n = t[0] - Ye;
    if (n < ze || n > Ie + 1)
      throw new TypeError("Invalid version for segwit address");
    if (n === 1) throw new TypeError("taproot");
    return !0;
  } catch {
  }
  return !1;
};
function $n(t, e) {
  if (!l.isBuffer(t)) throw new TypeError("output must be a Buffer");
  if (!e.bech32Opnet) throw new Error("Network does not support opnet");
  const n = t[0];
  let r = 1, c;
  if (t[1] < 76)
    c = t[1], r = 2;
  else if (t[1] === 76)
    c = t[2], r = 3;
  else
    throw new TypeError("Unsupported push opcode in script");
  const s = l.from(t.subarray(r, r + c));
  if (s.length < me || s.length > _e)
    throw new TypeError("Invalid program length for segwit address");
  const o = n === ve.OP_0 ? 0 : n >= ve.OP_1 && n <= ve.OP_16 ? n - (ve.OP_1 - 1) : -1;
  if (o < Ie || o > zn)
    throw new TypeError(`Invalid segwit version ${o}`);
  const i = [o, ...K.bech32m.toWords(s)];
  return K.bech32m.encode(e.bech32Opnet, i);
}
function Zn(t, e) {
  const n = l.from(t.subarray(2));
  if (n.length < me || n.length > _e)
    throw new TypeError("Invalid program length for segwit address");
  const r = t[0] - Ye;
  if (r < ze || r > Ie)
    throw new TypeError("Invalid version for segwit address");
  if (t[1] !== n.length)
    throw new TypeError(`Invalid script for segwit address ${t[1]} !== ${n.length}`);
  return Qn(n, r, e.bech32, e.bech32Opnet);
}
function Jn(t) {
  const e = l.from(Fn.decode(t));
  if (e.length < 21) throw new TypeError(t + " is too short");
  if (e.length > 21) throw new TypeError(t + " is too long");
  const n = e.readUInt8(0), r = l.from(e.subarray(1));
  return { version: n, hash: r };
}
function Is(t, e) {
  Ar(xr(Cr, Or), [t, e]);
  const n = l.allocUnsafe(21);
  return n.writeUInt8(e, 0), t.copy(n, 1), Fn.encode(n);
}
function Qn(t, e, n, r) {
  const c = K.bech32.toWords(t);
  return c.unshift(e), e === We && r ? K.bech32m.encode(r, c) : e === 0 ? K.bech32.encode(n, c) : K.bech32m.encode(n, c);
}
function er(t, e) {
  e = e || Ve;
  try {
    return we({ output: t, network: e }).address;
  } catch {
  }
  try {
    return Ee({ output: t, network: e }).address;
  } catch {
  }
  try {
    return Xe({ output: t, network: e }).address;
  } catch {
  }
  try {
    return be({ output: t, network: e }).address;
  } catch {
  }
  try {
    return Te({ output: t, network: e }).address;
  } catch {
  }
  try {
    return $n(t, e);
  } catch {
  }
  try {
    return Zn(t, e);
  } catch {
  }
  throw new Error(Ir(t) + " has no matching Address");
}
function tr(t, e) {
  e = e || Ve;
  let n, r;
  try {
    n = Jn(t);
  } catch {
  }
  if (n) {
    if (n.version === e.pubKeyHash)
      return we({ hash: n.hash }).output;
    if (n.version === e.scriptHash)
      return Ee({ hash: n.hash }).output;
  } else {
    try {
      r = Kn(t);
    } catch {
    }
    if (r) {
      if (r.prefix !== e.bech32 && e.bech32Opnet && r.prefix !== e.bech32Opnet)
        throw new Error(t + " has an invalid prefix");
      if (r.version === 0) {
        if (r.data.length === 20)
          return Xe({ hash: r.data }).output;
        if (r.data.length === 32)
          return be({ hash: r.data }).output;
      } else if (r.version === 1) {
        if (r.data.length === 32)
          return Te({ pubkey: r.data }).output;
      } else if (r.version === We) {
        if (!e.bech32Opnet) throw new Error(t + " has an invalid prefix");
        return On({
          program: r.data,
          network: e
        }).output;
      } else if (r.version >= ze && r.version <= Ie && r.data.length >= me && r.data.length <= _e)
        return r.version !== We && console.warn(ms), vr([
          r.version + Ye,
          r.data
        ]);
    }
  }
  throw new TypeError(t + " has no matching Script");
}
const Go = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  FUTURE_MAX_VERSION: zn,
  FUTURE_OPNET_VERSION: We,
  FUTURE_SEGWIT_MAX_SIZE: _e,
  FUTURE_SEGWIT_MAX_VERSION: Ie,
  FUTURE_SEGWIT_MIN_SIZE: me,
  FUTURE_SEGWIT_MIN_VERSION: ze,
  FUTURE_SEGWIT_VERSION_DIFF: Ye,
  _toFutureSegwitAddress: Zn,
  fromBase58Check: Jn,
  fromBech32: Kn,
  fromOutputScript: er,
  isUnknownSegwitVersion: Yn,
  toBase58Check: Is,
  toBech32: Qn,
  toFutureOPNetAddress: $n,
  toOutputScript: tr
}, Symbol.toStringTag, { value: "Module" }));
var Ae = {}, xe = {}, st = {}, oe = {}, ie = {}, ot = {}, Lt;
function A() {
  return Lt || (Lt = 1, (function(t) {
    Object.defineProperty(t, "__esModule", { value: !0 }), (function(e) {
      e[e.UNSIGNED_TX = 0] = "UNSIGNED_TX", e[e.GLOBAL_XPUB = 1] = "GLOBAL_XPUB";
    })(t.GlobalTypes || (t.GlobalTypes = {})), t.GLOBAL_TYPE_NAMES = ["unsignedTx", "globalXpub"], (function(e) {
      e[e.NON_WITNESS_UTXO = 0] = "NON_WITNESS_UTXO", e[e.WITNESS_UTXO = 1] = "WITNESS_UTXO", e[e.PARTIAL_SIG = 2] = "PARTIAL_SIG", e[e.SIGHASH_TYPE = 3] = "SIGHASH_TYPE", e[e.REDEEM_SCRIPT = 4] = "REDEEM_SCRIPT", e[e.WITNESS_SCRIPT = 5] = "WITNESS_SCRIPT", e[e.BIP32_DERIVATION = 6] = "BIP32_DERIVATION", e[e.FINAL_SCRIPTSIG = 7] = "FINAL_SCRIPTSIG", e[e.FINAL_SCRIPTWITNESS = 8] = "FINAL_SCRIPTWITNESS", e[e.POR_COMMITMENT = 9] = "POR_COMMITMENT", e[e.TAP_KEY_SIG = 19] = "TAP_KEY_SIG", e[e.TAP_SCRIPT_SIG = 20] = "TAP_SCRIPT_SIG", e[e.TAP_LEAF_SCRIPT = 21] = "TAP_LEAF_SCRIPT", e[e.TAP_BIP32_DERIVATION = 22] = "TAP_BIP32_DERIVATION", e[e.TAP_INTERNAL_KEY = 23] = "TAP_INTERNAL_KEY", e[e.TAP_MERKLE_ROOT = 24] = "TAP_MERKLE_ROOT";
    })(t.InputTypes || (t.InputTypes = {})), t.INPUT_TYPE_NAMES = [
      "nonWitnessUtxo",
      "witnessUtxo",
      "partialSig",
      "sighashType",
      "redeemScript",
      "witnessScript",
      "bip32Derivation",
      "finalScriptSig",
      "finalScriptWitness",
      "porCommitment",
      "tapKeySig",
      "tapScriptSig",
      "tapLeafScript",
      "tapBip32Derivation",
      "tapInternalKey",
      "tapMerkleRoot"
    ], (function(e) {
      e[e.REDEEM_SCRIPT = 0] = "REDEEM_SCRIPT", e[e.WITNESS_SCRIPT = 1] = "WITNESS_SCRIPT", e[e.BIP32_DERIVATION = 2] = "BIP32_DERIVATION", e[e.TAP_INTERNAL_KEY = 5] = "TAP_INTERNAL_KEY", e[e.TAP_TREE = 6] = "TAP_TREE", e[e.TAP_BIP32_DERIVATION = 7] = "TAP_BIP32_DERIVATION";
    })(t.OutputTypes || (t.OutputTypes = {})), t.OUTPUT_TYPE_NAMES = [
      "redeemScript",
      "witnessScript",
      "bip32Derivation",
      "tapInternalKey",
      "tapTree",
      "tapBip32Derivation"
    ];
  })(ot)), ot;
}
var W = {}, Mt;
function vs() {
  if (Mt) return W;
  Mt = 1, Object.defineProperty(W, "__esModule", { value: !0 });
  const t = A(), e = (o) => [...Array(o).keys()];
  function n(o) {
    if (o.key[0] !== t.GlobalTypes.GLOBAL_XPUB)
      throw new Error(
        "Decode Error: could not decode globalXpub with key 0x" + o.key.toString("hex")
      );
    if (o.key.length !== 79 || ![2, 3].includes(o.key[46]))
      throw new Error(
        "Decode Error: globalXpub has invalid extended pubkey in key 0x" + o.key.toString("hex")
      );
    if (o.value.length / 4 % 1 !== 0)
      throw new Error(
        "Decode Error: Global GLOBAL_XPUB value length should be multiple of 4"
      );
    const i = o.key.slice(1), a = {
      masterFingerprint: o.value.slice(0, 4),
      extendedPubkey: i,
      path: "m"
    };
    for (const u of e(o.value.length / 4 - 1)) {
      const f = o.value.readUInt32LE(u * 4 + 4), h = !!(f & 2147483648), g = f & 2147483647;
      a.path += "/" + g.toString(10) + (h ? "'" : "");
    }
    return a;
  }
  W.decode = n;
  function r(o) {
    const i = l.from([t.GlobalTypes.GLOBAL_XPUB]), a = l.concat([i, o.extendedPubkey]), u = o.path.split("/"), f = l.allocUnsafe(u.length * 4);
    o.masterFingerprint.copy(f, 0);
    let h = 4;
    return u.slice(1).forEach((g) => {
      const y = g.slice(-1) === "'";
      let p = 2147483647 & parseInt(y ? g.slice(0, -1) : g, 10);
      y && (p += 2147483648), f.writeUInt32LE(p, h), h += 4;
    }), {
      key: a,
      value: f
    };
  }
  W.encode = r, W.expected = "{ masterFingerprint: Buffer; extendedPubkey: Buffer; path: string; }";
  function c(o) {
    const i = o.extendedPubkey, a = o.masterFingerprint, u = o.path;
    return l.isBuffer(i) && i.length === 78 && [2, 3].indexOf(i[45]) > -1 && l.isBuffer(a) && a.length === 4 && typeof u == "string" && !!u.match(/^m(\/\d+'?)*$/);
  }
  W.check = c;
  function s(o, i, a) {
    const u = i.extendedPubkey.toString("hex");
    return a.has(u) ? !1 : (a.add(u), o.filter((f) => f.extendedPubkey.equals(i.extendedPubkey)).length === 0);
  }
  return W.canAddToArray = s, W;
}
var Ce = {}, Wt;
function ks() {
  if (Wt) return Ce;
  Wt = 1, Object.defineProperty(Ce, "__esModule", { value: !0 });
  const t = A();
  function e(n) {
    return {
      key: l.from([t.GlobalTypes.UNSIGNED_TX]),
      value: n.toBuffer()
    };
  }
  return Ce.encode = e, Ce;
}
var G = {}, Gt;
function As() {
  if (Gt) return G;
  Gt = 1, Object.defineProperty(G, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.FINAL_SCRIPTSIG)
      throw new Error(
        "Decode Error: could not decode finalScriptSig with key 0x" + s.key.toString("hex")
      );
    return s.value;
  }
  G.decode = e;
  function n(s) {
    return {
      key: l.from([t.InputTypes.FINAL_SCRIPTSIG]),
      value: s
    };
  }
  G.encode = n, G.expected = "Buffer";
  function r(s) {
    return l.isBuffer(s);
  }
  G.check = r;
  function c(s, o) {
    return !!s && !!o && s.finalScriptSig === void 0;
  }
  return G.canAdd = c, G;
}
var X = {}, Xt;
function xs() {
  if (Xt) return X;
  Xt = 1, Object.defineProperty(X, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.FINAL_SCRIPTWITNESS)
      throw new Error(
        "Decode Error: could not decode finalScriptWitness with key 0x" + s.key.toString("hex")
      );
    return s.value;
  }
  X.decode = e;
  function n(s) {
    return {
      key: l.from([t.InputTypes.FINAL_SCRIPTWITNESS]),
      value: s
    };
  }
  X.encode = n, X.expected = "Buffer";
  function r(s) {
    return l.isBuffer(s);
  }
  X.check = r;
  function c(s, o) {
    return !!s && !!o && s.finalScriptWitness === void 0;
  }
  return X.canAdd = c, X;
}
var D = {}, Dt;
function Cs() {
  if (Dt) return D;
  Dt = 1, Object.defineProperty(D, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.NON_WITNESS_UTXO)
      throw new Error(
        "Decode Error: could not decode nonWitnessUtxo with key 0x" + s.key.toString("hex")
      );
    return s.value;
  }
  D.decode = e;
  function n(s) {
    return {
      key: l.from([t.InputTypes.NON_WITNESS_UTXO]),
      value: s
    };
  }
  D.encode = n, D.expected = "Buffer";
  function r(s) {
    return l.isBuffer(s);
  }
  D.check = r;
  function c(s, o) {
    return !!s && !!o && s.nonWitnessUtxo === void 0;
  }
  return D.canAdd = c, D;
}
var q = {}, qt;
function Os() {
  if (qt) return q;
  qt = 1, Object.defineProperty(q, "__esModule", { value: !0 });
  const t = A();
  function e(o) {
    if (o.key[0] !== t.InputTypes.PARTIAL_SIG)
      throw new Error(
        "Decode Error: could not decode partialSig with key 0x" + o.key.toString("hex")
      );
    if (!(o.key.length === 34 || o.key.length === 66) || ![2, 3, 4].includes(o.key[1]))
      throw new Error(
        "Decode Error: partialSig has invalid pubkey in key 0x" + o.key.toString("hex")
      );
    return {
      pubkey: o.key.slice(1),
      signature: o.value
    };
  }
  q.decode = e;
  function n(o) {
    const i = l.from([t.InputTypes.PARTIAL_SIG]);
    return {
      key: l.concat([i, o.pubkey]),
      value: o.signature
    };
  }
  q.encode = n, q.expected = "{ pubkey: Buffer; signature: Buffer; }";
  function r(o) {
    return l.isBuffer(o.pubkey) && l.isBuffer(o.signature) && [33, 65].includes(o.pubkey.length) && [2, 3, 4].includes(o.pubkey[0]) && c(o.signature);
  }
  q.check = r;
  function c(o) {
    if (!l.isBuffer(o) || o.length < 9 || o[0] !== 48 || o.length !== o[1] + 3 || o[2] !== 2) return !1;
    const i = o[3];
    if (i > 33 || i < 1 || o[3 + i + 1] !== 2) return !1;
    const a = o[3 + i + 2];
    return !(a > 33 || a < 1 || o.length !== 3 + i + 2 + a + 2);
  }
  function s(o, i, a) {
    const u = i.pubkey.toString("hex");
    return a.has(u) ? !1 : (a.add(u), o.filter((f) => f.pubkey.equals(i.pubkey)).length === 0);
  }
  return q.canAddToArray = s, q;
}
var V = {}, Vt;
function Ps() {
  if (Vt) return V;
  Vt = 1, Object.defineProperty(V, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.POR_COMMITMENT)
      throw new Error(
        "Decode Error: could not decode porCommitment with key 0x" + s.key.toString("hex")
      );
    return s.value.toString("utf8");
  }
  V.decode = e;
  function n(s) {
    return {
      key: l.from([t.InputTypes.POR_COMMITMENT]),
      value: l.from(s, "utf8")
    };
  }
  V.encode = n, V.expected = "string";
  function r(s) {
    return typeof s == "string";
  }
  V.check = r;
  function c(s, o) {
    return !!s && !!o && s.porCommitment === void 0;
  }
  return V.canAdd = c, V;
}
var j = {}, jt;
function Ns() {
  if (jt) return j;
  jt = 1, Object.defineProperty(j, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.SIGHASH_TYPE)
      throw new Error(
        "Decode Error: could not decode sighashType with key 0x" + s.key.toString("hex")
      );
    return s.value.readUInt32LE(0);
  }
  j.decode = e;
  function n(s) {
    const o = l.from([t.InputTypes.SIGHASH_TYPE]), i = l.allocUnsafe(4);
    return i.writeUInt32LE(s, 0), {
      key: o,
      value: i
    };
  }
  j.encode = n, j.expected = "number";
  function r(s) {
    return typeof s == "number";
  }
  j.check = r;
  function c(s, o) {
    return !!s && !!o && s.sighashType === void 0;
  }
  return j.canAdd = c, j;
}
var z = {}, zt;
function Us() {
  if (zt) return z;
  zt = 1, Object.defineProperty(z, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.TAP_KEY_SIG || s.key.length !== 1)
      throw new Error(
        "Decode Error: could not decode tapKeySig with key 0x" + s.key.toString("hex")
      );
    if (!r(s.value))
      throw new Error(
        "Decode Error: tapKeySig not a valid 64-65-byte BIP340 signature"
      );
    return s.value;
  }
  z.decode = e;
  function n(s) {
    return { key: l.from([t.InputTypes.TAP_KEY_SIG]), value: s };
  }
  z.encode = n, z.expected = "Buffer";
  function r(s) {
    return l.isBuffer(s) && (s.length === 64 || s.length === 65);
  }
  z.check = r;
  function c(s, o) {
    return !!s && !!o && s.tapKeySig === void 0;
  }
  return z.canAdd = c, z;
}
var Y = {}, Yt;
function Hs() {
  if (Yt) return Y;
  Yt = 1, Object.defineProperty(Y, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.TAP_LEAF_SCRIPT)
      throw new Error(
        "Decode Error: could not decode tapLeafScript with key 0x" + s.key.toString("hex")
      );
    if ((s.key.length - 2) % 32 !== 0)
      throw new Error(
        "Decode Error: tapLeafScript has invalid control block in key 0x" + s.key.toString("hex")
      );
    const o = s.value[s.value.length - 1];
    if ((s.key[1] & 254) !== o)
      throw new Error(
        "Decode Error: tapLeafScript bad leaf version in key 0x" + s.key.toString("hex")
      );
    const i = s.value.slice(0, -1);
    return { controlBlock: s.key.slice(1), script: i, leafVersion: o };
  }
  Y.decode = e;
  function n(s) {
    const o = l.from([t.InputTypes.TAP_LEAF_SCRIPT]), i = l.from([s.leafVersion]);
    return {
      key: l.concat([o, s.controlBlock]),
      value: l.concat([s.script, i])
    };
  }
  Y.encode = n, Y.expected = "{ controlBlock: Buffer; leafVersion: number, script: Buffer; }";
  function r(s) {
    return l.isBuffer(s.controlBlock) && (s.controlBlock.length - 1) % 32 === 0 && (s.controlBlock[0] & 254) === s.leafVersion && l.isBuffer(s.script);
  }
  Y.check = r;
  function c(s, o, i) {
    const a = o.controlBlock.toString("hex");
    return i.has(a) ? !1 : (i.add(a), s.filter((u) => u.controlBlock.equals(o.controlBlock)).length === 0);
  }
  return Y.canAddToArray = c, Y;
}
var $ = {}, $t;
function Bs() {
  if ($t) return $;
  $t = 1, Object.defineProperty($, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.TAP_MERKLE_ROOT || s.key.length !== 1)
      throw new Error(
        "Decode Error: could not decode tapMerkleRoot with key 0x" + s.key.toString("hex")
      );
    if (!r(s.value))
      throw new Error("Decode Error: tapMerkleRoot not a 32-byte hash");
    return s.value;
  }
  $.decode = e;
  function n(s) {
    return { key: l.from([t.InputTypes.TAP_MERKLE_ROOT]), value: s };
  }
  $.encode = n, $.expected = "Buffer";
  function r(s) {
    return l.isBuffer(s) && s.length === 32;
  }
  $.check = r;
  function c(s, o) {
    return !!s && !!o && s.tapMerkleRoot === void 0;
  }
  return $.canAdd = c, $;
}
var Z = {}, Zt;
function Rs() {
  if (Zt) return Z;
  Zt = 1, Object.defineProperty(Z, "__esModule", { value: !0 });
  const t = A();
  function e(s) {
    if (s.key[0] !== t.InputTypes.TAP_SCRIPT_SIG)
      throw new Error(
        "Decode Error: could not decode tapScriptSig with key 0x" + s.key.toString("hex")
      );
    if (s.key.length !== 65)
      throw new Error(
        "Decode Error: tapScriptSig has invalid key 0x" + s.key.toString("hex")
      );
    if (s.value.length !== 64 && s.value.length !== 65)
      throw new Error(
        "Decode Error: tapScriptSig has invalid signature in key 0x" + s.key.toString("hex")
      );
    const o = s.key.slice(1, 33), i = s.key.slice(33);
    return {
      pubkey: o,
      leafHash: i,
      signature: s.value
    };
  }
  Z.decode = e;
  function n(s) {
    const o = l.from([t.InputTypes.TAP_SCRIPT_SIG]);
    return {
      key: l.concat([o, s.pubkey, s.leafHash]),
      value: s.signature
    };
  }
  Z.encode = n, Z.expected = "{ pubkey: Buffer; leafHash: Buffer; signature: Buffer; }";
  function r(s) {
    return l.isBuffer(s.pubkey) && l.isBuffer(s.leafHash) && l.isBuffer(s.signature) && s.pubkey.length === 32 && s.leafHash.length === 32 && (s.signature.length === 64 || s.signature.length === 65);
  }
  Z.check = r;
  function c(s, o, i) {
    const a = o.pubkey.toString("hex") + o.leafHash.toString("hex");
    return i.has(a) ? !1 : (i.add(a), s.filter(
      (u) => u.pubkey.equals(o.pubkey) && u.leafHash.equals(o.leafHash)
    ).length === 0);
  }
  return Z.canAddToArray = c, Z;
}
var J = {}, F = {}, ce = {}, Jt;
function fe() {
  if (Jt) return ce;
  Jt = 1, Object.defineProperty(ce, "__esModule", { value: !0 });
  const t = 9007199254740991;
  function e(s) {
    if (s < 0 || s > t || s % 1 !== 0)
      throw new RangeError("value out of range");
  }
  function n(s, o, i) {
    if (e(s), o || (o = l.allocUnsafe(c(s))), !l.isBuffer(o))
      throw new TypeError("buffer must be a Buffer instance");
    return i || (i = 0), s < 253 ? (o.writeUInt8(s, i), Object.assign(n, { bytes: 1 })) : s <= 65535 ? (o.writeUInt8(253, i), o.writeUInt16LE(s, i + 1), Object.assign(n, { bytes: 3 })) : s <= 4294967295 ? (o.writeUInt8(254, i), o.writeUInt32LE(s, i + 1), Object.assign(n, { bytes: 5 })) : (o.writeUInt8(255, i), o.writeUInt32LE(s >>> 0, i + 1), o.writeUInt32LE(s / 4294967296 | 0, i + 5), Object.assign(n, { bytes: 9 })), o;
  }
  ce.encode = n;
  function r(s, o) {
    if (!l.isBuffer(s))
      throw new TypeError("buffer must be a Buffer instance");
    o || (o = 0);
    const i = s.readUInt8(o);
    if (i < 253)
      return Object.assign(r, { bytes: 1 }), i;
    if (i === 253)
      return Object.assign(r, { bytes: 3 }), s.readUInt16LE(o + 1);
    if (i === 254)
      return Object.assign(r, { bytes: 5 }), s.readUInt32LE(o + 1);
    {
      Object.assign(r, { bytes: 9 });
      const a = s.readUInt32LE(o + 1), f = s.readUInt32LE(o + 5) * 4294967296 + a;
      return e(f), f;
    }
  }
  ce.decode = r;
  function c(s) {
    return e(s), s < 253 ? 1 : s <= 65535 ? 3 : s <= 4294967295 ? 5 : 9;
  }
  return ce.encodingLength = c, ce;
}
var Qt;
function mt() {
  if (Qt) return F;
  Qt = 1, Object.defineProperty(F, "__esModule", { value: !0 });
  const t = fe();
  F.range = (i) => [...Array(i).keys()];
  function e(i) {
    if (i.length < 1) return i;
    let a = i.length - 1, u = 0;
    for (let f = 0; f < i.length / 2; f++)
      u = i[f], i[f] = i[a], i[a] = u, a--;
    return i;
  }
  F.reverseBuffer = e;
  function n(i) {
    const a = i.map(r);
    return a.push(l.from([0])), l.concat(a);
  }
  F.keyValsToBuffer = n;
  function r(i) {
    const a = i.key.length, u = i.value.length, f = t.encodingLength(a), h = t.encodingLength(u), g = l.allocUnsafe(
      f + a + h + u
    );
    return t.encode(a, g, 0), i.key.copy(g, f), t.encode(u, g, f + a), i.value.copy(g, f + a + h), g;
  }
  F.keyValToBuffer = r;
  function c(i, a) {
    if (typeof i != "number")
      throw new Error("cannot write a non-number as a number");
    if (i < 0)
      throw new Error("specified a negative value for writing an unsigned value");
    if (i > a) throw new Error("RangeError: value out of range");
    if (Math.floor(i) !== i)
      throw new Error("value has a fractional component");
  }
  function s(i, a) {
    const u = i.readUInt32LE(a);
    let f = i.readUInt32LE(a + 4);
    return f *= 4294967296, c(f + u, 9007199254740991), f + u;
  }
  F.readUInt64LE = s;
  function o(i, a, u) {
    return c(a, 9007199254740991), i.writeInt32LE(a & -1, u), i.writeUInt32LE(Math.floor(a / 4294967296), u + 4), u + 8;
  }
  return F.writeUInt64LE = o, F;
}
var en;
function Fs() {
  if (en) return J;
  en = 1, Object.defineProperty(J, "__esModule", { value: !0 });
  const t = A(), e = mt(), n = fe();
  function r(i) {
    if (i.key[0] !== t.InputTypes.WITNESS_UTXO)
      throw new Error(
        "Decode Error: could not decode witnessUtxo with key 0x" + i.key.toString("hex")
      );
    const a = e.readUInt64LE(i.value, 0);
    let u = 8;
    const f = n.decode(i.value, u);
    u += n.encodingLength(f);
    const h = i.value.slice(u);
    if (h.length !== f)
      throw new Error("Decode Error: WITNESS_UTXO script is not proper length");
    return {
      script: h,
      value: a
    };
  }
  J.decode = r;
  function c(i) {
    const { script: a, value: u } = i, f = n.encodingLength(a.length), h = l.allocUnsafe(8 + f + a.length);
    return e.writeUInt64LE(h, u, 0), n.encode(a.length, h, 8), a.copy(h, 8 + f), {
      key: l.from([t.InputTypes.WITNESS_UTXO]),
      value: h
    };
  }
  J.encode = c, J.expected = "{ script: Buffer; value: number; }";
  function s(i) {
    return l.isBuffer(i.script) && typeof i.value == "number";
  }
  J.check = s;
  function o(i, a) {
    return !!i && !!a && i.witnessUtxo === void 0;
  }
  return J.canAdd = o, J;
}
var Q = {}, tn;
function Ks() {
  if (tn) return Q;
  tn = 1, Object.defineProperty(Q, "__esModule", { value: !0 });
  const t = A(), e = fe();
  function n(o) {
    if (o.key[0] !== t.OutputTypes.TAP_TREE || o.key.length !== 1)
      throw new Error(
        "Decode Error: could not decode tapTree with key 0x" + o.key.toString("hex")
      );
    let i = 0;
    const a = [];
    for (; i < o.value.length; ) {
      const u = o.value[i++], f = o.value[i++], h = e.decode(o.value, i);
      i += e.encodingLength(h), a.push({
        depth: u,
        leafVersion: f,
        script: o.value.slice(i, i + h)
      }), i += h;
    }
    return { leaves: a };
  }
  Q.decode = n;
  function r(o) {
    const i = l.from([t.OutputTypes.TAP_TREE]), a = [].concat(
      ...o.leaves.map((u) => [
        l.of(u.depth, u.leafVersion),
        e.encode(u.script.length),
        u.script
      ])
    );
    return {
      key: i,
      value: l.concat(a)
    };
  }
  Q.encode = r, Q.expected = "{ leaves: [{ depth: number; leafVersion: number, script: Buffer; }] }";
  function c(o) {
    return Array.isArray(o.leaves) && o.leaves.every(
      (i) => i.depth >= 0 && i.depth <= 128 && (i.leafVersion & 254) === i.leafVersion && l.isBuffer(i.script)
    );
  }
  Q.check = c;
  function s(o, i) {
    return !!o && !!i && o.tapTree === void 0;
  }
  return Q.canAdd = s, Q;
}
var Oe = {}, nn;
function nr() {
  if (nn) return Oe;
  nn = 1, Object.defineProperty(Oe, "__esModule", { value: !0 });
  const t = (r) => [...Array(r).keys()], e = (r) => r.length === 33 && [2, 3].includes(r[0]) || r.length === 65 && r[0] === 4;
  function n(r, c = e) {
    function s(f) {
      if (f.key[0] !== r)
        throw new Error(
          "Decode Error: could not decode bip32Derivation with key 0x" + f.key.toString("hex")
        );
      const h = f.key.slice(1);
      if (!c(h))
        throw new Error(
          "Decode Error: bip32Derivation has invalid pubkey in key 0x" + f.key.toString("hex")
        );
      if (f.value.length / 4 % 1 !== 0)
        throw new Error(
          "Decode Error: Input BIP32_DERIVATION value length should be multiple of 4"
        );
      const g = {
        masterFingerprint: f.value.slice(0, 4),
        pubkey: h,
        path: "m"
      };
      for (const y of t(f.value.length / 4 - 1)) {
        const p = f.value.readUInt32LE(y * 4 + 4), d = !!(p & 2147483648), E = p & 2147483647;
        g.path += "/" + E.toString(10) + (d ? "'" : "");
      }
      return g;
    }
    function o(f) {
      const h = l.from([r]), g = l.concat([h, f.pubkey]), y = f.path.split("/"), p = l.allocUnsafe(y.length * 4);
      f.masterFingerprint.copy(p, 0);
      let d = 4;
      return y.slice(1).forEach((E) => {
        const b = E.slice(-1) === "'";
        let S = 2147483647 & parseInt(b ? E.slice(0, -1) : E, 10);
        b && (S += 2147483648), p.writeUInt32LE(S, d), d += 4;
      }), {
        key: g,
        value: p
      };
    }
    const i = "{ masterFingerprint: Buffer; pubkey: Buffer; path: string; }";
    function a(f) {
      return l.isBuffer(f.pubkey) && l.isBuffer(f.masterFingerprint) && typeof f.path == "string" && c(f.pubkey) && f.masterFingerprint.length === 4;
    }
    function u(f, h, g) {
      const y = h.pubkey.toString("hex");
      return g.has(y) ? !1 : (g.add(y), f.filter((p) => p.pubkey.equals(h.pubkey)).length === 0);
    }
    return {
      decode: s,
      encode: o,
      check: a,
      expected: i,
      canAddToArray: u
    };
  }
  return Oe.makeConverter = n, Oe;
}
var Pe = {}, rn;
function Ls() {
  if (rn) return Pe;
  rn = 1, Object.defineProperty(Pe, "__esModule", { value: !0 });
  function t(e) {
    return n;
    function n(r) {
      let c;
      if (e.includes(r.key[0]) && (c = r.key.slice(1), !(c.length === 33 || c.length === 65) || ![2, 3, 4].includes(c[0])))
        throw new Error(
          "Format Error: invalid pubkey in key 0x" + r.key.toString("hex")
        );
      return c;
    }
  }
  return Pe.makeChecker = t, Pe;
}
var Ne = {}, sn;
function Ms() {
  if (sn) return Ne;
  sn = 1, Object.defineProperty(Ne, "__esModule", { value: !0 });
  function t(e) {
    function n(i) {
      if (i.key[0] !== e)
        throw new Error(
          "Decode Error: could not decode redeemScript with key 0x" + i.key.toString("hex")
        );
      return i.value;
    }
    function r(i) {
      return {
        key: l.from([e]),
        value: i
      };
    }
    const c = "Buffer";
    function s(i) {
      return l.isBuffer(i);
    }
    function o(i, a) {
      return !!i && !!a && i.redeemScript === void 0;
    }
    return {
      decode: n,
      encode: r,
      check: s,
      expected: c,
      canAdd: o
    };
  }
  return Ne.makeConverter = t, Ne;
}
var Ue = {}, on;
function Ws() {
  if (on) return Ue;
  on = 1, Object.defineProperty(Ue, "__esModule", { value: !0 });
  const t = fe(), e = nr(), n = (c) => c.length === 32;
  function r(c) {
    const s = e.makeConverter(c, n);
    function o(f) {
      const h = t.decode(f.value), g = t.encodingLength(h), y = s.decode({
        key: f.key,
        value: f.value.slice(g + h * 32)
      }), p = new Array(h);
      for (let d = 0, E = g; d < h; d++, E += 32)
        p[d] = f.value.slice(E, E + 32);
      return Object.assign({}, y, { leafHashes: p });
    }
    function i(f) {
      const h = s.encode(f), g = t.encodingLength(f.leafHashes.length), y = l.allocUnsafe(g);
      t.encode(f.leafHashes.length, y);
      const p = l.concat([y, ...f.leafHashes, h.value]);
      return Object.assign({}, h, { value: p });
    }
    const a = "{ masterFingerprint: Buffer; pubkey: Buffer; path: string; leafHashes: Buffer[]; }";
    function u(f) {
      return Array.isArray(f.leafHashes) && f.leafHashes.every(
        (h) => l.isBuffer(h) && h.length === 32
      ) && s.check(f);
    }
    return {
      decode: o,
      encode: i,
      check: u,
      expected: a,
      canAddToArray: s.canAddToArray
    };
  }
  return Ue.makeConverter = r, Ue;
}
var He = {}, cn;
function Gs() {
  if (cn) return He;
  cn = 1, Object.defineProperty(He, "__esModule", { value: !0 });
  function t(e) {
    function n(i) {
      if (i.key[0] !== e || i.key.length !== 1)
        throw new Error(
          "Decode Error: could not decode tapInternalKey with key 0x" + i.key.toString("hex")
        );
      if (i.value.length !== 32)
        throw new Error(
          "Decode Error: tapInternalKey not a 32-byte x-only pubkey"
        );
      return i.value;
    }
    function r(i) {
      return { key: l.from([e]), value: i };
    }
    const c = "Buffer";
    function s(i) {
      return l.isBuffer(i) && i.length === 32;
    }
    function o(i, a) {
      return !!i && !!a && i.tapInternalKey === void 0;
    }
    return {
      decode: n,
      encode: r,
      check: s,
      expected: c,
      canAdd: o
    };
  }
  return He.makeConverter = t, He;
}
var Be = {}, an;
function Xs() {
  if (an) return Be;
  an = 1, Object.defineProperty(Be, "__esModule", { value: !0 });
  function t(e) {
    function n(i) {
      if (i.key[0] !== e)
        throw new Error(
          "Decode Error: could not decode witnessScript with key 0x" + i.key.toString("hex")
        );
      return i.value;
    }
    function r(i) {
      return {
        key: l.from([e]),
        value: i
      };
    }
    const c = "Buffer";
    function s(i) {
      return l.isBuffer(i);
    }
    function o(i, a) {
      return !!i && !!a && i.witnessScript === void 0;
    }
    return {
      decode: n,
      encode: r,
      check: s,
      expected: c,
      canAdd: o
    };
  }
  return Be.makeConverter = t, Be;
}
var un;
function It() {
  if (un) return ie;
  un = 1, Object.defineProperty(ie, "__esModule", { value: !0 });
  const t = A(), e = vs(), n = ks(), r = As(), c = xs(), s = Cs(), o = Os(), i = Ps(), a = Ns(), u = Us(), f = Hs(), h = Bs(), g = Rs(), y = Fs(), p = Ks(), d = nr(), E = Ls(), b = Ms(), S = Ws(), w = Gs(), _ = Xs(), I = {
    unsignedTx: n,
    globalXpub: e,
    // pass an Array of key bytes that require pubkey beside the key
    checkPubkey: E.makeChecker([])
  };
  ie.globals = I;
  const P = {
    nonWitnessUtxo: s,
    partialSig: o,
    sighashType: a,
    finalScriptSig: r,
    finalScriptWitness: c,
    porCommitment: i,
    witnessUtxo: y,
    bip32Derivation: d.makeConverter(
      t.InputTypes.BIP32_DERIVATION
    ),
    redeemScript: b.makeConverter(
      t.InputTypes.REDEEM_SCRIPT
    ),
    witnessScript: _.makeConverter(
      t.InputTypes.WITNESS_SCRIPT
    ),
    checkPubkey: E.makeChecker([
      t.InputTypes.PARTIAL_SIG,
      t.InputTypes.BIP32_DERIVATION
    ]),
    tapKeySig: u,
    tapScriptSig: g,
    tapLeafScript: f,
    tapBip32Derivation: S.makeConverter(
      t.InputTypes.TAP_BIP32_DERIVATION
    ),
    tapInternalKey: w.makeConverter(
      t.InputTypes.TAP_INTERNAL_KEY
    ),
    tapMerkleRoot: h
  };
  ie.inputs = P;
  const m = {
    bip32Derivation: d.makeConverter(
      t.OutputTypes.BIP32_DERIVATION
    ),
    redeemScript: b.makeConverter(
      t.OutputTypes.REDEEM_SCRIPT
    ),
    witnessScript: _.makeConverter(
      t.OutputTypes.WITNESS_SCRIPT
    ),
    checkPubkey: E.makeChecker([
      t.OutputTypes.BIP32_DERIVATION
    ]),
    tapBip32Derivation: S.makeConverter(
      t.OutputTypes.TAP_BIP32_DERIVATION
    ),
    tapTree: p,
    tapInternalKey: w.makeConverter(
      t.OutputTypes.TAP_INTERNAL_KEY
    )
  };
  return ie.outputs = m, ie;
}
var fn;
function Ds() {
  if (fn) return oe;
  fn = 1, Object.defineProperty(oe, "__esModule", { value: !0 });
  const t = It(), e = mt(), n = fe(), r = A();
  function c(i, a) {
    let u = 0;
    function f() {
      const m = n.decode(i, u);
      u += n.encodingLength(m);
      const k = i.slice(u, u + m);
      return u += m, k;
    }
    function h() {
      const m = i.readUInt32BE(u);
      return u += 4, m;
    }
    function g() {
      const m = i.readUInt8(u);
      return u += 1, m;
    }
    function y() {
      const m = f(), k = f();
      return {
        key: m,
        value: k
      };
    }
    function p() {
      if (u >= i.length)
        throw new Error("Format Error: Unexpected End of PSBT");
      const m = i.readUInt8(u) === 0;
      return m && u++, m;
    }
    if (h() !== 1886610036)
      throw new Error("Format Error: Invalid Magic Number");
    if (g() !== 255)
      throw new Error(
        "Format Error: Magic Number must be followed by 0xff separator"
      );
    const d = [], E = {};
    for (; !p(); ) {
      const m = y(), k = m.key.toString("hex");
      if (E[k])
        throw new Error(
          "Format Error: Keys must be unique for global keymap: key " + k
        );
      E[k] = 1, d.push(m);
    }
    const b = d.filter(
      (m) => m.key[0] === r.GlobalTypes.UNSIGNED_TX
    );
    if (b.length !== 1)
      throw new Error("Format Error: Only one UNSIGNED_TX allowed");
    const S = a(b[0].value), { inputCount: w, outputCount: _ } = S.getInputOutputCounts(), I = [], P = [];
    for (const m of e.range(w)) {
      const k = {}, U = [];
      for (; !p(); ) {
        const B = y(), te = B.key.toString("hex");
        if (k[te])
          throw new Error(
            "Format Error: Keys must be unique for each input: input index " + m + " key " + te
          );
        k[te] = 1, U.push(B);
      }
      I.push(U);
    }
    for (const m of e.range(_)) {
      const k = {}, U = [];
      for (; !p(); ) {
        const B = y(), te = B.key.toString("hex");
        if (k[te])
          throw new Error(
            "Format Error: Keys must be unique for each output: output index " + m + " key " + te
          );
        k[te] = 1, U.push(B);
      }
      P.push(U);
    }
    return o(S, {
      globalMapKeyVals: d,
      inputKeyVals: I,
      outputKeyVals: P
    });
  }
  oe.psbtFromBuffer = c;
  function s(i, a, u) {
    if (!a.equals(l.from([u])))
      throw new Error(
        `Format Error: Invalid ${i} key: ${a.toString("hex")}`
      );
  }
  oe.checkKeyBuffer = s;
  function o(i, { globalMapKeyVals: a, inputKeyVals: u, outputKeyVals: f }) {
    const h = {
      unsignedTx: i
    };
    let g = 0;
    for (const b of a)
      switch (b.key[0]) {
        case r.GlobalTypes.UNSIGNED_TX:
          if (s(
            "global",
            b.key,
            r.GlobalTypes.UNSIGNED_TX
          ), g > 0)
            throw new Error("Format Error: GlobalMap has multiple UNSIGNED_TX");
          g++;
          break;
        case r.GlobalTypes.GLOBAL_XPUB:
          h.globalXpub === void 0 && (h.globalXpub = []), h.globalXpub.push(t.globals.globalXpub.decode(b));
          break;
        default:
          h.unknownKeyVals || (h.unknownKeyVals = []), h.unknownKeyVals.push(b);
      }
    const y = u.length, p = f.length, d = [], E = [];
    for (const b of e.range(y)) {
      const S = {};
      for (const w of u[b])
        switch (t.inputs.checkPubkey(w), w.key[0]) {
          case r.InputTypes.NON_WITNESS_UTXO:
            if (s(
              "input",
              w.key,
              r.InputTypes.NON_WITNESS_UTXO
            ), S.nonWitnessUtxo !== void 0)
              throw new Error(
                "Format Error: Input has multiple NON_WITNESS_UTXO"
              );
            S.nonWitnessUtxo = t.inputs.nonWitnessUtxo.decode(w);
            break;
          case r.InputTypes.WITNESS_UTXO:
            if (s(
              "input",
              w.key,
              r.InputTypes.WITNESS_UTXO
            ), S.witnessUtxo !== void 0)
              throw new Error("Format Error: Input has multiple WITNESS_UTXO");
            S.witnessUtxo = t.inputs.witnessUtxo.decode(w);
            break;
          case r.InputTypes.PARTIAL_SIG:
            S.partialSig === void 0 && (S.partialSig = []), S.partialSig.push(t.inputs.partialSig.decode(w));
            break;
          case r.InputTypes.SIGHASH_TYPE:
            if (s(
              "input",
              w.key,
              r.InputTypes.SIGHASH_TYPE
            ), S.sighashType !== void 0)
              throw new Error("Format Error: Input has multiple SIGHASH_TYPE");
            S.sighashType = t.inputs.sighashType.decode(w);
            break;
          case r.InputTypes.REDEEM_SCRIPT:
            if (s(
              "input",
              w.key,
              r.InputTypes.REDEEM_SCRIPT
            ), S.redeemScript !== void 0)
              throw new Error("Format Error: Input has multiple REDEEM_SCRIPT");
            S.redeemScript = t.inputs.redeemScript.decode(w);
            break;
          case r.InputTypes.WITNESS_SCRIPT:
            if (s(
              "input",
              w.key,
              r.InputTypes.WITNESS_SCRIPT
            ), S.witnessScript !== void 0)
              throw new Error("Format Error: Input has multiple WITNESS_SCRIPT");
            S.witnessScript = t.inputs.witnessScript.decode(w);
            break;
          case r.InputTypes.BIP32_DERIVATION:
            S.bip32Derivation === void 0 && (S.bip32Derivation = []), S.bip32Derivation.push(
              t.inputs.bip32Derivation.decode(w)
            );
            break;
          case r.InputTypes.FINAL_SCRIPTSIG:
            s(
              "input",
              w.key,
              r.InputTypes.FINAL_SCRIPTSIG
            ), S.finalScriptSig = t.inputs.finalScriptSig.decode(w);
            break;
          case r.InputTypes.FINAL_SCRIPTWITNESS:
            s(
              "input",
              w.key,
              r.InputTypes.FINAL_SCRIPTWITNESS
            ), S.finalScriptWitness = t.inputs.finalScriptWitness.decode(
              w
            );
            break;
          case r.InputTypes.POR_COMMITMENT:
            s(
              "input",
              w.key,
              r.InputTypes.POR_COMMITMENT
            ), S.porCommitment = t.inputs.porCommitment.decode(w);
            break;
          case r.InputTypes.TAP_KEY_SIG:
            s(
              "input",
              w.key,
              r.InputTypes.TAP_KEY_SIG
            ), S.tapKeySig = t.inputs.tapKeySig.decode(w);
            break;
          case r.InputTypes.TAP_SCRIPT_SIG:
            S.tapScriptSig === void 0 && (S.tapScriptSig = []), S.tapScriptSig.push(t.inputs.tapScriptSig.decode(w));
            break;
          case r.InputTypes.TAP_LEAF_SCRIPT:
            S.tapLeafScript === void 0 && (S.tapLeafScript = []), S.tapLeafScript.push(t.inputs.tapLeafScript.decode(w));
            break;
          case r.InputTypes.TAP_BIP32_DERIVATION:
            S.tapBip32Derivation === void 0 && (S.tapBip32Derivation = []), S.tapBip32Derivation.push(
              t.inputs.tapBip32Derivation.decode(w)
            );
            break;
          case r.InputTypes.TAP_INTERNAL_KEY:
            s(
              "input",
              w.key,
              r.InputTypes.TAP_INTERNAL_KEY
            ), S.tapInternalKey = t.inputs.tapInternalKey.decode(w);
            break;
          case r.InputTypes.TAP_MERKLE_ROOT:
            s(
              "input",
              w.key,
              r.InputTypes.TAP_MERKLE_ROOT
            ), S.tapMerkleRoot = t.inputs.tapMerkleRoot.decode(w);
            break;
          default:
            S.unknownKeyVals || (S.unknownKeyVals = []), S.unknownKeyVals.push(w);
        }
      d.push(S);
    }
    for (const b of e.range(p)) {
      const S = {};
      for (const w of f[b])
        switch (t.outputs.checkPubkey(w), w.key[0]) {
          case r.OutputTypes.REDEEM_SCRIPT:
            if (s(
              "output",
              w.key,
              r.OutputTypes.REDEEM_SCRIPT
            ), S.redeemScript !== void 0)
              throw new Error("Format Error: Output has multiple REDEEM_SCRIPT");
            S.redeemScript = t.outputs.redeemScript.decode(w);
            break;
          case r.OutputTypes.WITNESS_SCRIPT:
            if (s(
              "output",
              w.key,
              r.OutputTypes.WITNESS_SCRIPT
            ), S.witnessScript !== void 0)
              throw new Error("Format Error: Output has multiple WITNESS_SCRIPT");
            S.witnessScript = t.outputs.witnessScript.decode(w);
            break;
          case r.OutputTypes.BIP32_DERIVATION:
            S.bip32Derivation === void 0 && (S.bip32Derivation = []), S.bip32Derivation.push(
              t.outputs.bip32Derivation.decode(w)
            );
            break;
          case r.OutputTypes.TAP_INTERNAL_KEY:
            s(
              "output",
              w.key,
              r.OutputTypes.TAP_INTERNAL_KEY
            ), S.tapInternalKey = t.outputs.tapInternalKey.decode(w);
            break;
          case r.OutputTypes.TAP_TREE:
            s(
              "output",
              w.key,
              r.OutputTypes.TAP_TREE
            ), S.tapTree = t.outputs.tapTree.decode(w);
            break;
          case r.OutputTypes.TAP_BIP32_DERIVATION:
            S.tapBip32Derivation === void 0 && (S.tapBip32Derivation = []), S.tapBip32Derivation.push(
              t.outputs.tapBip32Derivation.decode(w)
            );
            break;
          default:
            S.unknownKeyVals || (S.unknownKeyVals = []), S.unknownKeyVals.push(w);
        }
      E.push(S);
    }
    return { globalMap: h, inputs: d, outputs: E };
  }
  return oe.psbtFromKeyVals = o, oe;
}
var he = {}, hn;
function qs() {
  if (hn) return he;
  hn = 1, Object.defineProperty(he, "__esModule", { value: !0 });
  const t = It(), e = mt();
  function n({ globalMap: o, inputs: i, outputs: a }) {
    const { globalKeyVals: u, inputKeyVals: f, outputKeyVals: h } = s({
      globalMap: o,
      inputs: i,
      outputs: a
    }), g = e.keyValsToBuffer(u), y = (b) => b.length === 0 ? [l.from([0])] : b.map(e.keyValsToBuffer), p = y(f), d = y(h), E = l.allocUnsafe(5);
    return E.writeUIntBE(482972169471, 0, 5), l.concat(
      [E, g].concat(p, d)
    );
  }
  he.psbtToBuffer = n;
  const r = (o, i) => o.key.compare(i.key);
  function c(o, i) {
    const a = /* @__PURE__ */ new Set(), u = Object.entries(o).reduce((h, [g, y]) => {
      if (g === "unknownKeyVals") return h;
      const p = i[g];
      if (p === void 0) return h;
      const d = (Array.isArray(y) ? y : [y]).map(
        p.encode
      );
      return d.map((b) => b.key.toString("hex")).forEach((b) => {
        if (a.has(b))
          throw new Error("Serialize Error: Duplicate key: " + b);
        a.add(b);
      }), h.concat(d);
    }, []), f = o.unknownKeyVals ? o.unknownKeyVals.filter((h) => !a.has(h.key.toString("hex"))) : [];
    return u.concat(f).sort(r);
  }
  function s({ globalMap: o, inputs: i, outputs: a }) {
    return {
      globalKeyVals: c(o, t.globals),
      inputKeyVals: i.map((u) => c(u, t.inputs)),
      outputKeyVals: a.map((u) => c(u, t.outputs))
    };
  }
  return he.psbtToKeyVals = s, he;
}
var pn;
function rr() {
  return pn || (pn = 1, (function(t) {
    function e(n) {
      for (var r in n) t.hasOwnProperty(r) || (t[r] = n[r]);
    }
    Object.defineProperty(t, "__esModule", { value: !0 }), e(Ds()), e(qs());
  })(st)), st;
}
var ln;
function Vs() {
  if (ln) return xe;
  ln = 1, Object.defineProperty(xe, "__esModule", { value: !0 });
  const t = rr();
  function e(s) {
    const o = s[0], i = t.psbtToKeyVals(o), a = s.slice(1);
    if (a.length === 0) throw new Error("Combine: Nothing to combine");
    const u = r(o);
    if (u === void 0)
      throw new Error("Combine: Self missing transaction");
    const f = c(i.globalKeyVals), h = i.inputKeyVals.map(c), g = i.outputKeyVals.map(c);
    for (const y of a) {
      const p = r(y);
      if (p === void 0 || !p.toBuffer().equals(u.toBuffer()))
        throw new Error(
          "Combine: One of the Psbts does not have the same transaction."
        );
      const d = t.psbtToKeyVals(y);
      c(d.globalKeyVals).forEach(
        n(
          f,
          i.globalKeyVals,
          d.globalKeyVals
        )
      ), d.inputKeyVals.map(c).forEach(
        (w, _) => w.forEach(
          n(
            h[_],
            i.inputKeyVals[_],
            d.inputKeyVals[_]
          )
        )
      ), d.outputKeyVals.map(c).forEach(
        (w, _) => w.forEach(
          n(
            g[_],
            i.outputKeyVals[_],
            d.outputKeyVals[_]
          )
        )
      );
    }
    return t.psbtFromKeyVals(u, {
      globalMapKeyVals: i.globalKeyVals,
      inputKeyVals: i.inputKeyVals,
      outputKeyVals: i.outputKeyVals
    });
  }
  xe.combine = e;
  function n(s, o, i) {
    return (a) => {
      if (s.has(a)) return;
      const u = i.filter((f) => f.key.toString("hex") === a)[0];
      o.push(u), s.add(a);
    };
  }
  function r(s) {
    return s.globalMap.unsignedTx;
  }
  function c(s) {
    const o = /* @__PURE__ */ new Set();
    return s.forEach((i) => {
      const a = i.key.toString("hex");
      if (o.has(a))
        throw new Error("Combine: KeyValue Map keys should be unique");
      o.add(a);
    }), o;
  }
  return xe;
}
var it = {}, dn;
function sr() {
  return dn || (dn = 1, (function(t) {
    Object.defineProperty(t, "__esModule", { value: !0 });
    const e = It();
    function n(y, p) {
      const d = y[p];
      if (d === void 0) throw new Error(`No input #${p}`);
      return d;
    }
    t.checkForInput = n;
    function r(y, p) {
      const d = y[p];
      if (d === void 0) throw new Error(`No output #${p}`);
      return d;
    }
    t.checkForOutput = r;
    function c(y, p, d) {
      if (y.key[0] < d)
        throw new Error(
          "Use the method for your specific key instead of addUnknownKeyVal*"
        );
      if (p && p.filter((E) => E.key.equals(y.key)).length !== 0)
        throw new Error(`Duplicate Key: ${y.key.toString("hex")}`);
    }
    t.checkHasKey = c;
    function s(y) {
      let p = 0;
      return Object.keys(y).forEach((d) => {
        Number(isNaN(Number(d))) && p++;
      }), p;
    }
    t.getEnumLength = s;
    function o(y, p) {
      let d = !1;
      if (p.nonWitnessUtxo || p.witnessUtxo) {
        const E = !!p.redeemScript, b = !!p.witnessScript, S = !E || !!p.finalScriptSig, w = !b || !!p.finalScriptWitness, _ = !!p.finalScriptSig || !!p.finalScriptWitness;
        d = S && w && _;
      }
      if (d === !1)
        throw new Error(
          `Input #${y} has too much or too little data to clean`
        );
    }
    t.inputCheckUncleanFinalized = o;
    function i(y, p, d, E) {
      throw new Error(
        `Data for ${y} key ${p} is incorrect: Expected ${d} and got ${JSON.stringify(E)}`
      );
    }
    function a(y) {
      return (p, d) => {
        for (const E of Object.keys(p)) {
          const b = p[E], { canAdd: S, canAddToArray: w, check: _, expected: I } = (
            // @ts-ignore
            e[y + "s"][E] || {}
          ), P = !!w;
          if (_)
            if (P) {
              if (!Array.isArray(b) || // @ts-ignore
              d[E] && !Array.isArray(d[E]))
                throw new Error(`Key type ${E} must be an array`);
              b.every(_) || i(y, E, I, b);
              const m = d[E] || [], k = /* @__PURE__ */ new Set();
              if (!b.every((U) => w(m, U, k)))
                throw new Error("Can not add duplicate data to array");
              d[E] = m.concat(b);
            } else {
              if (_(b) || i(y, E, I, b), !S(d, b))
                throw new Error(`Can not add duplicate data to ${y}`);
              d[E] = b;
            }
        }
      };
    }
    t.updateGlobal = a("global"), t.updateInput = a("input"), t.updateOutput = a("output");
    function u(y, p) {
      const d = y.length - 1, E = n(y, d);
      t.updateInput(p, E);
    }
    t.addInputAttributes = u;
    function f(y, p) {
      const d = y.length - 1, E = r(y, d);
      t.updateOutput(p, E);
    }
    t.addOutputAttributes = f;
    function h(y, p) {
      if (!l.isBuffer(p) || p.length < 4)
        throw new Error("Set Version: Invalid Transaction");
      return p.writeUInt32LE(y, 0), p;
    }
    t.defaultVersionSetter = h;
    function g(y, p) {
      if (!l.isBuffer(p) || p.length < 4)
        throw new Error("Set Locktime: Invalid Transaction");
      return p.writeUInt32LE(y, p.length - 4), p;
    }
    t.defaultLocktimeSetter = g;
  })(it)), it;
}
var gn;
function js() {
  if (gn) return Ae;
  gn = 1, Object.defineProperty(Ae, "__esModule", { value: !0 });
  const t = Vs(), e = rr(), n = A(), r = sr();
  class c {
    constructor(o) {
      this.inputs = [], this.outputs = [], this.globalMap = {
        unsignedTx: o
      };
    }
    static fromBase64(o, i) {
      const a = l.from(o, "base64");
      return this.fromBuffer(a, i);
    }
    static fromHex(o, i) {
      const a = l.from(o, "hex");
      return this.fromBuffer(a, i);
    }
    static fromBuffer(o, i) {
      const a = e.psbtFromBuffer(o, i), u = new this(a.globalMap.unsignedTx);
      return Object.assign(u, a), u;
    }
    toBase64() {
      return this.toBuffer().toString("base64");
    }
    toHex() {
      return this.toBuffer().toString("hex");
    }
    toBuffer() {
      return e.psbtToBuffer(this);
    }
    updateGlobal(o) {
      return r.updateGlobal(o, this.globalMap), this;
    }
    updateInput(o, i) {
      const a = r.checkForInput(this.inputs, o);
      return r.updateInput(i, a), this;
    }
    updateOutput(o, i) {
      const a = r.checkForOutput(this.outputs, o);
      return r.updateOutput(i, a), this;
    }
    addUnknownKeyValToGlobal(o) {
      return r.checkHasKey(
        o,
        this.globalMap.unknownKeyVals,
        r.getEnumLength(n.GlobalTypes)
      ), this.globalMap.unknownKeyVals || (this.globalMap.unknownKeyVals = []), this.globalMap.unknownKeyVals.push(o), this;
    }
    addUnknownKeyValToInput(o, i) {
      const a = r.checkForInput(this.inputs, o);
      return r.checkHasKey(
        i,
        a.unknownKeyVals,
        r.getEnumLength(n.InputTypes)
      ), a.unknownKeyVals || (a.unknownKeyVals = []), a.unknownKeyVals.push(i), this;
    }
    addUnknownKeyValToOutput(o, i) {
      const a = r.checkForOutput(this.outputs, o);
      return r.checkHasKey(
        i,
        a.unknownKeyVals,
        r.getEnumLength(n.OutputTypes)
      ), a.unknownKeyVals || (a.unknownKeyVals = []), a.unknownKeyVals.push(i), this;
    }
    addInput(o) {
      this.globalMap.unsignedTx.addInput(o), this.inputs.push({
        unknownKeyVals: []
      });
      const i = o.unknownKeyVals || [], a = this.inputs.length - 1;
      if (!Array.isArray(i))
        throw new Error("unknownKeyVals must be an Array");
      return i.forEach(
        (u) => this.addUnknownKeyValToInput(a, u)
      ), r.addInputAttributes(this.inputs, o), this;
    }
    addOutput(o) {
      this.globalMap.unsignedTx.addOutput(o), this.outputs.push({
        unknownKeyVals: []
      });
      const i = o.unknownKeyVals || [], a = this.outputs.length - 1;
      if (!Array.isArray(i))
        throw new Error("unknownKeyVals must be an Array");
      return i.forEach(
        (u) => this.addUnknownKeyValToOutput(a, u)
      ), r.addOutputAttributes(this.outputs, o), this;
    }
    clearFinalizedInput(o) {
      const i = r.checkForInput(this.inputs, o);
      r.inputCheckUncleanFinalized(o, i);
      for (const a of Object.keys(i))
        [
          "witnessUtxo",
          "nonWitnessUtxo",
          "finalScriptSig",
          "finalScriptWitness",
          "unknownKeyVals"
        ].includes(a) || delete i[a];
      return this;
    }
    combine(...o) {
      const i = t.combine([this].concat(o));
      return Object.assign(this, i), this;
    }
    getTransaction() {
      return this.globalMap.unsignedTx.toBuffer();
    }
  }
  return Ae.Psbt = c, Ae;
}
var yn = js(), dt = fe(), x = sr();
function ee(t) {
  return (e) => {
    try {
      return t({ output: e }), !0;
    } catch {
      return !1;
    }
  };
}
const zs = ee(De), Ys = ee(Pn), $s = ee(we), $e = ee(Xe), Sn = ee(be), or = ee(Ee), vt = ee(Te), Xo = ee(On), Do = (t) => t.length === 4 && t[0] === 81 && // OP_1
t[1] === 2 && // push 2 bytes
t[2] === 78 && t[3] === 115;
function ye(t) {
  let e = l.allocUnsafe(0);
  function n(o) {
    e = l.concat([e, l.from(o)]);
  }
  function r(o) {
    const i = e.length, a = dt.encodingLength(o);
    e = l.concat([e, l.allocUnsafe(a)]), dt.encode(o, e, i);
  }
  function c(o) {
    r(o.length), n(o);
  }
  function s(o) {
    r(o.length), o.forEach(c);
  }
  return s(t), e;
}
function ir(t, e) {
  const n = ue(e);
  if (n === null) throw new Error("Unknown script error");
  const r = et(t), c = ge(t), s = _s(t), o = s?.hybrid ? et(s.hybrid) : void 0, i = s?.uncompressed ? et(s.uncompressed) : void 0;
  return n.findIndex((a) => {
    if (typeof a == "number") return !1;
    if (ke(a, t) || ke(a, c) || a.equals(r) || s && (ke(a, s.uncompressed) || ke(a, s.hybrid) || o && a.equals(o) || i && a.equals(i)))
      return !0;
  });
}
function Ze(t, e) {
  return ir(t, e) !== -1;
}
function Zs(t, e) {
  return Js(t).some((r) => cr(r, de.decode, e));
}
function cr(t, e, n) {
  const { hashType: r } = e(t), c = [];
  switch (r & v.SIGHASH_ANYONECANPAY && c.push("addInput"), r & 31) {
    case v.SIGHASH_ALL:
      break;
    case v.SIGHASH_SINGLE:
    case v.SIGHASH_NONE:
      c.push("addOutput"), c.push("setInputSequence");
      break;
  }
  return c.indexOf(n) === -1;
}
function Js(t) {
  const { partialSig: e } = t;
  let n;
  if (!e || e.length === 0) {
    if (!t.finalScriptSig && !t.finalScriptWitness) return [];
    n = Qs(t);
  } else
    n = e;
  return n.map((r) => r.signature);
}
function Qs(t) {
  const e = t.finalScriptSig ? ue(t.finalScriptSig) || [] : [], n = t.finalScriptWitness ? ue(t.finalScriptWitness) || [] : [];
  return e.concat(n).filter((r) => l.isBuffer(r) && Un(r)).map((r) => ({ signature: r }));
}
function wn(t, e, n) {
  const r = lo(e, t, n);
  try {
    const s = ho(e, r).concat(r.script).concat(r.controlBlock);
    return { finalScriptWitness: ye(s) };
  } catch (c) {
    throw new Error(`Can not finalize taproot input #${t}: ${c}`);
  }
}
function Re(t, e) {
  const n = e ? l.from([e]) : l.from([]);
  return l.concat([t, n]);
}
function N(t) {
  return t && !!(t.tapInternalKey || t.tapMerkleRoot || t.tapLeafScript && t.tapLeafScript.length || t.tapBip32Derivation && t.tapBip32Derivation.length || t.witnessUtxo && vt(t.witnessUtxo.script));
}
function ct(t, e) {
  return t && !!(t.tapInternalKey || t.tapTree || t.tapBip32Derivation && t.tapBip32Derivation.length || e && vt(e));
}
function En(t, e, n) {
  ao(t, e, n), fo(t, e, n);
}
function bn(t, e, n) {
  uo(t, e, n), eo(t, e);
}
function eo(t, e) {
  if (!e.tapTree && !e.tapInternalKey) return;
  const n = e.tapInternalKey || t.tapInternalKey, r = e.tapTree || t.tapTree;
  if (n) {
    const c = t.script, s = to(n, r);
    if (c && !c.equals(s))
      throw new Error("Error adding output. Script or address missmatch.");
  }
}
function to(t, e) {
  const n = e && no(e.leaves), { output: r } = Te({
    internalPubkey: t,
    scriptTree: n
  });
  if (!r) throw new Error("Failed to generate taproot script pubkey");
  return r;
}
function qo(t, e) {
  const n = e.tapInternalKey, r = n && Tr(n, e.tapMerkleRoot);
  if (!r)
    throw new Error(
      `Cannot tweak tap internal key for input #${t}. Public key: ${n && n.toString("hex")}`
    );
  return r.x;
}
function Vo(t) {
  if (!Pr(t))
    throw new Error("Cannot convert taptree to tapleaf list. Expecting a tapree structure.");
  return gt(t);
}
function no(t = []) {
  return t.length === 1 && t[0].depth === 0 ? {
    output: t[0].script,
    version: t[0].leafVersion
  } : co(t);
}
function ro(t, e) {
  return oo(t).some((r) => cr(r, so, e));
}
function so(t) {
  return {
    signature: t.subarray(0, 64),
    hashType: t.subarray(64)[0] || v.SIGHASH_DEFAULT
  };
}
function oo(t) {
  const e = [];
  if (t.tapKeySig && e.push(t.tapKeySig), t.tapScriptSig && e.push(...t.tapScriptSig.map((n) => n.signature)), !e.length) {
    const n = io(t.finalScriptWitness);
    n && e.push(n);
  }
  return e;
}
function io(t) {
  if (!t) return;
  const e = t.subarray(2);
  if (e.length === 64 || e.length === 65) return e;
}
function gt(t, e = [], n = 0) {
  if (n > Nn) throw new Error("Max taptree depth exceeded.");
  return t ? Hn(t) ? (e.push({
    depth: n,
    leafVersion: t.version || _r,
    script: t.output
  }), e) : (t[0] && gt(t[0], e, n + 1), t[1] && gt(t[1], e, n + 1), e) : [];
}
function co(t) {
  let e;
  for (const n of t)
    if (e = yt(n, e), !e) throw new Error("No room left to insert tapleaf in tree");
  return e;
}
function yt(t, e, n = 0) {
  if (n > Nn) throw new Error("Max taptree depth exceeded.");
  if (t.depth === n)
    return e ? void 0 : {
      output: t.script,
      version: t.leafVersion
    };
  if (Hn(e)) return;
  const r = yt(t, e && e[0], n + 1);
  if (r) return [r, e && e[1]];
  const c = yt(t, e && e[1], n + 1);
  if (c) return [e && e[0], c];
}
function ao(t, e, n) {
  const r = N(t) && ae(e), c = ae(t) && N(e), s = t === e && N(e) && ae(e);
  if (r || c || s)
    throw new Error(
      `Invalid arguments for Psbt.${n}. Cannot use both taproot and non-taproot fields.`
    );
}
function uo(t, e, n) {
  const r = ct(t) && ae(e), c = ae(t) && ct(e), s = t === e && ct(e) && ae(e);
  if (r || c || s)
    throw new Error(
      `Invalid arguments for Psbt.${n}. Cannot use both taproot and non-taproot fields.`
    );
}
function fo(t, e, n) {
  if (e.tapMerkleRoot) {
    const r = (e.tapLeafScript || []).every(
      (s) => at(s, e.tapMerkleRoot)
    ), c = (t.tapLeafScript || []).every(
      (s) => at(s, e.tapMerkleRoot)
    );
    if (!r || !c)
      throw new Error(`Invalid arguments for Psbt.${n}. Tapleaf not part of taptree.`);
  } else if (t.tapMerkleRoot && !(e.tapLeafScript || []).every(
    (c) => at(c, t.tapMerkleRoot)
  ))
    throw new Error(`Invalid arguments for Psbt.${n}. Tapleaf not part of taptree.`);
}
function at(t, e) {
  if (!e) return !0;
  const n = qe({
    output: t.script,
    version: t.leafVersion
  });
  return mr(t.controlBlock, n).equals(e);
}
function ho(t, e) {
  const n = qe({
    output: e.script,
    version: e.leafVersion
  });
  return (t.tapScriptSig || []).filter((r) => r.leafHash.equals(n)).map((r) => po(e.script, r)).sort((r, c) => c.positionInScript - r.positionInScript).map((r) => r.signature);
}
function po(t, e) {
  return Object.assign(
    {
      positionInScript: ir(e.pubkey, t)
    },
    e
  );
}
function lo(t, e, n) {
  const { tapScriptSig: r } = t;
  if (!r || !r.length)
    throw new Error(
      `Can not finalize taproot input #${e}. No tapleaf script signature provided.`
    );
  const c = (t.tapLeafScript || []).sort((s, o) => s.controlBlock.length - o.controlBlock.length).find((s) => go(s, r, n));
  if (!c)
    throw new Error(
      `Can not finalize taproot input #${e}. Signature for tapleaf script not found.`
    );
  return c;
}
function go(t, e, n) {
  const r = qe({
    output: t.script,
    version: t.leafVersion
  });
  return (!n || n.equals(r)) && e.find((s) => s.leafHash.equals(r)) !== void 0;
}
function ae(t) {
  return t && !!(t.redeemScript || t.witnessScript || t.bip32Derivation && t.bip32Derivation.length);
}
const Tn = dt.decode, yo = {
  /**
   * A bitcoinjs Network object. This is only used if you pass an `address`
   * parameter to addOutput. Otherwise it is not needed and can be left default.
   */
  network: Ve,
  /**
   * When extractTransaction is called, the fee rate is checked.
   * THIS IS NOT TO BE RELIED ON.
   * It is only here as a last ditch effort to prevent sending a 500 BTC fee etc.
   */
  maximumFeeRate: 5e3
  // satoshi per byte
};
class St {
  constructor(e = {}, n = new yn.Psbt(new ar())) {
    this.data = n, this.opts = Object.assign({}, yo, e), this.__CACHE = {
      __NON_WITNESS_UTXO_TX_CACHE: [],
      __NON_WITNESS_UTXO_BUF_CACHE: [],
      __TX_IN_CACHE: {},
      // unsignedTx.tx property is dynamically added by PsbtBase
      __TX: this.data.globalMap.unsignedTx.tx,
      __UNSAFE_SIGN_NONSEGWIT: !1
    }, e.version === 3 ? this.setVersionTRUC() : this.data.inputs.length === 0 && this.setVersion(2);
    const r = (c, s, o, i) => {
      Object.defineProperty(c, s, {
        enumerable: o,
        writable: i
      });
    };
    r(this, "__CACHE", !1, !0), r(this, "opts", !1, !0);
  }
  get inputCount() {
    return this.data.inputs.length;
  }
  get version() {
    return this.__CACHE.__TX.version;
  }
  set version(e) {
    this.setVersion(e);
  }
  get locktime() {
    return this.__CACHE.__TX.locktime;
  }
  set locktime(e) {
    this.setLocktime(e);
  }
  get txInputs() {
    return this.__CACHE.__TX.ins.map((e) => ({
      hash: At(e.hash),
      index: e.index,
      sequence: e.sequence
    }));
  }
  get txOutputs() {
    return this.__CACHE.__TX.outs.map((e) => {
      let n;
      try {
        n = er(e.script, this.opts.network);
      } catch {
      }
      return {
        script: At(e.script),
        value: e.value,
        address: n
      };
    });
  }
  static fromBase64(e, n = {}) {
    const r = l.from(e, "base64");
    return this.fromBuffer(r, n);
  }
  static fromHex(e, n = {}) {
    const r = l.from(e, "hex");
    return this.fromBuffer(r, n);
  }
  static fromBuffer(e, n = {}) {
    const r = yn.Psbt.fromBuffer(e, So), c = new St(n, r);
    return _o(c.__CACHE.__TX, c.__CACHE), c;
  }
  combine(...e) {
    return this.data.combine(...e.map((n) => n.data)), this;
  }
  clone() {
    const e = JSON.parse(JSON.stringify(this.opts));
    return St.fromBuffer(this.data.toBuffer(), e);
  }
  setMaximumFeeRate(e) {
    Fe(e), this.opts.maximumFeeRate = e;
  }
  setVersion(e) {
    Fe(e), pe(this.data.inputs, "setVersion");
    const n = this.__CACHE;
    return n.__TX.version = e, n.__EXTRACTED_TX = void 0, this;
  }
  setVersionTRUC() {
    return this.setVersion(v.TRUC_VERSION);
  }
  setLocktime(e) {
    Fe(e), pe(this.data.inputs, "setLocktime");
    const n = this.__CACHE;
    return n.__TX.locktime = e, n.__EXTRACTED_TX = void 0, this;
  }
  setInputSequence(e, n) {
    Fe(n), pe(this.data.inputs, "setInputSequence");
    const r = this.__CACHE;
    if (r.__TX.ins.length <= e)
      throw new Error("Input index too high");
    return r.__TX.ins[e].sequence = n, r.__EXTRACTED_TX = void 0, this;
  }
  addInputs(e, n = !0) {
    return e.forEach((r) => this.addInput(r, n)), this;
  }
  addInput(e, n = !0) {
    if (!e || e.hash === void 0 || e.index === void 0)
      throw new Error(
        "Invalid arguments for Psbt.addInput. Requires single object with at least [hash] and [index]"
      );
    En(e, e, "addInput"), n && pe(this.data.inputs, "addInput"), e.witnessScript && Ge(e.witnessScript);
    const r = this.__CACHE;
    this.data.addInput(e);
    const c = r.__TX.ins[r.__TX.ins.length - 1];
    hr(r, c);
    const s = this.data.inputs.length - 1, o = this.data.inputs[s];
    return o.nonWitnessUtxo && Et(this.__CACHE, o, s), r.__FEE = void 0, r.__FEE_RATE = void 0, r.__EXTRACTED_TX = void 0, this;
  }
  addOutputs(e) {
    return e.forEach((n) => this.addOutput(n)), this;
  }
  addOutput(e) {
    const n = "address" in e, r = "script" in e;
    if (arguments.length > 1 || !e || e.value === void 0 || !n && !r)
      throw new Error(
        "Invalid arguments for Psbt.addOutput. Requires single object with at least [script or address] and [value]"
      );
    if (pe(this.data.inputs, "addOutput"), n) {
      const { address: s } = e, { network: o } = this.opts, i = tr(s, o);
      e = Object.assign({}, e, { script: i });
    }
    bn(e, e, "addOutput");
    const c = this.__CACHE;
    return this.data.addOutput(e), c.__FEE = void 0, c.__FEE_RATE = void 0, c.__EXTRACTED_TX = void 0, this;
  }
  extractTransaction(e, n) {
    if (n && (this.data.inputs = this.data.inputs.filter((s) => !s.partialSig)), !this.data.inputs.every(ur)) throw new Error("Not finalized");
    const r = this.__CACHE;
    if (e || Eo(this, r, this.opts), r.__EXTRACTED_TX) return r.__EXTRACTED_TX;
    const c = r.__TX.clone();
    return Sr(this.data.inputs, c, r, !0, n), c;
  }
  getFeeRate(e = !1) {
    return kn(
      "__FEE_RATE",
      "fee rate",
      this.data.inputs,
      this.__CACHE,
      e
    );
  }
  getFee(e = !1) {
    return kn("__FEE", "fee", this.data.inputs, this.__CACHE, e);
  }
  finalizeAllInputs() {
    return x.checkForInput(this.data.inputs, 0), le(this.data.inputs.length).forEach((e) => this.finalizeInput(e)), this;
  }
  finalizeInput(e, n, r) {
    const c = x.checkForInput(this.data.inputs, e);
    return N(c) ? this._finalizeTaprootInput(
      e,
      c,
      void 0,
      n
    ) : this._finalizeInput(
      e,
      c,
      n,
      r ?? !0
    );
  }
  finalizeTaprootInput(e, n, r = wn) {
    const c = x.checkForInput(this.data.inputs, e);
    if (N(c))
      return this._finalizeTaprootInput(
        e,
        c,
        n,
        r
      );
    throw new Error(`Cannot finalize input #${e}. Not Taproot.`);
  }
  getInputType(e) {
    const n = x.checkForInput(this.data.inputs, e), r = wr(e, n, this.__CACHE), c = Qe(
      r,
      e,
      "input",
      n.redeemScript || No(n.finalScriptSig),
      n.witnessScript || Uo(n.finalScriptWitness)
    ), s = c.type === "raw" ? "" : c.type + "-", o = br(c.meaningfulScript);
    return s + o;
  }
  inputHasPubkey(e, n) {
    const r = x.checkForInput(this.data.inputs, e);
    return Oo(n, r, e, this.__CACHE);
  }
  inputHasHDKey(e, n) {
    const r = x.checkForInput(this.data.inputs, e), c = mn(n);
    return !!r.bip32Derivation && r.bip32Derivation.some(c);
  }
  outputHasPubkey(e, n) {
    const r = x.checkForOutput(this.data.outputs, e);
    return Po(n, r, e, this.__CACHE);
  }
  outputHasHDKey(e, n) {
    const r = x.checkForOutput(this.data.outputs, e), c = mn(n);
    return !!r.bip32Derivation && r.bip32Derivation.some(c);
  }
  validateSignaturesOfAllInputs(e) {
    return x.checkForInput(this.data.inputs, 0), le(this.data.inputs.length).map(
      (r) => this.validateSignaturesOfInput(r, e)
    ).reduce((r, c) => c && r, !0);
  }
  validateSignaturesOfInput(e, n, r) {
    const c = this.data.inputs[e];
    return N(c) ? this.validateSignaturesOfTaprootInput(e, n, r) : this._validateSignaturesOfInput(e, n, r);
  }
  signAllInputsHD(e, n = [v.SIGHASH_ALL]) {
    if (!e || !e.publicKey || !e.fingerprint)
      throw new Error("Need HDSigner to sign input");
    const r = [];
    for (const c of le(this.data.inputs.length))
      try {
        this.signInputHD(c, e, n), r.push(!0);
      } catch {
        r.push(!1);
      }
    if (r.every((c) => !c))
      throw new Error("No inputs were signed");
    return this;
  }
  signAllInputsHDAsync(e, n = [v.SIGHASH_ALL]) {
    return new Promise((r, c) => {
      if (!e || !e.publicKey || !e.fingerprint)
        return c(new Error("Need HDSigner to sign input"));
      const s = [], o = [];
      for (const i of le(this.data.inputs.length))
        o.push(
          this.signInputHDAsync(i, e, n).then(
            () => {
              s.push(!0);
            },
            () => {
              s.push(!1);
            }
          )
        );
      return Promise.all(o).then(() => {
        if (s.every((i) => !i))
          return c(new Error("No inputs were signed"));
        r();
      });
    });
  }
  signInputHD(e, n, r = [v.SIGHASH_ALL]) {
    if (!n || !n.publicKey || !n.fingerprint)
      throw new Error("Need HDSigner to sign input");
    return Cn(e, this.data.inputs, n).forEach((s) => this.signInput(e, s, r)), this;
  }
  signInputHDAsync(e, n, r = [v.SIGHASH_ALL]) {
    return new Promise((c, s) => {
      if (!n || !n.publicKey || !n.fingerprint)
        return s(new Error("Need HDSigner to sign input"));
      const i = Cn(e, this.data.inputs, n).map(
        (a) => this.signInputAsync(e, a, r)
      );
      return Promise.all(i).then(() => {
        c();
      }).catch(s);
    });
  }
  signAllInputs(e, n) {
    if (!e || !e.publicKey) throw new Error("Need Signer to sign input");
    const r = [];
    for (const c of le(this.data.inputs.length))
      try {
        this.signInput(c, e, n), r.push(!0);
      } catch {
        r.push(!1);
      }
    if (r.every((c) => !c))
      throw new Error("No inputs were signed");
    return this;
  }
  signAllInputsAsync(e, n) {
    return new Promise((r, c) => {
      if (!e || !e.publicKey)
        return c(new Error("Need Signer to sign input"));
      const s = [], o = [];
      for (const [i] of this.data.inputs.entries())
        o.push(
          this.signInputAsync(i, e, n).then(
            () => {
              s.push(!0);
            },
            () => {
              s.push(!1);
            }
          )
        );
      return Promise.all(o).then(() => {
        if (s.every((i) => !i))
          return c(new Error("No inputs were signed"));
        r();
      });
    });
  }
  signInput(e, n, r) {
    if (!n || !n.publicKey)
      throw new Error("Need Signer to sign input");
    const c = x.checkForInput(this.data.inputs, e);
    return N(c) ? this._signTaprootInput(e, c, n, void 0, r) : this._signInput(e, n, r);
  }
  signTaprootInput(e, n, r, c) {
    if (!n || !n.publicKey)
      throw new Error("Need Signer to sign input");
    const s = x.checkForInput(this.data.inputs, e);
    if (N(s))
      return this._signTaprootInput(
        e,
        s,
        n,
        r,
        c
      );
    throw new Error(`Input #${e} is not of type Taproot.`);
  }
  signInputAsync(e, n, r) {
    return Promise.resolve().then(() => {
      if (!n || !n.publicKey) throw new Error("Need Signer to sign input");
      const c = x.checkForInput(this.data.inputs, e);
      return N(c) ? this._signTaprootInputAsync(
        e,
        c,
        n,
        void 0,
        r
      ) : this._signInputAsync(e, n, r);
    });
  }
  signTaprootInputAsync(e, n, r, c) {
    return Promise.resolve().then(() => {
      if (!n || !n.publicKey) throw new Error("Need Signer to sign input");
      const s = x.checkForInput(this.data.inputs, e);
      if (N(s))
        return this._signTaprootInputAsync(
          e,
          s,
          n,
          r,
          c
        );
      throw new Error(`Input #${e} is not of type Taproot.`);
    });
  }
  toBuffer() {
    return ut(this.__CACHE), this.data.toBuffer();
  }
  toHex() {
    return ut(this.__CACHE), this.data.toHex();
  }
  toBase64() {
    return ut(this.__CACHE), this.data.toBase64();
  }
  updateGlobal(e) {
    return this.data.updateGlobal(e), this;
  }
  updateInput(e, n) {
    return n.witnessScript && Ge(n.witnessScript), En(this.data.inputs[e], n, "updateInput"), this.data.updateInput(e, n), n.nonWitnessUtxo && Et(this.__CACHE, this.data.inputs[e], e), this;
  }
  updateOutput(e, n) {
    const r = this.data.outputs[e];
    return bn(r, n, "updateOutput"), this.data.updateOutput(e, n), this;
  }
  addUnknownKeyValToGlobal(e) {
    return this.data.addUnknownKeyValToGlobal(e), this;
  }
  addUnknownKeyValToInput(e, n) {
    return this.data.addUnknownKeyValToInput(e, n), this;
  }
  addUnknownKeyValToOutput(e, n) {
    return this.data.addUnknownKeyValToOutput(e, n), this;
  }
  clearFinalizedInput(e) {
    return this.data.clearFinalizedInput(e), this;
  }
  checkTaprootHashesForSig(e, n, r, c, s) {
    if (typeof r.signSchnorr != "function")
      throw new Error(`Need Schnorr Signer to sign taproot input #${e}.`);
    const o = l.isBuffer(r.publicKey) ? r.publicKey : l.from(r.publicKey), i = wt(
      e,
      n,
      this.data.inputs,
      o,
      this.__CACHE,
      c,
      s
    );
    if (!i || !i.length)
      throw new Error(
        `Can not sign for input #${e} with the key ${o.toString("hex")}`
      );
    return i;
  }
  _finalizeInput(e, n, r = mo, c = !0) {
    const { script: s, isP2SH: o, isP2WSH: i, isSegwit: a } = Ao(
      e,
      n,
      this.__CACHE
    );
    if (!s) throw new Error(`No script found for input #${e}`);
    bo(n);
    const { finalScriptSig: u, finalScriptWitness: f } = r(
      e,
      n,
      s,
      a,
      o,
      i,
      c
    );
    if (u && this.data.updateInput(e, { finalScriptSig: u }), f && this.data.updateInput(e, { finalScriptWitness: f }), !u && !f)
      throw new Error(`Unknown error finalizing input #${e}`);
    return this.data.clearFinalizedInput(e), this;
  }
  _finalizeTaprootInput(e, n, r, c = wn) {
    if (!n.witnessUtxo)
      throw new Error(`Cannot finalize input #${e}. Missing witness utxo.`);
    if (n.tapKeySig) {
      const s = Te({
        output: n.witnessUtxo.script,
        signature: n.tapKeySig
      });
      if (!s.witness) throw new Error("Cannot finalize taproot key spend");
      const o = ye(s.witness);
      this.data.updateInput(e, { finalScriptWitness: o });
    } else {
      const { finalScriptWitness: s } = c(
        e,
        n,
        r
      );
      this.data.updateInput(e, { finalScriptWitness: s });
    }
    return this.data.clearFinalizedInput(e), this;
  }
  _validateSignaturesOfInput(e, n, r) {
    const c = this.data.inputs[e], s = (c || {}).partialSig;
    if (!c || !s || s.length < 1)
      throw new Error("No signatures to validate");
    if (typeof n != "function")
      throw new Error("Need validator function to validate signatures");
    const o = r ? s.filter((h) => h.pubkey.equals(r)) : s;
    if (o.length < 1) throw new Error("No signatures for this pubkey");
    const i = [];
    let a, u, f;
    for (const h of o) {
      const g = de.decode(h.signature), { hash: y, script: p } = f !== g.hashType || !a || !u ? lr(
        e,
        Object.assign({}, c, {
          sighashType: g.hashType
        }),
        this.__CACHE,
        !0
      ) : { hash: a, script: u };
      f = g.hashType, a = y, u = p, fr(h.pubkey, p, "verify"), i.push(n(h.pubkey, y, g.signature));
    }
    return i.every((h) => h);
  }
  validateSignaturesOfTaprootInput(e, n, r) {
    const c = this.data.inputs[e], s = (c || {}).tapKeySig, o = (c || {}).tapScriptSig;
    if (!c && !s && !(o && !o.length))
      throw new Error("No signatures to validate");
    if (typeof n != "function")
      throw new Error("Need validator function to validate signatures");
    r = r && ge(r);
    const i = r ? wt(e, c, this.data.inputs, r, this.__CACHE) : vo(e, c, this.data.inputs, this.__CACHE);
    if (!i.length) throw new Error("No signatures for this pubkey");
    const a = i.find((f) => !f.leafHash);
    let u = 0;
    if (s && a) {
      if (!n(
        a.pubkey,
        a.hash,
        xn(s)
      )) return !1;
      u++;
    }
    if (o)
      for (const f of o) {
        const h = i.find((g) => f.pubkey.equals(g.pubkey));
        if (h) {
          if (!n(
            f.pubkey,
            h.hash,
            xn(f.signature)
          )) return !1;
          u++;
        }
      }
    return u > 0;
  }
  _signInput(e, n, r = [v.SIGHASH_ALL]) {
    const c = l.isBuffer(n.publicKey) ? n.publicKey : l.from(n.publicKey), { hash: s, sighashType: o } = An(
      this.data.inputs,
      e,
      c,
      this.__CACHE,
      r
    ), i = n.sign(s), a = [
      {
        pubkey: c,
        signature: de.encode(
          l.isBuffer(i) ? i : l.from(i),
          o
        )
      }
    ];
    return this.data.updateInput(e, { partialSig: a }), this;
  }
  _signTaprootInput(e, n, r, c, s = [v.SIGHASH_DEFAULT]) {
    const o = l.isBuffer(r.publicKey) ? r.publicKey : l.from(r.publicKey), i = this.checkTaprootHashesForSig(
      e,
      n,
      r,
      c,
      s
    ), a = r.signSchnorr.bind(r), u = (g) => l.isBuffer(g) ? g : l.from(g), f = i.filter((g) => !g.leafHash).map(
      (g) => Re(u(a(g.hash)), n.sighashType)
    )[0], h = i.filter((g) => !!g.leafHash).map(
      (g) => ({
        pubkey: ge(o),
        signature: Re(
          u(a(g.hash)),
          n.sighashType
        ),
        leafHash: g.leafHash
      })
    );
    return f && this.data.updateInput(e, { tapKeySig: f }), h.length && this.data.updateInput(e, { tapScriptSig: h }), this;
  }
  _signInputAsync(e, n, r = [v.SIGHASH_ALL]) {
    const c = l.isBuffer(n.publicKey) ? n.publicKey : l.from(n.publicKey), { hash: s, sighashType: o } = An(
      this.data.inputs,
      e,
      c,
      this.__CACHE,
      r
    );
    return Promise.resolve(n.sign(s)).then((i) => {
      const a = l.isBuffer(i) ? i : l.from(i), u = [
        {
          pubkey: c,
          signature: de.encode(a, o)
        }
      ];
      this.data.updateInput(e, { partialSig: u });
    });
  }
  async _signTaprootInputAsync(e, n, r, c, s = [v.SIGHASH_DEFAULT]) {
    const o = l.isBuffer(r.publicKey) ? r.publicKey : l.from(r.publicKey), i = this.checkTaprootHashesForSig(
      e,
      n,
      r,
      c,
      s
    ), a = r.signSchnorr.bind(r), u = (p) => l.isBuffer(p) ? p : l.from(p), f = [], h = i.filter((p) => !p.leafHash)[0];
    if (h) {
      const p = Promise.resolve(a(h.hash)).then((d) => ({
        tapKeySig: Re(u(d), n.sighashType)
      }));
      f.push(p);
    }
    const g = i.filter((p) => !!p.leafHash);
    if (g.length) {
      const p = g.map(async (d) => {
        const E = await a(d.hash);
        return { tapScriptSig: [
          {
            pubkey: ge(o),
            signature: Re(
              u(E),
              n.sighashType
            ),
            leafHash: d.leafHash
          }
        ] };
      });
      f.push(...p);
    }
    const y = await Promise.all(f);
    for (const p of y)
      this.data.updateInput(e, p);
  }
}
const So = (t) => new ar(t);
class ar {
  constructor(e = l.from([2, 0, 0, 0, 0, 0, 0, 0, 0, 0])) {
    this.tx = v.fromBuffer(e), To(this.tx), Object.defineProperty(this, "tx", {
      enumerable: !1,
      writable: !0
    });
  }
  getInputOutputCounts() {
    return {
      inputCount: this.tx.ins.length,
      outputCount: this.tx.outs.length
    };
  }
  addInput(e) {
    if (e.hash === void 0 || e.index === void 0 || !l.isBuffer(e.hash) && typeof e.hash != "string" || typeof e.index != "number")
      throw new Error("Error adding input.");
    const n = typeof e.hash == "string" ? Bn(l.from(e.hash, "hex")) : e.hash;
    this.tx.addInput(n, e.index, e.sequence);
  }
  addOutput(e) {
    if (e.script === void 0 || e.value === void 0 || !l.isBuffer(e.script) || typeof e.value != "number")
      throw new Error("Error adding output.");
    this.tx.addOutput(e.script, e.value);
  }
  toBuffer() {
    return this.tx.toBuffer();
  }
}
function wo(t, e, n) {
  switch (n) {
    case "pubkey":
    case "pubkeyhash":
    case "witnesspubkeyhash":
      return _n(1, t.partialSig);
    case "multisig": {
      const r = De({
        output: e
      });
      return _n(r.m, t.partialSig, r.pubkeys);
    }
    case "nonstandard":
      return !0;
    default:
      return !1;
  }
}
function ut(t) {
  if (t.__UNSAFE_SIGN_NONSEGWIT)
    throw new Error("Not BIP174 compliant, can not export");
}
function _n(t, e, n) {
  if (!e) return !1;
  let r;
  if (n ? r = n.map((c) => {
    const s = Ho(c);
    return e.find((o) => o.pubkey.equals(s));
  }).filter((c) => !!c) : r = e, r.length > t) throw new Error("Too many signatures");
  return r.length === t;
}
function ur(t) {
  return !!t.finalScriptSig || !!t.finalScriptWitness;
}
function mn(t) {
  return (e) => {
    const n = l.isBuffer(t.fingerprint) ? t.fingerprint : l.from(t.fingerprint);
    if (!e.masterFingerprint.equals(n)) return !1;
    const r = t.derivePath(e.path).publicKey;
    return !!(l.isBuffer(r) ? r : l.from(r)).equals(e.pubkey);
  };
}
function Fe(t) {
  if (typeof t != "number" || t !== Math.floor(t) || t > 4294967295 || t < 0)
    throw new Error("Invalid 32 bit integer");
}
function Eo(t, e, n) {
  const r = e.__FEE_RATE || t.getFeeRate(), c = e.__EXTRACTED_TX.virtualSize(), s = r * c;
  if (r >= n.maximumFeeRate)
    throw new Error(
      `Warning: You are paying around ${(s / 1e8).toFixed(8)} in fees, which is ${r} satoshi per byte for a transaction with a VSize of ${c} bytes (segwit counted as 0.25 byte per byte). Use setMaximumFeeRate method to raise your threshold, or pass true to the first arg of extractTransaction.`
    );
}
function pe(t, e) {
  t.forEach((n) => {
    if (N(n) ? ro(n, e) : Zs(n, e)) throw new Error("Can not modify transaction, signatures exist.");
  });
}
function bo(t) {
  if (!t.sighashType || !t.partialSig) return;
  const { partialSig: e, sighashType: n } = t;
  e.forEach((r) => {
    const { hashType: c } = de.decode(r.signature);
    if (n !== c)
      throw new Error("Signature sighash does not match input sighash type");
  });
}
function fr(t, e, n) {
  if (!Ze(t, e))
    throw new Error(`Can not ${n} for this input with the key ${t.toString("hex")}`);
}
function To(t) {
  if (!t.ins.every(
    (n) => n.script && n.script.length === 0 && n.witness && n.witness.length === 0
  ))
    throw new Error("Format Error: Transaction ScriptSigs are not empty");
}
function _o(t, e) {
  t.ins.forEach((n) => {
    hr(e, n);
  });
}
function hr(t, e) {
  const n = `${Bn(l.from(e.hash)).toString("hex")}:${e.index}`;
  if (t.__TX_IN_CACHE[n]) throw new Error("Duplicate input detected.");
  t.__TX_IN_CACHE[n] = 1;
}
function pr(t, e) {
  return (n, r, c, s) => {
    const o = t({
      redeem: { output: c }
    }).output;
    if (!r.equals(o))
      throw new Error(
        `${e} for ${s} #${n} doesn't match the scriptPubKey in the prevout`
      );
  };
}
const In = pr(Ee, "Redeem script"), vn = pr(be, "Witness script");
function kn(t, e, n, r, c = !1) {
  if (!n.every(ur)) throw new Error(`PSBT must be finalized to calculate ${e}`);
  if (t === "__FEE_RATE" && r.__FEE_RATE) return r.__FEE_RATE;
  if (t === "__FEE" && r.__FEE) return r.__FEE;
  let s, o = !0;
  r.__EXTRACTED_TX ? (s = r.__EXTRACTED_TX, o = !1) : s = r.__TX.clone(), Sr(n, s, r, o, c);
  const i = t === "__FEE_RATE" ? r.__FEE_RATE : r.__FEE;
  if (i === void 0) throw new Error(`Failed to calculate ${e}`);
  return i;
}
function mo(t, e, n, r, c, s, o = !0, i) {
  const a = br(n);
  if (!wo(e, n, a) && o)
    throw new Error(`Can not finalize input #${t}`);
  return Io(
    n,
    a,
    e.partialSig,
    r,
    c,
    s,
    i
  );
}
function Io(t, e, n, r, c, s, o) {
  let i, a;
  const u = ko(t, e, n), f = s ? be({ redeem: u }) : null, h = c ? Ee({ redeem: f || u }) : null;
  return r ? (f ? a = ye(f.witness) : u ? a = ye(u.witness) : a = ye(o ?? [l.from([0])]), h && (i = h?.input)) : h ? i = h?.input : u ? i = u.input : i = Array.isArray(o) && o[0] ? o[0] : l.from([1]), {
    finalScriptSig: i,
    finalScriptWitness: a
  };
}
function An(t, e, n, r, c) {
  const s = x.checkForInput(t, e), { hash: o, sighashType: i, script: a } = lr(
    e,
    s,
    r,
    !1,
    c
  );
  return fr(n, a, "sign"), {
    hash: o,
    sighashType: i
  };
}
function lr(t, e, n, r, c) {
  const s = n.__TX, o = e.sighashType || v.SIGHASH_ALL;
  gr(o, c);
  let i, a;
  if (e.nonWitnessUtxo) {
    const h = Je(n, e, t), g = s.ins[t].hash, y = h.getHash();
    if (!g.equals(y))
      throw new Error(
        `Non-witness UTXO hash for input #${t} doesn't match the hash specified in the prevout`
      );
    const p = s.ins[t].index;
    a = h.outs[p];
  } else if (e.witnessUtxo)
    a = e.witnessUtxo;
  else
    throw new Error("Need a Utxo input item for signing");
  const { meaningfulScript: u, type: f } = Qe(
    a.script,
    t,
    "input",
    e.redeemScript,
    e.witnessScript
  );
  if (["p2sh-p2wsh", "p2wsh"].indexOf(f) >= 0)
    i = s.hashForWitnessV0(
      t,
      u,
      a.value,
      o
    );
  else if ($e(u)) {
    const h = we({
      hash: u.subarray(2)
    }).output;
    i = s.hashForWitnessV0(t, h, a.value, o);
  } else {
    if (e.nonWitnessUtxo === void 0 && !n.__UNSAFE_SIGN_NONSEGWIT)
      throw new Error(
        `Input #${t} has witnessUtxo but non-segwit script: ` + u.toString("hex")
      );
    !r && n.__UNSAFE_SIGN_NONSEGWIT && console.warn(
      `Warning: Signing non-segwit inputs without the full parent transaction means there is a chance that a miner could feed you incorrect information to trick you into paying large fees. This behavior is the same as Psbt's predecessor (TransactionBuilder - now removed) when signing non-segwit scripts. You are not able to export this Psbt with toBuffer|toBase64|toHex since it is not BIP174 compliant.
*********************
PROCEED WITH CAUTION!
*********************`
    ), i = s.hashForSignature(t, u, o);
  }
  return {
    script: u,
    sighashType: o,
    hash: i
  };
}
function vo(t, e, n, r) {
  const c = [];
  if (e.tapInternalKey) {
    const o = dr(t, e, r);
    o && c.push(o);
  }
  if (e.tapScriptSig) {
    const o = e.tapScriptSig.map((i) => i.pubkey);
    c.push(...o);
  }
  return c.map(
    (o) => wt(t, e, n, o, r)
  ).flat();
}
function dr(t, e, n) {
  const { script: r } = kt(t, e, n);
  return vt(r) ? l.from(r.subarray(2, 34)) : null;
}
function xn(t) {
  return t.length === 64 ? t : l.from(t.subarray(0, 64));
}
function wt(t, e, n, r, c, s, o) {
  const i = c.__TX, a = e.sighashType || v.SIGHASH_DEFAULT;
  gr(a, o);
  const u = n.map(
    (p, d) => kt(d, p, c)
  ), f = u.map((p) => p.script), h = u.map((p) => p.value), g = [];
  if (e.tapInternalKey && !s) {
    const p = dr(t, e, c) || l.from([]);
    if (ge(r).equals(p)) {
      const d = i.hashForWitnessV1(
        t,
        f,
        h,
        a
      );
      g.push({ pubkey: r, hash: d });
    }
  }
  const y = (e.tapLeafScript || []).filter((p) => Ze(r, p.script)).map((p) => {
    const d = qe({
      output: p.script,
      version: p.leafVersion
    });
    return Object.assign({ hash: d }, p);
  }).filter((p) => !s || s.equals(p.hash)).map((p) => {
    const d = i.hashForWitnessV1(
      t,
      f,
      h,
      a,
      p.hash
    );
    return {
      pubkey: r,
      hash: d,
      leafHash: p.hash
    };
  });
  return g.concat(y);
}
function gr(t, e) {
  if (e && e.indexOf(t) < 0) {
    const n = Co(t);
    throw new Error(
      `Sighash type is not allowed. Retry the sign method passing the sighashTypes array of whitelisted types. Sighash type: ${n}`
    );
  }
}
function ko(t, e, n) {
  let r;
  switch (e) {
    case "multisig": {
      const c = xo(t, n);
      r = De({
        output: t,
        signatures: c
      });
      break;
    }
    case "pubkey":
      r = Pn({
        output: t,
        signature: n[0].signature
      });
      break;
    case "pubkeyhash":
      r = we({
        output: t,
        pubkey: n[0].pubkey,
        signature: n[0].signature
      });
      break;
    case "witnesspubkeyhash":
      r = Xe({
        output: t,
        pubkey: n[0].pubkey,
        signature: n[0].signature
      });
      break;
  }
  return r;
}
function Ao(t, e, n) {
  const r = n.__TX, c = {
    script: null,
    isSegwit: !1,
    isP2SH: !1,
    isP2WSH: !1
  };
  if (c.isP2SH = !!e.redeemScript, c.isP2WSH = !!e.witnessScript, e.witnessScript)
    c.script = e.witnessScript;
  else if (e.redeemScript)
    c.script = e.redeemScript;
  else if (e.nonWitnessUtxo) {
    const s = Je(n, e, t), o = r.ins[t].index;
    c.script = s.outs[o].script;
  } else e.witnessUtxo && (c.script = e.witnessUtxo.script);
  if (e.witnessScript || $e(c.script))
    c.isSegwit = !0;
  else
    try {
      const s = c.script;
      if (!s) throw new TypeError("Invalid script for segwit address");
      c.isSegwit = Yn(s);
    } catch {
    }
  return c;
}
function Cn(t, e, n) {
  const r = x.checkForInput(e, t);
  if (!r.bip32Derivation || r.bip32Derivation.length === 0)
    throw new Error("Need bip32Derivation to sign with HD");
  const c = r.bip32Derivation.map((s) => {
    if (s.masterFingerprint.equals(n.fingerprint))
      return s;
  }).filter((s) => !!s);
  if (c.length === 0)
    throw new Error(
      "Need one bip32Derivation masterFingerprint to match the HDSigner fingerprint"
    );
  return c.map((s) => {
    const o = n.derivePath(s.path);
    if (!s.pubkey.equals(o.publicKey))
      throw new Error("pubkey did not match bip32Derivation");
    return o;
  });
}
function xo(t, e) {
  return De({ output: t }).pubkeys.map((r) => (e.filter((c) => c.pubkey.equals(r))[0] || {}).signature).filter((r) => !!r);
}
function yr(t) {
  let e = 0;
  function n(o) {
    return e += o, t.subarray(e - o, e);
  }
  function r() {
    const o = Tn(t, e);
    return e += Tn.bytes, o;
  }
  function c() {
    return n(r());
  }
  function s() {
    const o = r(), i = [];
    for (let a = 0; a < o; a++) i.push(c());
    return i;
  }
  return s();
}
function Co(t) {
  let e = t & v.SIGHASH_ANYONECANPAY ? "SIGHASH_ANYONECANPAY | " : "";
  switch (t & 31) {
    case v.SIGHASH_ALL:
      e += "SIGHASH_ALL";
      break;
    case v.SIGHASH_SINGLE:
      e += "SIGHASH_SINGLE";
      break;
    case v.SIGHASH_NONE:
      e += "SIGHASH_NONE";
      break;
  }
  return e;
}
function Et(t, e, n) {
  t.__NON_WITNESS_UTXO_BUF_CACHE[n] = e.nonWitnessUtxo, t.__NON_WITNESS_UTXO_TX_CACHE[n] = v.fromBuffer(e.nonWitnessUtxo);
  const r = t, c = n;
  delete e.nonWitnessUtxo, Object.defineProperty(e, "nonWitnessUtxo", {
    enumerable: !0,
    get() {
      const s = r.__NON_WITNESS_UTXO_BUF_CACHE[c], o = r.__NON_WITNESS_UTXO_TX_CACHE[c];
      if (s !== void 0)
        return s;
      {
        const i = o.toBuffer();
        return r.__NON_WITNESS_UTXO_BUF_CACHE[c] = i, i;
      }
    },
    set(s) {
      r.__NON_WITNESS_UTXO_BUF_CACHE[c] = s;
    }
  });
}
function Sr(t, e, n, r, c) {
  let s = 0;
  t.forEach((u, f) => {
    if (r && u.finalScriptSig && (e.ins[f].script = u.finalScriptSig), r && u.finalScriptWitness && (e.ins[f].witness = yr(u.finalScriptWitness)), u.witnessUtxo)
      s += u.witnessUtxo.value;
    else if (u.nonWitnessUtxo) {
      const h = Je(n, u, f), g = e.ins[f].index, y = h.outs[g];
      s += y.value;
    }
  });
  const o = e.outs.reduce((u, f) => u + f.value, 0), i = s - o;
  if (!c && i < 0)
    throw new Error(
      `Outputs are spending more than Inputs ${s} < ${o}`
    );
  const a = e.virtualSize();
  n.__FEE = i, n.__EXTRACTED_TX = e, n.__FEE_RATE = Math.floor(i / a);
}
function Je(t, e, n) {
  const r = t.__NON_WITNESS_UTXO_TX_CACHE;
  return r[n] || Et(t, e, n), r[n];
}
function wr(t, e, n) {
  const { script: r } = kt(t, e, n);
  return r;
}
function kt(t, e, n) {
  if (e.witnessUtxo !== void 0)
    return {
      script: e.witnessUtxo.script,
      value: e.witnessUtxo.value
    };
  if (e.nonWitnessUtxo !== void 0) {
    const c = Je(n, e, t).outs[n.__TX.ins[t].index];
    return { script: c.script, value: c.value };
  } else
    throw new Error("Can't find pubkey in input without Utxo data");
}
function Oo(t, e, n, r) {
  const c = wr(n, e, r), { meaningfulScript: s } = Qe(
    c,
    n,
    "input",
    e.redeemScript,
    e.witnessScript
  );
  return Ze(t, s);
}
function Po(t, e, n, r) {
  const c = r.__TX.outs[n].script, { meaningfulScript: s } = Qe(
    c,
    n,
    "output",
    e.redeemScript,
    e.witnessScript
  );
  return Ze(t, s);
}
function No(t) {
  if (!t) return;
  const e = ue(t);
  if (!e) return;
  const n = e[e.length - 1];
  if (!(!l.isBuffer(n) || Er(n) || Bo(n) || !ue(n)))
    return n;
}
function Uo(t) {
  if (!t) return;
  const e = yr(t), n = e[e.length - 1];
  if (!(Er(n) || !ue(n)))
    return n;
}
function Ho(t) {
  if (t.length === 65) {
    const e = t[64] & 1, n = l.from(t.subarray(0, 33));
    return n[0] = 2 | e, n;
  }
  return l.from(t);
}
function Er(t) {
  return t.length === 33 && kr(t);
}
function Bo(t) {
  return Un(t);
}
function Qe(t, e, n, r, c) {
  const s = or(t), o = s && r && Sn(r), i = Sn(t);
  if (s && r === void 0)
    throw new Error("scriptPubkey is P2SH but redeemScript missing");
  if ((i || o) && c === void 0)
    throw new Error("scriptPubkey or redeemScript is P2WSH but witnessScript missing");
  let a;
  return o ? (a = c, In(e, t, r, n), vn(e, r, c, n), Ge(a)) : i ? (a = c, vn(e, t, c, n), Ge(a)) : s ? (a = r, In(e, t, r, n)) : a = t, {
    meaningfulScript: a,
    type: o ? "p2sh-p2wsh" : s ? "p2sh" : i ? "p2wsh" : "raw"
  };
}
function Ge(t) {
  if ($e(t) || or(t))
    throw new Error("P2WPKH or P2SH can not be contained within P2WSH");
}
function br(t) {
  return $e(t) ? "witnesspubkeyhash" : $s(t) ? "pubkeyhash" : zs(t) ? "multisig" : Ys(t) ? "pubkey" : "nonstandard";
}
function le(t) {
  return [...Array(t).keys()];
}
export {
  ze as A,
  Ye as B,
  Yn as C,
  $n as D,
  Jn as E,
  _e as F,
  Is as G,
  Qn as H,
  er as I,
  tr as J,
  zs as K,
  Ys as L,
  $s as M,
  $e as N,
  Sn as O,
  St as P,
  or as Q,
  vt as R,
  Xo as S,
  Do as T,
  ye as U,
  ir as V,
  Ze as W,
  Zs as X,
  cr as Y,
  Qs as Z,
  Zn as _,
  _s as a,
  Ve as b,
  Go as c,
  K as d,
  Io as e,
  Kn as f,
  mo as g,
  wn as h,
  Fn as i,
  N as j,
  ct as k,
  En as l,
  bn as m,
  Wo as n,
  qo as o,
  ke as p,
  Vo as q,
  no as r,
  Re as s,
  ge as t,
  ro as u,
  io as v,
  me as w,
  Ie as x,
  zn as y,
  We as z
};
