// @btc-vision/visitor-as/dist/decorator.js
import { PathTransformVisitor, mergeTransformer } from "./transformer.js";
import { decorates, not, isStdlib } from "./utils.js";
export function registerDecorator(decorator) {
    TopLevelDecorator.registerVisitor(decorator);
    return TopLevelDecorator;
}
export class TopLevelDecorator extends PathTransformVisitor {
    static _visitor;
    static registerVisitor(visitor) {
        TopLevelDecorator._visitor = visitor;
    }
    get visitor() {
        return TopLevelDecorator._visitor;
    }
    visitDecoratorNode(node) {
        if (this.visitor.decoratorMatcher(node)) {
            this.visitor.currentPath = this.currentParentPath;
            this.visitor.visit(this.currentParent);
        }
    }
    afterParse(_) {
        mergeTransformer(this, this.visitor);
        this.visit(this.program.sources.filter(this.visitor.sourceFilter));
    }
}
export class Decorator extends PathTransformVisitor {
    /**
     * Default filter that removes library files
     */
    get sourceFilter() {
        return not(isStdlib);
    }
    get decoratorMatcher() {
        return (node) => decorates(node, this.name);
    }
    get name() { return ""; }
    getDecorator(node) {
        return node.decorators && node.decorators.find(this.decoratorMatcher) || null;
    }
}
export class ClassDecorator extends Decorator {
}
export class FunctionDecorator extends Decorator {
}
export class VariableDecorator extends Decorator {
}
