// @btc-vision/bitcoin/build/psbt/psbtutils.js
import * as varuint from 'bip174/src/lib/converter/varint.js';
import { hash160 } from '../crypto.js';
import { p2ms } from '../payments/p2ms.js';
import { p2pk } from '../payments/p2pk.js';
import { p2pkh } from '../payments/p2pkh.js';
import { p2sh } from '../payments/p2sh.js';
import { p2tr } from '../payments/p2tr.js';
import { p2wpkh } from '../payments/p2wpkh.js';
import { p2wsh } from '../payments/p2wsh.js';
import { p2op } from '../payments/p2op.js';
import { decompressPublicKey, pubkeysMatch, toXOnly } from '../pubkey.js';
import * as bscript from '../script.js';
import { Transaction } from '../transaction.js';
function isPaymentFactory(payment) {
    return (script) => {
        try {
            payment({ output: script });
            return true;
        }
        catch {
            return false;
        }
    };
}
export const isP2MS = isPaymentFactory(p2ms);
export const isP2PK = isPaymentFactory(p2pk);
export const isP2PKH = isPaymentFactory(p2pkh);
export const isP2WPKH = isPaymentFactory(p2wpkh);
export const isP2WSHScript = isPaymentFactory(p2wsh);
export const isP2SHScript = isPaymentFactory(p2sh);
export const isP2TR = isPaymentFactory(p2tr);
export const isP2OP = isPaymentFactory(p2op);
export const isP2A = (script) => {
    return (script.length === 4 &&
        script[0] === 0x51 &&
        script[1] === 0x02 &&
        script[2] === 0x4e &&
        script[3] === 0x73);
};
export function witnessStackToScriptWitness(witness) {
    let buffer = Buffer.allocUnsafe(0);
    function writeSlice(slice) {
        buffer = Buffer.concat([buffer, Buffer.from(slice)]);
    }
    function writeVarInt(i) {
        const currentLen = buffer.length;
        const varintLen = varuint.encodingLength(i);
        buffer = Buffer.concat([buffer, Buffer.allocUnsafe(varintLen)]);
        varuint.encode(i, buffer, currentLen);
    }
    function writeVarSlice(slice) {
        writeVarInt(slice.length);
        writeSlice(slice);
    }
    function writeVector(vector) {
        writeVarInt(vector.length);
        vector.forEach(writeVarSlice);
    }
    writeVector(witness);
    return buffer;
}
export function pubkeyPositionInScript(pubkey, script) {
    const decompiled = bscript.decompile(script);
    if (decompiled === null)
        throw new Error('Unknown script error');
    const pubkeyHash = hash160(pubkey);
    const pubkeyXOnly = toXOnly(pubkey);
    const uncompressed = decompressPublicKey(pubkey);
    const pubkeyHybridHash = uncompressed?.hybrid ? hash160(uncompressed.hybrid) : undefined;
    const pubkeyUncompressedHash = uncompressed?.uncompressed
        ? hash160(uncompressed.uncompressed)
        : undefined;
    return decompiled.findIndex((element) => {
        if (typeof element === 'number')
            return false;
        if (pubkeysMatch(element, pubkey))
            return true;
        if (pubkeysMatch(element, pubkeyXOnly))
            return true;
        if (element.equals(pubkeyHash)) {
            return true;
        }
        if (uncompressed) {
            if (pubkeysMatch(element, uncompressed.uncompressed))
                return true;
            if (pubkeysMatch(element, uncompressed.hybrid))
                return true;
            if ((pubkeyHybridHash && element.equals(pubkeyHybridHash)) ||
                (pubkeyUncompressedHash && element.equals(pubkeyUncompressedHash))) {
                return true;
            }
        }
    });
}
export function pubkeyInScript(pubkey, script) {
    return pubkeyPositionInScript(pubkey, script) !== -1;
}
export function checkInputForSig(input, action) {
    const pSigs = extractPartialSigs(input);
    return pSigs.some((pSig) => signatureBlocksAction(pSig, bscript.signature.decode, action));
}
export function signatureBlocksAction(signature, signatureDecodeFn, action) {
    const { hashType } = signatureDecodeFn(signature);
    const whitelist = [];
    const isAnyoneCanPay = hashType & Transaction.SIGHASH_ANYONECANPAY;
    if (isAnyoneCanPay)
        whitelist.push('addInput');
    const hashMod = hashType & 0x1f;
    switch (hashMod) {
        case Transaction.SIGHASH_ALL:
            break;
        case Transaction.SIGHASH_SINGLE:
        case Transaction.SIGHASH_NONE:
            whitelist.push('addOutput');
            whitelist.push('setInputSequence');
            break;
    }
    return whitelist.indexOf(action) === -1;
}
function extractPartialSigs(input) {
    const { partialSig } = input;
    let pSigs;
    if (!partialSig || partialSig.length === 0) {
        if (!input.finalScriptSig && !input.finalScriptWitness)
            return [];
        pSigs = getPsigsFromInputFinalScripts(input);
    }
    else {
        pSigs = partialSig;
    }
    return pSigs.map((p) => p.signature);
}
export function getPsigsFromInputFinalScripts(input) {
    const scriptItems = !input.finalScriptSig ? [] : bscript.decompile(input.finalScriptSig) || [];
    const witnessItems = !input.finalScriptWitness
        ? []
        : bscript.decompile(input.finalScriptWitness) || [];
    return scriptItems
        .concat(witnessItems)
        .filter((item) => {
        return Buffer.isBuffer(item) && bscript.isCanonicalScriptSignature(item);
    })
        .map((sig) => ({ signature: sig }));
}
