// @btc-vision/transaction/build/mnemonic/BIPStandard.js
export var BIPStandard;
(function (BIPStandard) {
    BIPStandard[BIPStandard["BIP44"] = 44] = "BIP44";
    BIPStandard[BIPStandard["BIP49"] = 49] = "BIP49";
    BIPStandard[BIPStandard["BIP84"] = 84] = "BIP84";
    BIPStandard[BIPStandard["BIP86"] = 86] = "BIP86";
})(BIPStandard || (BIPStandard = {}));
export function getBIPDescription(standard) {
    switch (standard) {
        case BIPStandard.BIP44:
            return 'BIP44: Legacy addresses (P2PKH), widely used by Unisat and other wallets';
        case BIPStandard.BIP49:
            return 'BIP49: Wrapped SegWit addresses (P2SH-P2WPKH)';
        case BIPStandard.BIP84:
            return 'BIP84: Native SegWit addresses (P2WPKH) - DEFAULT';
        case BIPStandard.BIP86:
            return 'BIP86: Taproot addresses (P2TR)';
        default:
            return 'Unknown BIP standard';
    }
}
export function buildBIPPath(standard, coinType, account, change, addressIndex) {
    return `m/${standard}'/${coinType}'/${account}'/${change}/${addressIndex}`;
}
