// @btc-vision/bitcoin/browser/index.js
import { c as s, n as a } from "./chunks/psbt-BCNk7JUx.js";
import { y as P, z as I, F as m, x as _, w as F, A as f, B as h, P as R, _ as k, X as A, l as U, u as d, m as l, a as O, E as g, f as y, I as B, g as H, Z as N, v as M, T as W, K as w, S as G, L as b, M as K, Q as L, R as V, N as x, O as X, j as D, k as C, C as v, e as z, W as Z, V as j, p as q, s as J, Y as Q, h as Y, r as $, q as ss, G as as, H as ts, D as rs, J as es, t as os, o as is, U as ps } from "./chunks/psbt-BCNk7JUx.js";
import { i as t } from "./chunks/payments-B1wlSccx.js";
import { L as ns, M as Ss, P as Ts, l as Es, j as us, r as Ps, m as Is, h as ms, k as _s, t as Fs } from "./chunks/payments-B1wlSccx.js";
import { g as r } from "./chunks/script-DyPItFEl.js";
import { o as hs } from "./chunks/script-DyPItFEl.js";
import { c as e } from "./chunks/crypto-BhCpKpek.js";
import { e as ks, T as As, h as Us, b as ds, r as ls, d as Os, s as gs, t as ys } from "./chunks/crypto-BhCpKpek.js";
import { a as o } from "./chunks/transaction-C_UbhMGn.js";
import { B as Hs, T as Ns } from "./chunks/transaction-C_UbhMGn.js";
import { B as Ws, n as ws, T as Gs, c as bs, E as Ks, r as Ls, D as Vs, G as xs } from "./chunks/utils-DNZi-T5W.js";
const T = {
  networks: a,
  address: s,
  payments: t,
  script: r,
  crypto: e,
  Transaction: o
};
export {
  Hs as Block,
  Ws as BufferReader,
  ws as BufferWriter,
  P as FUTURE_MAX_VERSION,
  I as FUTURE_OPNET_VERSION,
  m as FUTURE_SEGWIT_MAX_SIZE,
  _ as FUTURE_SEGWIT_MAX_VERSION,
  F as FUTURE_SEGWIT_MIN_SIZE,
  f as FUTURE_SEGWIT_MIN_VERSION,
  h as FUTURE_SEGWIT_VERSION_DIFF,
  ns as LEAF_VERSION_TAPSCRIPT,
  Ss as MAX_TAPTREE_DEPTH,
  Ts as PaymentType,
  R as Psbt,
  ks as TAGGED_HASH_PREFIXES,
  As as TAGS,
  Gs as TAPLEAF_VERSION_MASK,
  Ns as Transaction,
  k as _toFutureSegwitAddress,
  s as address,
  A as checkInputForSig,
  U as checkTaprootInputFields,
  d as checkTaprootInputForSigs,
  l as checkTaprootOutputFields,
  bs as cloneBuffer,
  e as crypto,
  O as decompressPublicKey,
  T as default,
  Es as findScriptPath,
  g as fromBase58Check,
  y as fromBech32,
  B as fromOutputScript,
  H as getFinalScripts,
  N as getPsigsFromInputFinalScripts,
  M as getTapKeySigFromWitness,
  Us as hash160,
  ds as hash256,
  us as initEccLib,
  W as isP2A,
  w as isP2MS,
  G as isP2OP,
  b as isP2PK,
  K as isP2PKH,
  L as isP2SHScript,
  V as isP2TR,
  x as isP2WPKH,
  X as isP2WSHScript,
  D as isTaprootInput,
  C as isTaprootOutput,
  v as isUnknownSegwitVersion,
  a as networks,
  hs as opcodes,
  t as payments,
  z as prepareFinalScripts,
  Z as pubkeyInScript,
  j as pubkeyPositionInScript,
  q as pubkeysMatch,
  Ks as readUInt64LE,
  Ls as reverseBuffer,
  ls as ripemd160,
  Ps as rootHashFromPath,
  r as script,
  J as serializeTaprootSignature,
  Os as sha1,
  gs as sha256,
  Q as signatureBlocksAction,
  ys as taggedHash,
  Y as tapScriptFinalizer,
  $ as tapTreeFromList,
  ss as tapTreeToList,
  Is as tapTweakHash,
  ms as tapleafHash,
  as as toBase58Check,
  ts as toBech32,
  rs as toFutureOPNetAddress,
  _s as toHashTree,
  es as toOutputScript,
  os as toXOnly,
  is as tweakInternalPubKey,
  Fs as tweakKey,
  Vs as varuint,
  ps as witnessStackToScriptWitness,
  xs as writeUInt64LE
};
