// opnet/build/providers/websocket/types/ConnectionState.js
export var ConnectionState;
(function (ConnectionState) {
    ConnectionState[ConnectionState["DISCONNECTED"] = 0] = "DISCONNECTED";
    ConnectionState[ConnectionState["CONNECTING"] = 1] = "CONNECTING";
    ConnectionState[ConnectionState["CONNECTED"] = 2] = "CONNECTED";
    ConnectionState[ConnectionState["HANDSHAKING"] = 3] = "HANDSHAKING";
    ConnectionState[ConnectionState["READY"] = 4] = "READY";
    ConnectionState[ConnectionState["RECONNECTING"] = 5] = "RECONNECTING";
    ConnectionState[ConnectionState["CLOSING"] = 6] = "CLOSING";
})(ConnectionState || (ConnectionState = {}));
export function getConnectionStateName(state) {
    switch (state) {
        case ConnectionState.DISCONNECTED:
            return 'DISCONNECTED';
        case ConnectionState.CONNECTING:
            return 'CONNECTING';
        case ConnectionState.CONNECTED:
            return 'CONNECTED';
        case ConnectionState.HANDSHAKING:
            return 'HANDSHAKING';
        case ConnectionState.READY:
            return 'READY';
        case ConnectionState.RECONNECTING:
            return 'RECONNECTING';
        case ConnectionState.CLOSING:
            return 'CLOSING';
        default:
            return 'UNKNOWN';
    }
}
