// opnet/build/providers/websocket/types/WebSocketErrorCodes.js
export var ProtocolError;
(function (ProtocolError) {
    ProtocolError[ProtocolError["MALFORMED_MESSAGE"] = 1] = "MALFORMED_MESSAGE";
    ProtocolError[ProtocolError["UNKNOWN_OPCODE"] = 2] = "UNKNOWN_OPCODE";
    ProtocolError[ProtocolError["HANDSHAKE_REQUIRED"] = 3] = "HANDSHAKE_REQUIRED";
    ProtocolError[ProtocolError["HANDSHAKE_ALREADY_COMPLETED"] = 4] = "HANDSHAKE_ALREADY_COMPLETED";
    ProtocolError[ProtocolError["UNSUPPORTED_PROTOCOL_VERSION"] = 5] = "UNSUPPORTED_PROTOCOL_VERSION";
    ProtocolError[ProtocolError["PAYLOAD_TOO_LARGE"] = 6] = "PAYLOAD_TOO_LARGE";
    ProtocolError[ProtocolError["INVALID_REQUEST_ID"] = 7] = "INVALID_REQUEST_ID";
    ProtocolError[ProtocolError["CONNECTION_CLOSING"] = 8] = "CONNECTION_CLOSING";
    ProtocolError[ProtocolError["RATE_LIMIT_EXCEEDED"] = 9] = "RATE_LIMIT_EXCEEDED";
    ProtocolError[ProtocolError["TOO_MANY_PENDING_REQUESTS"] = 10] = "TOO_MANY_PENDING_REQUESTS";
    ProtocolError[ProtocolError["REQUEST_TIMEOUT"] = 11] = "REQUEST_TIMEOUT";
    ProtocolError[ProtocolError["INVALID_MESSAGE_FORMAT"] = 12] = "INVALID_MESSAGE_FORMAT";
})(ProtocolError || (ProtocolError = {}));
export var AuthError;
(function (AuthError) {
    AuthError[AuthError["AUTHENTICATION_REQUIRED"] = 1000] = "AUTHENTICATION_REQUIRED";
    AuthError[AuthError["INVALID_CREDENTIALS"] = 1001] = "INVALID_CREDENTIALS";
    AuthError[AuthError["SESSION_EXPIRED"] = 1002] = "SESSION_EXPIRED";
    AuthError[AuthError["PERMISSION_DENIED"] = 1003] = "PERMISSION_DENIED";
    AuthError[AuthError["INVALID_CLIENT_INFO"] = 1004] = "INVALID_CLIENT_INFO";
})(AuthError || (AuthError = {}));
export var ResourceError;
(function (ResourceError) {
    ResourceError[ResourceError["NOT_FOUND"] = 2000] = "NOT_FOUND";
    ResourceError[ResourceError["BLOCK_NOT_FOUND"] = 2001] = "BLOCK_NOT_FOUND";
    ResourceError[ResourceError["TRANSACTION_NOT_FOUND"] = 2002] = "TRANSACTION_NOT_FOUND";
    ResourceError[ResourceError["ADDRESS_NOT_FOUND"] = 2003] = "ADDRESS_NOT_FOUND";
    ResourceError[ResourceError["CONTRACT_NOT_FOUND"] = 2004] = "CONTRACT_NOT_FOUND";
    ResourceError[ResourceError["EPOCH_NOT_FOUND"] = 2005] = "EPOCH_NOT_FOUND";
    ResourceError[ResourceError["SUBSCRIPTION_NOT_FOUND"] = 2006] = "SUBSCRIPTION_NOT_FOUND";
    ResourceError[ResourceError["ALREADY_EXISTS"] = 2100] = "ALREADY_EXISTS";
    ResourceError[ResourceError["SUBSCRIPTION_ALREADY_EXISTS"] = 2101] = "SUBSCRIPTION_ALREADY_EXISTS";
    ResourceError[ResourceError["MAX_SUBSCRIPTIONS_REACHED"] = 2102] = "MAX_SUBSCRIPTIONS_REACHED";
})(ResourceError || (ResourceError = {}));
export var ValidationError;
(function (ValidationError) {
    ValidationError[ValidationError["INVALID_PARAMS"] = 3000] = "INVALID_PARAMS";
    ValidationError[ValidationError["MISSING_REQUIRED_FIELD"] = 3001] = "MISSING_REQUIRED_FIELD";
    ValidationError[ValidationError["VALUE_OUT_OF_RANGE"] = 3002] = "VALUE_OUT_OF_RANGE";
    ValidationError[ValidationError["INVALID_ADDRESS"] = 3003] = "INVALID_ADDRESS";
    ValidationError[ValidationError["INVALID_HASH"] = 3004] = "INVALID_HASH";
    ValidationError[ValidationError["INVALID_BLOCK_IDENTIFIER"] = 3005] = "INVALID_BLOCK_IDENTIFIER";
    ValidationError[ValidationError["INVALID_TRANSACTION_DATA"] = 3006] = "INVALID_TRANSACTION_DATA";
    ValidationError[ValidationError["INVALID_SIGNATURE"] = 3007] = "INVALID_SIGNATURE";
    ValidationError[ValidationError["INVALID_CALLDATA"] = 3008] = "INVALID_CALLDATA";
    ValidationError[ValidationError["INVALID_BLOCK_HEIGHT"] = 3009] = "INVALID_BLOCK_HEIGHT";
    ValidationError[ValidationError["INVALID_EPOCH_NUMBER"] = 3010] = "INVALID_EPOCH_NUMBER";
    ValidationError[ValidationError["INVALID_POINTER"] = 3011] = "INVALID_POINTER";
    ValidationError[ValidationError["INVALID_PUBLIC_KEY"] = 3012] = "INVALID_PUBLIC_KEY";
})(ValidationError || (ValidationError = {}));
export var InternalError;
(function (InternalError) {
    InternalError[InternalError["INTERNAL_ERROR"] = 4000] = "INTERNAL_ERROR";
    InternalError[InternalError["DATABASE_ERROR"] = 4001] = "DATABASE_ERROR";
    InternalError[InternalError["STORAGE_ERROR"] = 4002] = "STORAGE_ERROR";
    InternalError[InternalError["SERIALIZATION_ERROR"] = 4003] = "SERIALIZATION_ERROR";
    InternalError[InternalError["DESERIALIZATION_ERROR"] = 4004] = "DESERIALIZATION_ERROR";
    InternalError[InternalError["VM_ERROR"] = 4005] = "VM_ERROR";
    InternalError[InternalError["NETWORK_ERROR"] = 4006] = "NETWORK_ERROR";
    InternalError[InternalError["SERVICE_UNAVAILABLE"] = 4007] = "SERVICE_UNAVAILABLE";
    InternalError[InternalError["NOT_IMPLEMENTED"] = 4008] = "NOT_IMPLEMENTED";
    InternalError[InternalError["TIMEOUT"] = 4009] = "TIMEOUT";
})(InternalError || (InternalError = {}));
export const ErrorMessages = {
    [ProtocolError.MALFORMED_MESSAGE]: 'Malformed message',
    [ProtocolError.UNKNOWN_OPCODE]: 'Unknown opcode',
    [ProtocolError.HANDSHAKE_REQUIRED]: 'Handshake required before making requests',
    [ProtocolError.HANDSHAKE_ALREADY_COMPLETED]: 'Handshake has already been completed',
    [ProtocolError.UNSUPPORTED_PROTOCOL_VERSION]: 'Unsupported protocol version',
    [ProtocolError.PAYLOAD_TOO_LARGE]: 'Message payload too large',
    [ProtocolError.INVALID_REQUEST_ID]: 'Invalid or missing request ID',
    [ProtocolError.CONNECTION_CLOSING]: 'Connection is closing',
    [ProtocolError.RATE_LIMIT_EXCEEDED]: 'Rate limit exceeded',
    [ProtocolError.TOO_MANY_PENDING_REQUESTS]: 'Too many pending requests',
    [ProtocolError.REQUEST_TIMEOUT]: 'Request timed out',
    [ProtocolError.INVALID_MESSAGE_FORMAT]: 'Invalid message format',
    [AuthError.AUTHENTICATION_REQUIRED]: 'Authentication required',
    [AuthError.INVALID_CREDENTIALS]: 'Invalid credentials',
    [AuthError.SESSION_EXPIRED]: 'Session has expired',
    [AuthError.PERMISSION_DENIED]: 'Permission denied',
    [AuthError.INVALID_CLIENT_INFO]: 'Invalid client information',
    [ResourceError.NOT_FOUND]: 'Resource not found',
    [ResourceError.BLOCK_NOT_FOUND]: 'Block not found',
    [ResourceError.TRANSACTION_NOT_FOUND]: 'Transaction not found',
    [ResourceError.ADDRESS_NOT_FOUND]: 'Address not found',
    [ResourceError.CONTRACT_NOT_FOUND]: 'Contract not found',
    [ResourceError.EPOCH_NOT_FOUND]: 'Epoch not found',
    [ResourceError.SUBSCRIPTION_NOT_FOUND]: 'Subscription not found',
    [ResourceError.ALREADY_EXISTS]: 'Resource already exists',
    [ResourceError.SUBSCRIPTION_ALREADY_EXISTS]: 'Subscription already exists',
    [ResourceError.MAX_SUBSCRIPTIONS_REACHED]: 'Maximum subscriptions reached',
    [ValidationError.INVALID_PARAMS]: 'Invalid parameters',
    [ValidationError.MISSING_REQUIRED_FIELD]: 'Missing required field',
    [ValidationError.VALUE_OUT_OF_RANGE]: 'Value out of valid range',
    [ValidationError.INVALID_ADDRESS]: 'Invalid address format',
    [ValidationError.INVALID_HASH]: 'Invalid hash format',
    [ValidationError.INVALID_BLOCK_IDENTIFIER]: 'Invalid block identifier',
    [ValidationError.INVALID_TRANSACTION_DATA]: 'Invalid transaction data',
    [ValidationError.INVALID_SIGNATURE]: 'Invalid signature',
    [ValidationError.INVALID_CALLDATA]: 'Invalid calldata',
    [ValidationError.INVALID_BLOCK_HEIGHT]: 'Invalid block height',
    [ValidationError.INVALID_EPOCH_NUMBER]: 'Invalid epoch number',
    [ValidationError.INVALID_POINTER]: 'Invalid pointer format',
    [ValidationError.INVALID_PUBLIC_KEY]: 'Invalid public key',
    [InternalError.INTERNAL_ERROR]: 'Internal server error',
    [InternalError.DATABASE_ERROR]: 'Database error',
    [InternalError.STORAGE_ERROR]: 'Storage error',
    [InternalError.SERIALIZATION_ERROR]: 'Serialization error',
    [InternalError.DESERIALIZATION_ERROR]: 'Deserialization error',
    [InternalError.VM_ERROR]: 'VM execution error',
    [InternalError.NETWORK_ERROR]: 'Network error',
    [InternalError.SERVICE_UNAVAILABLE]: 'Service temporarily unavailable',
    [InternalError.NOT_IMPLEMENTED]: 'Method not implemented',
    [InternalError.TIMEOUT]: 'Operation timed out',
};
export function getErrorMessage(code) {
    return ErrorMessages[code] ?? 'Unknown error';
}
