// opnet/build/providers/websocket/types/WebSocketOpcodes.js
export var WebSocketRequestOpcode;
(function (WebSocketRequestOpcode) {
    WebSocketRequestOpcode[WebSocketRequestOpcode["PING"] = 0] = "PING";
    WebSocketRequestOpcode[WebSocketRequestOpcode["HANDSHAKE"] = 1] = "HANDSHAKE";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_BLOCK_NUMBER"] = 16] = "GET_BLOCK_NUMBER";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_BLOCK_BY_NUMBER"] = 17] = "GET_BLOCK_BY_NUMBER";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_BLOCK_BY_HASH"] = 18] = "GET_BLOCK_BY_HASH";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_BLOCK_BY_CHECKSUM"] = 19] = "GET_BLOCK_BY_CHECKSUM";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_BLOCK_WITNESS"] = 20] = "GET_BLOCK_WITNESS";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_GAS"] = 21] = "GET_GAS";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_TRANSACTION_BY_HASH"] = 32] = "GET_TRANSACTION_BY_HASH";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_TRANSACTION_RECEIPT"] = 33] = "GET_TRANSACTION_RECEIPT";
    WebSocketRequestOpcode[WebSocketRequestOpcode["BROADCAST_TRANSACTION"] = 34] = "BROADCAST_TRANSACTION";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_PREIMAGE"] = 35] = "GET_PREIMAGE";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_BALANCE"] = 48] = "GET_BALANCE";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_UTXOS"] = 49] = "GET_UTXOS";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_PUBLIC_KEY_INFO"] = 50] = "GET_PUBLIC_KEY_INFO";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_CHAIN_ID"] = 64] = "GET_CHAIN_ID";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_REORG"] = 65] = "GET_REORG";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_CODE"] = 80] = "GET_CODE";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_STORAGE_AT"] = 81] = "GET_STORAGE_AT";
    WebSocketRequestOpcode[WebSocketRequestOpcode["CALL"] = 82] = "CALL";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_LATEST_EPOCH"] = 96] = "GET_LATEST_EPOCH";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_EPOCH_BY_NUMBER"] = 97] = "GET_EPOCH_BY_NUMBER";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_EPOCH_BY_HASH"] = 98] = "GET_EPOCH_BY_HASH";
    WebSocketRequestOpcode[WebSocketRequestOpcode["GET_EPOCH_TEMPLATE"] = 99] = "GET_EPOCH_TEMPLATE";
    WebSocketRequestOpcode[WebSocketRequestOpcode["SUBMIT_EPOCH"] = 100] = "SUBMIT_EPOCH";
    WebSocketRequestOpcode[WebSocketRequestOpcode["SUBSCRIBE_BLOCKS"] = 112] = "SUBSCRIBE_BLOCKS";
    WebSocketRequestOpcode[WebSocketRequestOpcode["SUBSCRIBE_EPOCHS"] = 113] = "SUBSCRIBE_EPOCHS";
    WebSocketRequestOpcode[WebSocketRequestOpcode["UNSUBSCRIBE"] = 127] = "UNSUBSCRIBE";
})(WebSocketRequestOpcode || (WebSocketRequestOpcode = {}));
export var WebSocketResponseOpcode;
(function (WebSocketResponseOpcode) {
    WebSocketResponseOpcode[WebSocketResponseOpcode["ERROR"] = 128] = "ERROR";
    WebSocketResponseOpcode[WebSocketResponseOpcode["PONG"] = 129] = "PONG";
    WebSocketResponseOpcode[WebSocketResponseOpcode["HANDSHAKE_ACK"] = 130] = "HANDSHAKE_ACK";
    WebSocketResponseOpcode[WebSocketResponseOpcode["BLOCK_NUMBER"] = 144] = "BLOCK_NUMBER";
    WebSocketResponseOpcode[WebSocketResponseOpcode["BLOCK"] = 145] = "BLOCK";
    WebSocketResponseOpcode[WebSocketResponseOpcode["BLOCK_WITNESS"] = 146] = "BLOCK_WITNESS";
    WebSocketResponseOpcode[WebSocketResponseOpcode["GAS"] = 147] = "GAS";
    WebSocketResponseOpcode[WebSocketResponseOpcode["TRANSACTION"] = 160] = "TRANSACTION";
    WebSocketResponseOpcode[WebSocketResponseOpcode["TRANSACTION_RECEIPT"] = 161] = "TRANSACTION_RECEIPT";
    WebSocketResponseOpcode[WebSocketResponseOpcode["BROADCAST_RESULT"] = 162] = "BROADCAST_RESULT";
    WebSocketResponseOpcode[WebSocketResponseOpcode["PREIMAGE"] = 163] = "PREIMAGE";
    WebSocketResponseOpcode[WebSocketResponseOpcode["BALANCE"] = 176] = "BALANCE";
    WebSocketResponseOpcode[WebSocketResponseOpcode["UTXOS"] = 177] = "UTXOS";
    WebSocketResponseOpcode[WebSocketResponseOpcode["PUBLIC_KEY_INFO"] = 178] = "PUBLIC_KEY_INFO";
    WebSocketResponseOpcode[WebSocketResponseOpcode["CHAIN_ID"] = 192] = "CHAIN_ID";
    WebSocketResponseOpcode[WebSocketResponseOpcode["REORG"] = 193] = "REORG";
    WebSocketResponseOpcode[WebSocketResponseOpcode["CODE"] = 208] = "CODE";
    WebSocketResponseOpcode[WebSocketResponseOpcode["STORAGE"] = 209] = "STORAGE";
    WebSocketResponseOpcode[WebSocketResponseOpcode["CALL_RESULT"] = 210] = "CALL_RESULT";
    WebSocketResponseOpcode[WebSocketResponseOpcode["EPOCH"] = 224] = "EPOCH";
    WebSocketResponseOpcode[WebSocketResponseOpcode["EPOCH_TEMPLATE"] = 225] = "EPOCH_TEMPLATE";
    WebSocketResponseOpcode[WebSocketResponseOpcode["EPOCH_SUBMIT_RESULT"] = 226] = "EPOCH_SUBMIT_RESULT";
    WebSocketResponseOpcode[WebSocketResponseOpcode["SUBSCRIPTION_CREATED"] = 240] = "SUBSCRIPTION_CREATED";
    WebSocketResponseOpcode[WebSocketResponseOpcode["UNSUBSCRIBE_RESULT"] = 241] = "UNSUBSCRIBE_RESULT";
    WebSocketResponseOpcode[WebSocketResponseOpcode["NEW_BLOCK_NOTIFICATION"] = 248] = "NEW_BLOCK_NOTIFICATION";
    WebSocketResponseOpcode[WebSocketResponseOpcode["NEW_EPOCH_NOTIFICATION"] = 249] = "NEW_EPOCH_NOTIFICATION";
})(WebSocketResponseOpcode || (WebSocketResponseOpcode = {}));
