// @btc-vision/bip32/src/cjs/quantum/index.cjs
"use strict";
/**
 * Quantum-resistant BIP32 implementation using ML-DSA
 *
 * This module provides hierarchical deterministic key derivation using
 * ML-DSA (FIPS 204) for post-quantum security.
 *
 * Key features:
 * - Uses BIP32 path derivation (e.g., m/360'/0'/0'/0/0)
 * - Supports ML-DSA-44 (default), ML-DSA-65, and ML-DSA-87
 * - ML-DSA-44 provides Level 2 security (128-bit classical security)
 * - ML-DSA-65 provides Level 3 security (192-bit classical security)
 * - ML-DSA-87 provides Level 5 security (256-bit classical security)
 * - Compatible with standard BIP32 mnemonic seeds
 *
 * Usage:
 * ```typescript
 * import {
 *   QuantumBIP32Factory,
 *   MLDSASecurityLevel,
 *   QuantumDerivationPath
 * } from '@btc-vision/bip32/quantum';
 *
 * const seed = ...; // Your BIP39 seed
 *
 * // Default: ML-DSA-44 (Level 2 security)
 * const master = QuantumBIP32Factory.fromSeed(seed);
 *
 * // Or specify security level explicitly using enum
 * const master87 = QuantumBIP32Factory.fromSeed(seed, MLDSASecurityLevel.LEVEL5);
 * const master65 = QuantumBIP32Factory.fromSeed(seed, MLDSASecurityLevel.LEVEL3);
 * const master44 = QuantumBIP32Factory.fromSeed(seed, MLDSASecurityLevel.LEVEL2);
 *
 * // Use standard derivation paths
 * const child = master.derivePath(QuantumDerivationPath.STANDARD);
 *
 * const signature = child.sign(messageHash);
 * const isValid = child.verify(messageHash, signature);
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBitcoinPath = exports.getQuantumPath = exports.QuantumDerivationPath = exports.DerivationPath = exports.getMLDSAConfig = exports.DEFAULT_SECURITY_LEVEL = exports.MLDSASecurityLevel = exports.QuantumBIP32Factory = void 0;
// Export the generic ML-DSA implementation (supports 44, 65, 87)
var mldsa_js_1 = require("./mldsa.cjs");
Object.defineProperty(exports, "QuantumBIP32Factory", { enumerable: true, get: function () { return mldsa_js_1.QuantumBIP32Factory; } });
var config_js_1 = require("./config.cjs");
Object.defineProperty(exports, "MLDSASecurityLevel", { enumerable: true, get: function () { return config_js_1.MLDSASecurityLevel; } });
Object.defineProperty(exports, "DEFAULT_SECURITY_LEVEL", { enumerable: true, get: function () { return config_js_1.DEFAULT_SECURITY_LEVEL; } });
Object.defineProperty(exports, "getMLDSAConfig", { enumerable: true, get: function () { return config_js_1.getMLDSAConfig; } });
// Re-export derivation path enums for convenience
var derivation_paths_js_1 = require("../derivation-paths.cjs");
Object.defineProperty(exports, "DerivationPath", { enumerable: true, get: function () { return derivation_paths_js_1.DerivationPath; } });
Object.defineProperty(exports, "QuantumDerivationPath", { enumerable: true, get: function () { return derivation_paths_js_1.QuantumDerivationPath; } });
Object.defineProperty(exports, "getQuantumPath", { enumerable: true, get: function () { return derivation_paths_js_1.getQuantumPath; } });
Object.defineProperty(exports, "getBitcoinPath", { enumerable: true, get: function () { return derivation_paths_js_1.getBitcoinPath; } });
